// sklearn/build/generated/impute/KNNImputer.js
// src/generated/impute/KNNImputer.ts
import crypto from "node:crypto";
var KNNImputer = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `KNNImputer${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("KNNImputer.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.impute import KNNImputer
try: bridgeKNNImputer
except NameError: bridgeKNNImputer = {}
`;
    await this._py.ex`ctor_KNNImputer = {'missing_values': ${this.opts["missing_values"] ?? void 0}, 'n_neighbors': ${this.opts["n_neighbors"] ?? void 0}, 'weights': ${this.opts["weights"] ?? void 0}, 'metric': ${this.opts["metric"] ?? void 0}, 'copy': ${this.opts["copy"] ?? void 0}, 'add_indicator': ${this.opts["add_indicator"] ?? void 0}, 'keep_empty_features': ${this.opts["keep_empty_features"] ?? void 0}}

ctor_KNNImputer = {k: v for k, v in ctor_KNNImputer.items() if v is not None}`;
    await this._py.ex`bridgeKNNImputer[${this.id}] = KNNImputer(**ctor_KNNImputer)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeKNNImputer[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the imputer on X.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("KNNImputer must call init() before fit()");
    }
    await this._py.ex`pms_KNNImputer_fit = {'X': ${opts["X"] ?? void 0}, 'y': ${opts["y"] ?? void 0}}

pms_KNNImputer_fit = {k: v for k, v in pms_KNNImputer_fit.items() if v is not None}`;
    await this._py.ex`res_KNNImputer_fit = bridgeKNNImputer[${this.id}].fit(**pms_KNNImputer_fit)`;
    return this._py`res_KNNImputer_fit.tolist() if hasattr(res_KNNImputer_fit, 'tolist') else res_KNNImputer_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("KNNImputer must call init() before fit_transform()");
    }
    await this._py.ex`pms_KNNImputer_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_KNNImputer_fit_transform = {k: v for k, v in pms_KNNImputer_fit_transform.items() if v is not None}`;
    await this._py.ex`res_KNNImputer_fit_transform = bridgeKNNImputer[${this.id}].fit_transform(**pms_KNNImputer_fit_transform)`;
    return this._py`res_KNNImputer_fit_transform.tolist() if hasattr(res_KNNImputer_fit_transform, 'tolist') else res_KNNImputer_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "KNNImputer must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_KNNImputer_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_KNNImputer_get_feature_names_out = {k: v for k, v in pms_KNNImputer_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_KNNImputer_get_feature_names_out = bridgeKNNImputer[${this.id}].get_feature_names_out(**pms_KNNImputer_get_feature_names_out)`;
    return this._py`res_KNNImputer_get_feature_names_out.tolist() if hasattr(res_KNNImputer_get_feature_names_out, 'tolist') else res_KNNImputer_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "KNNImputer must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_KNNImputer_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_KNNImputer_get_metadata_routing = {k: v for k, v in pms_KNNImputer_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_KNNImputer_get_metadata_routing = bridgeKNNImputer[${this.id}].get_metadata_routing(**pms_KNNImputer_get_metadata_routing)`;
    return this._py`res_KNNImputer_get_metadata_routing.tolist() if hasattr(res_KNNImputer_get_metadata_routing, 'tolist') else res_KNNImputer_get_metadata_routing`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("KNNImputer must call init() before set_output()");
    }
    await this._py.ex`pms_KNNImputer_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_KNNImputer_set_output = {k: v for k, v in pms_KNNImputer_set_output.items() if v is not None}`;
    await this._py.ex`res_KNNImputer_set_output = bridgeKNNImputer[${this.id}].set_output(**pms_KNNImputer_set_output)`;
    return this._py`res_KNNImputer_set_output.tolist() if hasattr(res_KNNImputer_set_output, 'tolist') else res_KNNImputer_set_output`;
  }
  /**
    Impute all missing values in X.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("KNNImputer must call init() before transform()");
    }
    await this._py.ex`pms_KNNImputer_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_KNNImputer_transform = {k: v for k, v in pms_KNNImputer_transform.items() if v is not None}`;
    await this._py.ex`res_KNNImputer_transform = bridgeKNNImputer[${this.id}].transform(**pms_KNNImputer_transform)`;
    return this._py`res_KNNImputer_transform.tolist() if hasattr(res_KNNImputer_transform, 'tolist') else res_KNNImputer_transform`;
  }
  /**
    Indicator used to add binary indicators for missing values. `undefined` if add\_indicator is `false`.
   */
  get indicator_() {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("KNNImputer must call init() before accessing indicator_");
    }
    return (async () => {
      await this._py.ex`attr_KNNImputer_indicator_ = bridgeKNNImputer[${this.id}].indicator_`;
      return this._py`attr_KNNImputer_indicator_.tolist() if hasattr(attr_KNNImputer_indicator_, 'tolist') else attr_KNNImputer_indicator_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "KNNImputer must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_KNNImputer_n_features_in_ = bridgeKNNImputer[${this.id}].n_features_in_`;
      return this._py`attr_KNNImputer_n_features_in_.tolist() if hasattr(attr_KNNImputer_n_features_in_, 'tolist') else attr_KNNImputer_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This KNNImputer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "KNNImputer must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_KNNImputer_feature_names_in_ = bridgeKNNImputer[${this.id}].feature_names_in_`;
      return this._py`attr_KNNImputer_feature_names_in_.tolist() if hasattr(attr_KNNImputer_feature_names_in_, 'tolist') else attr_KNNImputer_feature_names_in_`;
    })();
  }
};
export {
  KNNImputer
};
