// sklearn/build/generated/impute/MissingIndicator.js
// src/generated/impute/MissingIndicator.ts
import crypto from "node:crypto";
var MissingIndicator = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `MissingIndicator${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("MissingIndicator.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.impute import MissingIndicator
try: bridgeMissingIndicator
except NameError: bridgeMissingIndicator = {}
`;
    await this._py.ex`ctor_MissingIndicator = {'missing_values': ${this.opts["missing_values"] ?? void 0}, 'features': ${this.opts["features"] ?? void 0}, 'sparse': ${this.opts["sparse"] ?? void 0}, 'error_on_new': ${this.opts["error_on_new"] ?? void 0}}

ctor_MissingIndicator = {k: v for k, v in ctor_MissingIndicator.items() if v is not None}`;
    await this._py.ex`bridgeMissingIndicator[${this.id}] = MissingIndicator(**ctor_MissingIndicator)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeMissingIndicator[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the transformer on `X`.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("MissingIndicator must call init() before fit()");
    }
    await this._py.ex`pms_MissingIndicator_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_MissingIndicator_fit = {k: v for k, v in pms_MissingIndicator_fit.items() if v is not None}`;
    await this._py.ex`res_MissingIndicator_fit = bridgeMissingIndicator[${this.id}].fit(**pms_MissingIndicator_fit)`;
    return this._py`res_MissingIndicator_fit.tolist() if hasattr(res_MissingIndicator_fit, 'tolist') else res_MissingIndicator_fit`;
  }
  /**
    Generate missing values indicator for `X`.
   */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "MissingIndicator must call init() before fit_transform()"
      );
    }
    await this._py.ex`pms_MissingIndicator_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_MissingIndicator_fit_transform = {k: v for k, v in pms_MissingIndicator_fit_transform.items() if v is not None}`;
    await this._py.ex`res_MissingIndicator_fit_transform = bridgeMissingIndicator[${this.id}].fit_transform(**pms_MissingIndicator_fit_transform)`;
    return this._py`res_MissingIndicator_fit_transform.tolist() if hasattr(res_MissingIndicator_fit_transform, 'tolist') else res_MissingIndicator_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "MissingIndicator must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_MissingIndicator_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_MissingIndicator_get_feature_names_out = {k: v for k, v in pms_MissingIndicator_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_MissingIndicator_get_feature_names_out = bridgeMissingIndicator[${this.id}].get_feature_names_out(**pms_MissingIndicator_get_feature_names_out)`;
    return this._py`res_MissingIndicator_get_feature_names_out.tolist() if hasattr(res_MissingIndicator_get_feature_names_out, 'tolist') else res_MissingIndicator_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "MissingIndicator must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_MissingIndicator_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_MissingIndicator_get_metadata_routing = {k: v for k, v in pms_MissingIndicator_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_MissingIndicator_get_metadata_routing = bridgeMissingIndicator[${this.id}].get_metadata_routing(**pms_MissingIndicator_get_metadata_routing)`;
    return this._py`res_MissingIndicator_get_metadata_routing.tolist() if hasattr(res_MissingIndicator_get_metadata_routing, 'tolist') else res_MissingIndicator_get_metadata_routing`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("MissingIndicator must call init() before set_output()");
    }
    await this._py.ex`pms_MissingIndicator_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_MissingIndicator_set_output = {k: v for k, v in pms_MissingIndicator_set_output.items() if v is not None}`;
    await this._py.ex`res_MissingIndicator_set_output = bridgeMissingIndicator[${this.id}].set_output(**pms_MissingIndicator_set_output)`;
    return this._py`res_MissingIndicator_set_output.tolist() if hasattr(res_MissingIndicator_set_output, 'tolist') else res_MissingIndicator_set_output`;
  }
  /**
    Generate missing values indicator for `X`.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("MissingIndicator must call init() before transform()");
    }
    await this._py.ex`pms_MissingIndicator_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_MissingIndicator_transform = {k: v for k, v in pms_MissingIndicator_transform.items() if v is not None}`;
    await this._py.ex`res_MissingIndicator_transform = bridgeMissingIndicator[${this.id}].transform(**pms_MissingIndicator_transform)`;
    return this._py`res_MissingIndicator_transform.tolist() if hasattr(res_MissingIndicator_transform, 'tolist') else res_MissingIndicator_transform`;
  }
  /**
    The features indices which will be returned when calling [`transform`](#sklearn.impute.MissingIndicator.transform "sklearn.impute.MissingIndicator.transform"). They are computed during [`fit`](#sklearn.impute.MissingIndicator.fit "sklearn.impute.MissingIndicator.fit"). If `features='all'`, `features\_` is equal to `range(n\_features)`.
   */
  get features_() {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "MissingIndicator must call init() before accessing features_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MissingIndicator_features_ = bridgeMissingIndicator[${this.id}].features_`;
      return this._py`attr_MissingIndicator_features_.tolist() if hasattr(attr_MissingIndicator_features_, 'tolist') else attr_MissingIndicator_features_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "MissingIndicator must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MissingIndicator_n_features_in_ = bridgeMissingIndicator[${this.id}].n_features_in_`;
      return this._py`attr_MissingIndicator_n_features_in_.tolist() if hasattr(attr_MissingIndicator_n_features_in_, 'tolist') else attr_MissingIndicator_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This MissingIndicator instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "MissingIndicator must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MissingIndicator_feature_names_in_ = bridgeMissingIndicator[${this.id}].feature_names_in_`;
      return this._py`attr_MissingIndicator_feature_names_in_.tolist() if hasattr(attr_MissingIndicator_feature_names_in_, 'tolist') else attr_MissingIndicator_feature_names_in_`;
    })();
  }
};
export {
  MissingIndicator
};
