// sklearn/build/generated/linear_model/LarsCV.js
// src/generated/linear_model/LarsCV.ts
import crypto from "node:crypto";
var LarsCV = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `LarsCV${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("LarsCV.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.linear_model import LarsCV
try: bridgeLarsCV
except NameError: bridgeLarsCV = {}
`;
    await this._py.ex`ctor_LarsCV = {'fit_intercept': ${this.opts["fit_intercept"] ?? void 0}, 'verbose': ${this.opts["verbose"] ?? void 0}, 'max_iter': ${this.opts["max_iter"] ?? void 0}, 'normalize': ${this.opts["normalize"] ?? void 0}, 'precompute': ${this.opts["precompute"] ?? void 0}, 'cv': ${this.opts["cv"] ?? void 0}, 'max_n_alphas': ${this.opts["max_n_alphas"] ?? void 0}, 'n_jobs': ${this.opts["n_jobs"] ?? void 0}, 'eps': ${this.opts["eps"] ?? void 0}, 'copy_X': ${this.opts["copy_X"] ?? void 0}}

ctor_LarsCV = {k: v for k, v in ctor_LarsCV.items() if v is not None}`;
    await this._py.ex`bridgeLarsCV[${this.id}] = LarsCV(**ctor_LarsCV)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeLarsCV[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the model using X, y as training data.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before fit()");
    }
    await this._py.ex`pms_LarsCV_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None}

pms_LarsCV_fit = {k: v for k, v in pms_LarsCV_fit.items() if v is not None}`;
    await this._py.ex`res_LarsCV_fit = bridgeLarsCV[${this.id}].fit(**pms_LarsCV_fit)`;
    return this._py`res_LarsCV_fit.tolist() if hasattr(res_LarsCV_fit, 'tolist') else res_LarsCV_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before get_metadata_routing()");
    }
    await this._py.ex`pms_LarsCV_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_LarsCV_get_metadata_routing = {k: v for k, v in pms_LarsCV_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_LarsCV_get_metadata_routing = bridgeLarsCV[${this.id}].get_metadata_routing(**pms_LarsCV_get_metadata_routing)`;
    return this._py`res_LarsCV_get_metadata_routing.tolist() if hasattr(res_LarsCV_get_metadata_routing, 'tolist') else res_LarsCV_get_metadata_routing`;
  }
  /**
    Predict using the linear model.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before predict()");
    }
    await this._py.ex`pms_LarsCV_predict = {'X': ${opts["X"] ?? void 0}}

pms_LarsCV_predict = {k: v for k, v in pms_LarsCV_predict.items() if v is not None}`;
    await this._py.ex`res_LarsCV_predict = bridgeLarsCV[${this.id}].predict(**pms_LarsCV_predict)`;
    return this._py`res_LarsCV_predict.tolist() if hasattr(res_LarsCV_predict, 'tolist') else res_LarsCV_predict`;
  }
  /**
      Return the coefficient of determination of the prediction.
  
      The coefficient of determination \\(R^2\\) is defined as \\((1 - \\frac{u}{v})\\), where \\(u\\) is the residual sum of squares `((y\_true \- y\_pred)\*\* 2).sum()` and \\(v\\) is the total sum of squares `((y\_true \- y\_true.mean()) \*\* 2).sum()`. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of `y`, disregarding the input features, would get a \\(R^2\\) score of 0.0.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before score()");
    }
    await this._py.ex`pms_LarsCV_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_LarsCV_score = {k: v for k, v in pms_LarsCV_score.items() if v is not None}`;
    await this._py.ex`res_LarsCV_score = bridgeLarsCV[${this.id}].score(**pms_LarsCV_score)`;
    return this._py`res_LarsCV_score.tolist() if hasattr(res_LarsCV_score, 'tolist') else res_LarsCV_score`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before set_fit_request()");
    }
    await this._py.ex`pms_LarsCV_set_fit_request = {'Xy': ${opts["Xy"] ?? void 0}}

pms_LarsCV_set_fit_request = {k: v for k, v in pms_LarsCV_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_LarsCV_set_fit_request = bridgeLarsCV[${this.id}].set_fit_request(**pms_LarsCV_set_fit_request)`;
    return this._py`res_LarsCV_set_fit_request.tolist() if hasattr(res_LarsCV_set_fit_request, 'tolist') else res_LarsCV_set_fit_request`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before set_score_request()");
    }
    await this._py.ex`pms_LarsCV_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_LarsCV_set_score_request = {k: v for k, v in pms_LarsCV_set_score_request.items() if v is not None}`;
    await this._py.ex`res_LarsCV_set_score_request = bridgeLarsCV[${this.id}].set_score_request(**pms_LarsCV_set_score_request)`;
    return this._py`res_LarsCV_set_score_request.tolist() if hasattr(res_LarsCV_set_score_request, 'tolist') else res_LarsCV_set_score_request`;
  }
  /**
    Indices of active variables at the end of the path. If this is a list of lists, the outer list length is `n\_targets`.
   */
  get active_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing active_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_active_ = bridgeLarsCV[${this.id}].active_`;
      return this._py`attr_LarsCV_active_.tolist() if hasattr(attr_LarsCV_active_, 'tolist') else attr_LarsCV_active_`;
    })();
  }
  /**
    parameter vector (w in the formulation formula)
   */
  get coef_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing coef_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_coef_ = bridgeLarsCV[${this.id}].coef_`;
      return this._py`attr_LarsCV_coef_.tolist() if hasattr(attr_LarsCV_coef_, 'tolist') else attr_LarsCV_coef_`;
    })();
  }
  /**
    independent term in decision function
   */
  get intercept_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing intercept_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_intercept_ = bridgeLarsCV[${this.id}].intercept_`;
      return this._py`attr_LarsCV_intercept_.tolist() if hasattr(attr_LarsCV_intercept_, 'tolist') else attr_LarsCV_intercept_`;
    })();
  }
  /**
    the varying values of the coefficients along the path
   */
  get coef_path_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing coef_path_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_coef_path_ = bridgeLarsCV[${this.id}].coef_path_`;
      return this._py`attr_LarsCV_coef_path_.tolist() if hasattr(attr_LarsCV_coef_path_, 'tolist') else attr_LarsCV_coef_path_`;
    })();
  }
  /**
    the estimated regularization parameter alpha
   */
  get alpha_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing alpha_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_alpha_ = bridgeLarsCV[${this.id}].alpha_`;
      return this._py`attr_LarsCV_alpha_.tolist() if hasattr(attr_LarsCV_alpha_, 'tolist') else attr_LarsCV_alpha_`;
    })();
  }
  /**
    the different values of alpha along the path
   */
  get alphas_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing alphas_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_alphas_ = bridgeLarsCV[${this.id}].alphas_`;
      return this._py`attr_LarsCV_alphas_.tolist() if hasattr(attr_LarsCV_alphas_, 'tolist') else attr_LarsCV_alphas_`;
    })();
  }
  /**
    all the values of alpha along the path for the different folds
   */
  get cv_alphas_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing cv_alphas_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_cv_alphas_ = bridgeLarsCV[${this.id}].cv_alphas_`;
      return this._py`attr_LarsCV_cv_alphas_.tolist() if hasattr(attr_LarsCV_cv_alphas_, 'tolist') else attr_LarsCV_cv_alphas_`;
    })();
  }
  /**
    the mean square error on left-out for each fold along the path (alpha values given by `cv\_alphas`)
   */
  get mse_path_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing mse_path_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_mse_path_ = bridgeLarsCV[${this.id}].mse_path_`;
      return this._py`attr_LarsCV_mse_path_.tolist() if hasattr(attr_LarsCV_mse_path_, 'tolist') else attr_LarsCV_mse_path_`;
    })();
  }
  /**
    the number of iterations run by Lars with the optimal alpha.
   */
  get n_iter_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing n_iter_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_n_iter_ = bridgeLarsCV[${this.id}].n_iter_`;
      return this._py`attr_LarsCV_n_iter_.tolist() if hasattr(attr_LarsCV_n_iter_, 'tolist') else attr_LarsCV_n_iter_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LarsCV must call init() before accessing n_features_in_");
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_n_features_in_ = bridgeLarsCV[${this.id}].n_features_in_`;
      return this._py`attr_LarsCV_n_features_in_.tolist() if hasattr(attr_LarsCV_n_features_in_, 'tolist') else attr_LarsCV_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This LarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LarsCV must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LarsCV_feature_names_in_ = bridgeLarsCV[${this.id}].feature_names_in_`;
      return this._py`attr_LarsCV_feature_names_in_.tolist() if hasattr(attr_LarsCV_feature_names_in_, 'tolist') else attr_LarsCV_feature_names_in_`;
    })();
  }
};
export {
  LarsCV
};
