// sklearn/build/generated/linear_model/LassoLarsCV.js
// src/generated/linear_model/LassoLarsCV.ts
import crypto from "node:crypto";
var LassoLarsCV = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `LassoLarsCV${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("LassoLarsCV.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.linear_model import LassoLarsCV
try: bridgeLassoLarsCV
except NameError: bridgeLassoLarsCV = {}
`;
    await this._py.ex`ctor_LassoLarsCV = {'fit_intercept': ${this.opts["fit_intercept"] ?? void 0}, 'verbose': ${this.opts["verbose"] ?? void 0}, 'max_iter': ${this.opts["max_iter"] ?? void 0}, 'normalize': ${this.opts["normalize"] ?? void 0}, 'precompute': ${this.opts["precompute"] ?? void 0}, 'cv': ${this.opts["cv"] ?? void 0}, 'max_n_alphas': ${this.opts["max_n_alphas"] ?? void 0}, 'n_jobs': ${this.opts["n_jobs"] ?? void 0}, 'eps': ${this.opts["eps"] ?? void 0}, 'copy_X': ${this.opts["copy_X"] ?? void 0}, 'positive': ${this.opts["positive"] ?? void 0}}

ctor_LassoLarsCV = {k: v for k, v in ctor_LassoLarsCV.items() if v is not None}`;
    await this._py.ex`bridgeLassoLarsCV[${this.id}] = LassoLarsCV(**ctor_LassoLarsCV)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeLassoLarsCV[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the model using X, y as training data.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before fit()");
    }
    await this._py.ex`pms_LassoLarsCV_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None}

pms_LassoLarsCV_fit = {k: v for k, v in pms_LassoLarsCV_fit.items() if v is not None}`;
    await this._py.ex`res_LassoLarsCV_fit = bridgeLassoLarsCV[${this.id}].fit(**pms_LassoLarsCV_fit)`;
    return this._py`res_LassoLarsCV_fit.tolist() if hasattr(res_LassoLarsCV_fit, 'tolist') else res_LassoLarsCV_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LassoLarsCV must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_LassoLarsCV_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_LassoLarsCV_get_metadata_routing = {k: v for k, v in pms_LassoLarsCV_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_LassoLarsCV_get_metadata_routing = bridgeLassoLarsCV[${this.id}].get_metadata_routing(**pms_LassoLarsCV_get_metadata_routing)`;
    return this._py`res_LassoLarsCV_get_metadata_routing.tolist() if hasattr(res_LassoLarsCV_get_metadata_routing, 'tolist') else res_LassoLarsCV_get_metadata_routing`;
  }
  /**
    Predict using the linear model.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before predict()");
    }
    await this._py.ex`pms_LassoLarsCV_predict = {'X': ${opts["X"] ?? void 0}}

pms_LassoLarsCV_predict = {k: v for k, v in pms_LassoLarsCV_predict.items() if v is not None}`;
    await this._py.ex`res_LassoLarsCV_predict = bridgeLassoLarsCV[${this.id}].predict(**pms_LassoLarsCV_predict)`;
    return this._py`res_LassoLarsCV_predict.tolist() if hasattr(res_LassoLarsCV_predict, 'tolist') else res_LassoLarsCV_predict`;
  }
  /**
      Return the coefficient of determination of the prediction.
  
      The coefficient of determination \\(R^2\\) is defined as \\((1 - \\frac{u}{v})\\), where \\(u\\) is the residual sum of squares `((y\_true \- y\_pred)\*\* 2).sum()` and \\(v\\) is the total sum of squares `((y\_true \- y\_true.mean()) \*\* 2).sum()`. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of `y`, disregarding the input features, would get a \\(R^2\\) score of 0.0.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before score()");
    }
    await this._py.ex`pms_LassoLarsCV_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_LassoLarsCV_score = {k: v for k, v in pms_LassoLarsCV_score.items() if v is not None}`;
    await this._py.ex`res_LassoLarsCV_score = bridgeLassoLarsCV[${this.id}].score(**pms_LassoLarsCV_score)`;
    return this._py`res_LassoLarsCV_score.tolist() if hasattr(res_LassoLarsCV_score, 'tolist') else res_LassoLarsCV_score`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before set_fit_request()");
    }
    await this._py.ex`pms_LassoLarsCV_set_fit_request = {'Xy': ${opts["Xy"] ?? void 0}}

pms_LassoLarsCV_set_fit_request = {k: v for k, v in pms_LassoLarsCV_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_LassoLarsCV_set_fit_request = bridgeLassoLarsCV[${this.id}].set_fit_request(**pms_LassoLarsCV_set_fit_request)`;
    return this._py`res_LassoLarsCV_set_fit_request.tolist() if hasattr(res_LassoLarsCV_set_fit_request, 'tolist') else res_LassoLarsCV_set_fit_request`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before set_score_request()");
    }
    await this._py.ex`pms_LassoLarsCV_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_LassoLarsCV_set_score_request = {k: v for k, v in pms_LassoLarsCV_set_score_request.items() if v is not None}`;
    await this._py.ex`res_LassoLarsCV_set_score_request = bridgeLassoLarsCV[${this.id}].set_score_request(**pms_LassoLarsCV_set_score_request)`;
    return this._py`res_LassoLarsCV_set_score_request.tolist() if hasattr(res_LassoLarsCV_set_score_request, 'tolist') else res_LassoLarsCV_set_score_request`;
  }
  /**
    parameter vector (w in the formulation formula)
   */
  get coef_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before accessing coef_");
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_coef_ = bridgeLassoLarsCV[${this.id}].coef_`;
      return this._py`attr_LassoLarsCV_coef_.tolist() if hasattr(attr_LassoLarsCV_coef_, 'tolist') else attr_LassoLarsCV_coef_`;
    })();
  }
  /**
    independent term in decision function.
   */
  get intercept_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LassoLarsCV must call init() before accessing intercept_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_intercept_ = bridgeLassoLarsCV[${this.id}].intercept_`;
      return this._py`attr_LassoLarsCV_intercept_.tolist() if hasattr(attr_LassoLarsCV_intercept_, 'tolist') else attr_LassoLarsCV_intercept_`;
    })();
  }
  /**
    the varying values of the coefficients along the path
   */
  get coef_path_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LassoLarsCV must call init() before accessing coef_path_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_coef_path_ = bridgeLassoLarsCV[${this.id}].coef_path_`;
      return this._py`attr_LassoLarsCV_coef_path_.tolist() if hasattr(attr_LassoLarsCV_coef_path_, 'tolist') else attr_LassoLarsCV_coef_path_`;
    })();
  }
  /**
    the estimated regularization parameter alpha
   */
  get alpha_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before accessing alpha_");
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_alpha_ = bridgeLassoLarsCV[${this.id}].alpha_`;
      return this._py`attr_LassoLarsCV_alpha_.tolist() if hasattr(attr_LassoLarsCV_alpha_, 'tolist') else attr_LassoLarsCV_alpha_`;
    })();
  }
  /**
    the different values of alpha along the path
   */
  get alphas_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before accessing alphas_");
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_alphas_ = bridgeLassoLarsCV[${this.id}].alphas_`;
      return this._py`attr_LassoLarsCV_alphas_.tolist() if hasattr(attr_LassoLarsCV_alphas_, 'tolist') else attr_LassoLarsCV_alphas_`;
    })();
  }
  /**
    all the values of alpha along the path for the different folds
   */
  get cv_alphas_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LassoLarsCV must call init() before accessing cv_alphas_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_cv_alphas_ = bridgeLassoLarsCV[${this.id}].cv_alphas_`;
      return this._py`attr_LassoLarsCV_cv_alphas_.tolist() if hasattr(attr_LassoLarsCV_cv_alphas_, 'tolist') else attr_LassoLarsCV_cv_alphas_`;
    })();
  }
  /**
    the mean square error on left-out for each fold along the path (alpha values given by `cv\_alphas`)
   */
  get mse_path_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before accessing mse_path_");
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_mse_path_ = bridgeLassoLarsCV[${this.id}].mse_path_`;
      return this._py`attr_LassoLarsCV_mse_path_.tolist() if hasattr(attr_LassoLarsCV_mse_path_, 'tolist') else attr_LassoLarsCV_mse_path_`;
    })();
  }
  /**
    the number of iterations run by Lars with the optimal alpha.
   */
  get n_iter_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before accessing n_iter_");
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_n_iter_ = bridgeLassoLarsCV[${this.id}].n_iter_`;
      return this._py`attr_LassoLarsCV_n_iter_.tolist() if hasattr(attr_LassoLarsCV_n_iter_, 'tolist') else attr_LassoLarsCV_n_iter_`;
    })();
  }
  /**
    Indices of active variables at the end of the path.
   */
  get active_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LassoLarsCV must call init() before accessing active_");
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_active_ = bridgeLassoLarsCV[${this.id}].active_`;
      return this._py`attr_LassoLarsCV_active_.tolist() if hasattr(attr_LassoLarsCV_active_, 'tolist') else attr_LassoLarsCV_active_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LassoLarsCV must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_n_features_in_ = bridgeLassoLarsCV[${this.id}].n_features_in_`;
      return this._py`attr_LassoLarsCV_n_features_in_.tolist() if hasattr(attr_LassoLarsCV_n_features_in_, 'tolist') else attr_LassoLarsCV_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This LassoLarsCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LassoLarsCV must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LassoLarsCV_feature_names_in_ = bridgeLassoLarsCV[${this.id}].feature_names_in_`;
      return this._py`attr_LassoLarsCV_feature_names_in_.tolist() if hasattr(attr_LassoLarsCV_feature_names_in_, 'tolist') else attr_LassoLarsCV_feature_names_in_`;
    })();
  }
};
export {
  LassoLarsCV
};
