// sklearn/build/generated/linear_model/LinearRegression.js
// src/generated/linear_model/LinearRegression.ts
import crypto from "node:crypto";
var LinearRegression = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `LinearRegression${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("LinearRegression.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.linear_model import LinearRegression
try: bridgeLinearRegression
except NameError: bridgeLinearRegression = {}
`;
    await this._py.ex`ctor_LinearRegression = {'fit_intercept': ${this.opts["fit_intercept"] ?? void 0}, 'copy_X': ${this.opts["copy_X"] ?? void 0}, 'n_jobs': ${this.opts["n_jobs"] ?? void 0}, 'positive': ${this.opts["positive"] ?? void 0}}

ctor_LinearRegression = {k: v for k, v in ctor_LinearRegression.items() if v is not None}`;
    await this._py.ex`bridgeLinearRegression[${this.id}] = LinearRegression(**ctor_LinearRegression)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeLinearRegression[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit linear model.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("LinearRegression must call init() before fit()");
    }
    await this._py.ex`pms_LinearRegression_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_LinearRegression_fit = {k: v for k, v in pms_LinearRegression_fit.items() if v is not None}`;
    await this._py.ex`res_LinearRegression_fit = bridgeLinearRegression[${this.id}].fit(**pms_LinearRegression_fit)`;
    return this._py`res_LinearRegression_fit.tolist() if hasattr(res_LinearRegression_fit, 'tolist') else res_LinearRegression_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_LinearRegression_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_LinearRegression_get_metadata_routing = {k: v for k, v in pms_LinearRegression_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_LinearRegression_get_metadata_routing = bridgeLinearRegression[${this.id}].get_metadata_routing(**pms_LinearRegression_get_metadata_routing)`;
    return this._py`res_LinearRegression_get_metadata_routing.tolist() if hasattr(res_LinearRegression_get_metadata_routing, 'tolist') else res_LinearRegression_get_metadata_routing`;
  }
  /**
    Predict using the linear model.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("LinearRegression must call init() before predict()");
    }
    await this._py.ex`pms_LinearRegression_predict = {'X': ${opts["X"] ?? void 0}}

pms_LinearRegression_predict = {k: v for k, v in pms_LinearRegression_predict.items() if v is not None}`;
    await this._py.ex`res_LinearRegression_predict = bridgeLinearRegression[${this.id}].predict(**pms_LinearRegression_predict)`;
    return this._py`res_LinearRegression_predict.tolist() if hasattr(res_LinearRegression_predict, 'tolist') else res_LinearRegression_predict`;
  }
  /**
      Return the coefficient of determination of the prediction.
  
      The coefficient of determination \\(R^2\\) is defined as \\((1 - \\frac{u}{v})\\), where \\(u\\) is the residual sum of squares `((y\_true \- y\_pred)\*\* 2).sum()` and \\(v\\) is the total sum of squares `((y\_true \- y\_true.mean()) \*\* 2).sum()`. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of `y`, disregarding the input features, would get a \\(R^2\\) score of 0.0.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("LinearRegression must call init() before score()");
    }
    await this._py.ex`pms_LinearRegression_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_LinearRegression_score = {k: v for k, v in pms_LinearRegression_score.items() if v is not None}`;
    await this._py.ex`res_LinearRegression_score = bridgeLinearRegression[${this.id}].score(**pms_LinearRegression_score)`;
    return this._py`res_LinearRegression_score.tolist() if hasattr(res_LinearRegression_score, 'tolist') else res_LinearRegression_score`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before set_fit_request()"
      );
    }
    await this._py.ex`pms_LinearRegression_set_fit_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_LinearRegression_set_fit_request = {k: v for k, v in pms_LinearRegression_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_LinearRegression_set_fit_request = bridgeLinearRegression[${this.id}].set_fit_request(**pms_LinearRegression_set_fit_request)`;
    return this._py`res_LinearRegression_set_fit_request.tolist() if hasattr(res_LinearRegression_set_fit_request, 'tolist') else res_LinearRegression_set_fit_request`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before set_score_request()"
      );
    }
    await this._py.ex`pms_LinearRegression_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_LinearRegression_set_score_request = {k: v for k, v in pms_LinearRegression_set_score_request.items() if v is not None}`;
    await this._py.ex`res_LinearRegression_set_score_request = bridgeLinearRegression[${this.id}].set_score_request(**pms_LinearRegression_set_score_request)`;
    return this._py`res_LinearRegression_set_score_request.tolist() if hasattr(res_LinearRegression_set_score_request, 'tolist') else res_LinearRegression_set_score_request`;
  }
  /**
    Estimated coefficients for the linear regression problem. If multiple targets are passed during the fit (y 2D), this is a 2D array of shape (n\_targets, n\_features), while if only one target is passed, this is a 1D array of length n\_features.
   */
  get coef_() {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before accessing coef_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LinearRegression_coef_ = bridgeLinearRegression[${this.id}].coef_`;
      return this._py`attr_LinearRegression_coef_.tolist() if hasattr(attr_LinearRegression_coef_, 'tolist') else attr_LinearRegression_coef_`;
    })();
  }
  /**
    Rank of matrix `X`. Only available when `X` is dense.
   */
  get rank_() {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before accessing rank_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LinearRegression_rank_ = bridgeLinearRegression[${this.id}].rank_`;
      return this._py`attr_LinearRegression_rank_.tolist() if hasattr(attr_LinearRegression_rank_, 'tolist') else attr_LinearRegression_rank_`;
    })();
  }
  /**
    Singular values of `X`. Only available when `X` is dense.
   */
  get singular_() {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before accessing singular_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LinearRegression_singular_ = bridgeLinearRegression[${this.id}].singular_`;
      return this._py`attr_LinearRegression_singular_.tolist() if hasattr(attr_LinearRegression_singular_, 'tolist') else attr_LinearRegression_singular_`;
    })();
  }
  /**
    Independent term in the linear model. Set to 0.0 if `fit\_intercept \= False`.
   */
  get intercept_() {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before accessing intercept_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LinearRegression_intercept_ = bridgeLinearRegression[${this.id}].intercept_`;
      return this._py`attr_LinearRegression_intercept_.tolist() if hasattr(attr_LinearRegression_intercept_, 'tolist') else attr_LinearRegression_intercept_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LinearRegression_n_features_in_ = bridgeLinearRegression[${this.id}].n_features_in_`;
      return this._py`attr_LinearRegression_n_features_in_.tolist() if hasattr(attr_LinearRegression_n_features_in_, 'tolist') else attr_LinearRegression_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This LinearRegression instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LinearRegression must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LinearRegression_feature_names_in_ = bridgeLinearRegression[${this.id}].feature_names_in_`;
      return this._py`attr_LinearRegression_feature_names_in_.tolist() if hasattr(attr_LinearRegression_feature_names_in_, 'tolist') else attr_LinearRegression_feature_names_in_`;
    })();
  }
};
export {
  LinearRegression
};
