// sklearn/build/generated/linear_model/OrthogonalMatchingPursuit.js
// src/generated/linear_model/OrthogonalMatchingPursuit.ts
import crypto from "node:crypto";
var OrthogonalMatchingPursuit = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `OrthogonalMatchingPursuit${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error(
        "OrthogonalMatchingPursuit.init requires a PythonBridge instance"
      );
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.linear_model import OrthogonalMatchingPursuit
try: bridgeOrthogonalMatchingPursuit
except NameError: bridgeOrthogonalMatchingPursuit = {}
`;
    await this._py.ex`ctor_OrthogonalMatchingPursuit = {'n_nonzero_coefs': ${this.opts["n_nonzero_coefs"] ?? void 0}, 'tol': ${this.opts["tol"] ?? void 0}, 'fit_intercept': ${this.opts["fit_intercept"] ?? void 0}, 'normalize': ${this.opts["normalize"] ?? void 0}, 'precompute': ${this.opts["precompute"] ?? void 0}}

ctor_OrthogonalMatchingPursuit = {k: v for k, v in ctor_OrthogonalMatchingPursuit.items() if v is not None}`;
    await this._py.ex`bridgeOrthogonalMatchingPursuit[${this.id}] = OrthogonalMatchingPursuit(**ctor_OrthogonalMatchingPursuit)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeOrthogonalMatchingPursuit[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the model using X, y as training data.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("OrthogonalMatchingPursuit must call init() before fit()");
    }
    await this._py.ex`pms_OrthogonalMatchingPursuit_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None}

pms_OrthogonalMatchingPursuit_fit = {k: v for k, v in pms_OrthogonalMatchingPursuit_fit.items() if v is not None}`;
    await this._py.ex`res_OrthogonalMatchingPursuit_fit = bridgeOrthogonalMatchingPursuit[${this.id}].fit(**pms_OrthogonalMatchingPursuit_fit)`;
    return this._py`res_OrthogonalMatchingPursuit_fit.tolist() if hasattr(res_OrthogonalMatchingPursuit_fit, 'tolist') else res_OrthogonalMatchingPursuit_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_OrthogonalMatchingPursuit_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_OrthogonalMatchingPursuit_get_metadata_routing = {k: v for k, v in pms_OrthogonalMatchingPursuit_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_OrthogonalMatchingPursuit_get_metadata_routing = bridgeOrthogonalMatchingPursuit[${this.id}].get_metadata_routing(**pms_OrthogonalMatchingPursuit_get_metadata_routing)`;
    return this._py`res_OrthogonalMatchingPursuit_get_metadata_routing.tolist() if hasattr(res_OrthogonalMatchingPursuit_get_metadata_routing, 'tolist') else res_OrthogonalMatchingPursuit_get_metadata_routing`;
  }
  /**
    Predict using the linear model.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before predict()"
      );
    }
    await this._py.ex`pms_OrthogonalMatchingPursuit_predict = {'X': ${opts["X"] ?? void 0}}

pms_OrthogonalMatchingPursuit_predict = {k: v for k, v in pms_OrthogonalMatchingPursuit_predict.items() if v is not None}`;
    await this._py.ex`res_OrthogonalMatchingPursuit_predict = bridgeOrthogonalMatchingPursuit[${this.id}].predict(**pms_OrthogonalMatchingPursuit_predict)`;
    return this._py`res_OrthogonalMatchingPursuit_predict.tolist() if hasattr(res_OrthogonalMatchingPursuit_predict, 'tolist') else res_OrthogonalMatchingPursuit_predict`;
  }
  /**
      Return the coefficient of determination of the prediction.
  
      The coefficient of determination \\(R^2\\) is defined as \\((1 - \\frac{u}{v})\\), where \\(u\\) is the residual sum of squares `((y\_true \- y\_pred)\*\* 2).sum()` and \\(v\\) is the total sum of squares `((y\_true \- y\_true.mean()) \*\* 2).sum()`. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of `y`, disregarding the input features, would get a \\(R^2\\) score of 0.0.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before score()"
      );
    }
    await this._py.ex`pms_OrthogonalMatchingPursuit_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_OrthogonalMatchingPursuit_score = {k: v for k, v in pms_OrthogonalMatchingPursuit_score.items() if v is not None}`;
    await this._py.ex`res_OrthogonalMatchingPursuit_score = bridgeOrthogonalMatchingPursuit[${this.id}].score(**pms_OrthogonalMatchingPursuit_score)`;
    return this._py`res_OrthogonalMatchingPursuit_score.tolist() if hasattr(res_OrthogonalMatchingPursuit_score, 'tolist') else res_OrthogonalMatchingPursuit_score`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before set_score_request()"
      );
    }
    await this._py.ex`pms_OrthogonalMatchingPursuit_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_OrthogonalMatchingPursuit_set_score_request = {k: v for k, v in pms_OrthogonalMatchingPursuit_set_score_request.items() if v is not None}`;
    await this._py.ex`res_OrthogonalMatchingPursuit_set_score_request = bridgeOrthogonalMatchingPursuit[${this.id}].set_score_request(**pms_OrthogonalMatchingPursuit_set_score_request)`;
    return this._py`res_OrthogonalMatchingPursuit_set_score_request.tolist() if hasattr(res_OrthogonalMatchingPursuit_set_score_request, 'tolist') else res_OrthogonalMatchingPursuit_set_score_request`;
  }
  /**
    Parameter vector (w in the formula).
   */
  get coef_() {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before accessing coef_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrthogonalMatchingPursuit_coef_ = bridgeOrthogonalMatchingPursuit[${this.id}].coef_`;
      return this._py`attr_OrthogonalMatchingPursuit_coef_.tolist() if hasattr(attr_OrthogonalMatchingPursuit_coef_, 'tolist') else attr_OrthogonalMatchingPursuit_coef_`;
    })();
  }
  /**
    Independent term in decision function.
   */
  get intercept_() {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before accessing intercept_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrthogonalMatchingPursuit_intercept_ = bridgeOrthogonalMatchingPursuit[${this.id}].intercept_`;
      return this._py`attr_OrthogonalMatchingPursuit_intercept_.tolist() if hasattr(attr_OrthogonalMatchingPursuit_intercept_, 'tolist') else attr_OrthogonalMatchingPursuit_intercept_`;
    })();
  }
  /**
    Number of active features across every target.
   */
  get n_iter_() {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before accessing n_iter_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrthogonalMatchingPursuit_n_iter_ = bridgeOrthogonalMatchingPursuit[${this.id}].n_iter_`;
      return this._py`attr_OrthogonalMatchingPursuit_n_iter_.tolist() if hasattr(attr_OrthogonalMatchingPursuit_n_iter_, 'tolist') else attr_OrthogonalMatchingPursuit_n_iter_`;
    })();
  }
  /**
    The number of non-zero coefficients in the solution. If `n\_nonzero\_coefs` is `undefined` and `tol` is `undefined` this value is either set to 10% of `n\_features` or 1, whichever is greater.
   */
  get n_nonzero_coefs_() {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before accessing n_nonzero_coefs_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrthogonalMatchingPursuit_n_nonzero_coefs_ = bridgeOrthogonalMatchingPursuit[${this.id}].n_nonzero_coefs_`;
      return this._py`attr_OrthogonalMatchingPursuit_n_nonzero_coefs_.tolist() if hasattr(attr_OrthogonalMatchingPursuit_n_nonzero_coefs_, 'tolist') else attr_OrthogonalMatchingPursuit_n_nonzero_coefs_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrthogonalMatchingPursuit_n_features_in_ = bridgeOrthogonalMatchingPursuit[${this.id}].n_features_in_`;
      return this._py`attr_OrthogonalMatchingPursuit_n_features_in_.tolist() if hasattr(attr_OrthogonalMatchingPursuit_n_features_in_, 'tolist') else attr_OrthogonalMatchingPursuit_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This OrthogonalMatchingPursuit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrthogonalMatchingPursuit must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrthogonalMatchingPursuit_feature_names_in_ = bridgeOrthogonalMatchingPursuit[${this.id}].feature_names_in_`;
      return this._py`attr_OrthogonalMatchingPursuit_feature_names_in_.tolist() if hasattr(attr_OrthogonalMatchingPursuit_feature_names_in_, 'tolist') else attr_OrthogonalMatchingPursuit_feature_names_in_`;
    })();
  }
};
export {
  OrthogonalMatchingPursuit
};
