// sklearn/build/generated/linear_model/Ridge.js
// src/generated/linear_model/Ridge.ts
import crypto from "node:crypto";
var Ridge = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `Ridge${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("Ridge.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.linear_model import Ridge
try: bridgeRidge
except NameError: bridgeRidge = {}
`;
    await this._py.ex`ctor_Ridge = {'alpha': np.array(${this.opts["alpha"] ?? void 0}) if ${this.opts["alpha"] !== void 0} else None, 'fit_intercept': ${this.opts["fit_intercept"] ?? void 0}, 'copy_X': ${this.opts["copy_X"] ?? void 0}, 'max_iter': ${this.opts["max_iter"] ?? void 0}, 'tol': ${this.opts["tol"] ?? void 0}, 'solver': ${this.opts["solver"] ?? void 0}, 'positive': ${this.opts["positive"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}}

ctor_Ridge = {k: v for k, v in ctor_Ridge.items() if v is not None}`;
    await this._py.ex`bridgeRidge[${this.id}] = Ridge(**ctor_Ridge)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeRidge[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit Ridge regression model.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before fit()");
    }
    await this._py.ex`pms_Ridge_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_Ridge_fit = {k: v for k, v in pms_Ridge_fit.items() if v is not None}`;
    await this._py.ex`res_Ridge_fit = bridgeRidge[${this.id}].fit(**pms_Ridge_fit)`;
    return this._py`res_Ridge_fit.tolist() if hasattr(res_Ridge_fit, 'tolist') else res_Ridge_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before get_metadata_routing()");
    }
    await this._py.ex`pms_Ridge_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_Ridge_get_metadata_routing = {k: v for k, v in pms_Ridge_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_Ridge_get_metadata_routing = bridgeRidge[${this.id}].get_metadata_routing(**pms_Ridge_get_metadata_routing)`;
    return this._py`res_Ridge_get_metadata_routing.tolist() if hasattr(res_Ridge_get_metadata_routing, 'tolist') else res_Ridge_get_metadata_routing`;
  }
  /**
    Predict using the linear model.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before predict()");
    }
    await this._py.ex`pms_Ridge_predict = {'X': ${opts["X"] ?? void 0}}

pms_Ridge_predict = {k: v for k, v in pms_Ridge_predict.items() if v is not None}`;
    await this._py.ex`res_Ridge_predict = bridgeRidge[${this.id}].predict(**pms_Ridge_predict)`;
    return this._py`res_Ridge_predict.tolist() if hasattr(res_Ridge_predict, 'tolist') else res_Ridge_predict`;
  }
  /**
      Return the coefficient of determination of the prediction.
  
      The coefficient of determination \\(R^2\\) is defined as \\((1 - \\frac{u}{v})\\), where \\(u\\) is the residual sum of squares `((y\_true \- y\_pred)\*\* 2).sum()` and \\(v\\) is the total sum of squares `((y\_true \- y\_true.mean()) \*\* 2).sum()`. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of `y`, disregarding the input features, would get a \\(R^2\\) score of 0.0.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before score()");
    }
    await this._py.ex`pms_Ridge_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_Ridge_score = {k: v for k, v in pms_Ridge_score.items() if v is not None}`;
    await this._py.ex`res_Ridge_score = bridgeRidge[${this.id}].score(**pms_Ridge_score)`;
    return this._py`res_Ridge_score.tolist() if hasattr(res_Ridge_score, 'tolist') else res_Ridge_score`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before set_fit_request()");
    }
    await this._py.ex`pms_Ridge_set_fit_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_Ridge_set_fit_request = {k: v for k, v in pms_Ridge_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_Ridge_set_fit_request = bridgeRidge[${this.id}].set_fit_request(**pms_Ridge_set_fit_request)`;
    return this._py`res_Ridge_set_fit_request.tolist() if hasattr(res_Ridge_set_fit_request, 'tolist') else res_Ridge_set_fit_request`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before set_score_request()");
    }
    await this._py.ex`pms_Ridge_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_Ridge_set_score_request = {k: v for k, v in pms_Ridge_set_score_request.items() if v is not None}`;
    await this._py.ex`res_Ridge_set_score_request = bridgeRidge[${this.id}].set_score_request(**pms_Ridge_set_score_request)`;
    return this._py`res_Ridge_set_score_request.tolist() if hasattr(res_Ridge_set_score_request, 'tolist') else res_Ridge_set_score_request`;
  }
  /**
    Weight vector(s).
   */
  get coef_() {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before accessing coef_");
    }
    return (async () => {
      await this._py.ex`attr_Ridge_coef_ = bridgeRidge[${this.id}].coef_`;
      return this._py`attr_Ridge_coef_.tolist() if hasattr(attr_Ridge_coef_, 'tolist') else attr_Ridge_coef_`;
    })();
  }
  /**
    Independent term in decision function. Set to 0.0 if `fit\_intercept \= False`.
   */
  get intercept_() {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before accessing intercept_");
    }
    return (async () => {
      await this._py.ex`attr_Ridge_intercept_ = bridgeRidge[${this.id}].intercept_`;
      return this._py`attr_Ridge_intercept_.tolist() if hasattr(attr_Ridge_intercept_, 'tolist') else attr_Ridge_intercept_`;
    })();
  }
  /**
    Actual number of iterations for each target. Available only for sag and lsqr solvers. Other solvers will return `undefined`.
   */
  get n_iter_() {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before accessing n_iter_");
    }
    return (async () => {
      await this._py.ex`attr_Ridge_n_iter_ = bridgeRidge[${this.id}].n_iter_`;
      return this._py`attr_Ridge_n_iter_.tolist() if hasattr(attr_Ridge_n_iter_, 'tolist') else attr_Ridge_n_iter_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Ridge must call init() before accessing n_features_in_");
    }
    return (async () => {
      await this._py.ex`attr_Ridge_n_features_in_ = bridgeRidge[${this.id}].n_features_in_`;
      return this._py`attr_Ridge_n_features_in_.tolist() if hasattr(attr_Ridge_n_features_in_, 'tolist') else attr_Ridge_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This Ridge instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "Ridge must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_Ridge_feature_names_in_ = bridgeRidge[${this.id}].feature_names_in_`;
      return this._py`attr_Ridge_feature_names_in_.tolist() if hasattr(attr_Ridge_feature_names_in_, 'tolist') else attr_Ridge_feature_names_in_`;
    })();
  }
};
export {
  Ridge
};
