// sklearn/build/generated/multioutput/ClassifierChain.js
// src/generated/multioutput/ClassifierChain.ts
import crypto from "node:crypto";
var ClassifierChain = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `ClassifierChain${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("ClassifierChain.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.multioutput import ClassifierChain
try: bridgeClassifierChain
except NameError: bridgeClassifierChain = {}
`;
    await this._py.ex`ctor_ClassifierChain = {'base_estimator': ${this.opts["base_estimator"] ?? void 0}, 'order': np.array(${this.opts["order"] ?? void 0}) if ${this.opts["order"] !== void 0} else None, 'cv': ${this.opts["cv"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}, 'verbose': ${this.opts["verbose"] ?? void 0}}

ctor_ClassifierChain = {k: v for k, v in ctor_ClassifierChain.items() if v is not None}`;
    await this._py.ex`bridgeClassifierChain[${this.id}] = ClassifierChain(**ctor_ClassifierChain)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeClassifierChain[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Evaluate the decision\_function of the models in the chain.
   */
  async decision_function(opts) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before decision_function()"
      );
    }
    await this._py.ex`pms_ClassifierChain_decision_function = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_ClassifierChain_decision_function = {k: v for k, v in pms_ClassifierChain_decision_function.items() if v is not None}`;
    await this._py.ex`res_ClassifierChain_decision_function = bridgeClassifierChain[${this.id}].decision_function(**pms_ClassifierChain_decision_function)`;
    return this._py`res_ClassifierChain_decision_function.tolist() if hasattr(res_ClassifierChain_decision_function, 'tolist') else res_ClassifierChain_decision_function`;
  }
  /**
    Fit the model to data matrix X and targets Y.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("ClassifierChain must call init() before fit()");
    }
    await this._py.ex`pms_ClassifierChain_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'Y': np.array(${opts["Y"] ?? void 0}) if ${opts["Y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_ClassifierChain_fit = {k: v for k, v in pms_ClassifierChain_fit.items() if v is not None}`;
    await this._py.ex`res_ClassifierChain_fit = bridgeClassifierChain[${this.id}].fit(**pms_ClassifierChain_fit)`;
    return this._py`res_ClassifierChain_fit.tolist() if hasattr(res_ClassifierChain_fit, 'tolist') else res_ClassifierChain_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_ClassifierChain_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_ClassifierChain_get_metadata_routing = {k: v for k, v in pms_ClassifierChain_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_ClassifierChain_get_metadata_routing = bridgeClassifierChain[${this.id}].get_metadata_routing(**pms_ClassifierChain_get_metadata_routing)`;
    return this._py`res_ClassifierChain_get_metadata_routing.tolist() if hasattr(res_ClassifierChain_get_metadata_routing, 'tolist') else res_ClassifierChain_get_metadata_routing`;
  }
  /**
    Predict on the data matrix X using the ClassifierChain model.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("ClassifierChain must call init() before predict()");
    }
    await this._py.ex`pms_ClassifierChain_predict = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_ClassifierChain_predict = {k: v for k, v in pms_ClassifierChain_predict.items() if v is not None}`;
    await this._py.ex`res_ClassifierChain_predict = bridgeClassifierChain[${this.id}].predict(**pms_ClassifierChain_predict)`;
    return this._py`res_ClassifierChain_predict.tolist() if hasattr(res_ClassifierChain_predict, 'tolist') else res_ClassifierChain_predict`;
  }
  /**
    Predict probability estimates.
   */
  async predict_proba(opts) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("ClassifierChain must call init() before predict_proba()");
    }
    await this._py.ex`pms_ClassifierChain_predict_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_ClassifierChain_predict_proba = {k: v for k, v in pms_ClassifierChain_predict_proba.items() if v is not None}`;
    await this._py.ex`res_ClassifierChain_predict_proba = bridgeClassifierChain[${this.id}].predict_proba(**pms_ClassifierChain_predict_proba)`;
    return this._py`res_ClassifierChain_predict_proba.tolist() if hasattr(res_ClassifierChain_predict_proba, 'tolist') else res_ClassifierChain_predict_proba`;
  }
  /**
      Return the mean accuracy on the given test data and labels.
  
      In multi-label classification, this is the subset accuracy which is a harsh metric since you require for each sample that each label set be correctly predicted.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("ClassifierChain must call init() before score()");
    }
    await this._py.ex`pms_ClassifierChain_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_ClassifierChain_score = {k: v for k, v in pms_ClassifierChain_score.items() if v is not None}`;
    await this._py.ex`res_ClassifierChain_score = bridgeClassifierChain[${this.id}].score(**pms_ClassifierChain_score)`;
    return this._py`res_ClassifierChain_score.tolist() if hasattr(res_ClassifierChain_score, 'tolist') else res_ClassifierChain_score`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before set_score_request()"
      );
    }
    await this._py.ex`pms_ClassifierChain_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_ClassifierChain_set_score_request = {k: v for k, v in pms_ClassifierChain_set_score_request.items() if v is not None}`;
    await this._py.ex`res_ClassifierChain_set_score_request = bridgeClassifierChain[${this.id}].set_score_request(**pms_ClassifierChain_set_score_request)`;
    return this._py`res_ClassifierChain_set_score_request.tolist() if hasattr(res_ClassifierChain_set_score_request, 'tolist') else res_ClassifierChain_set_score_request`;
  }
  /**
    A list of arrays of length `len(estimators\_)` containing the class labels for each estimator in the chain.
   */
  get classes_() {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before accessing classes_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ClassifierChain_classes_ = bridgeClassifierChain[${this.id}].classes_`;
      return this._py`attr_ClassifierChain_classes_.tolist() if hasattr(attr_ClassifierChain_classes_, 'tolist') else attr_ClassifierChain_classes_`;
    })();
  }
  /**
    A list of clones of base\_estimator.
   */
  get estimators_() {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before accessing estimators_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ClassifierChain_estimators_ = bridgeClassifierChain[${this.id}].estimators_`;
      return this._py`attr_ClassifierChain_estimators_.tolist() if hasattr(attr_ClassifierChain_estimators_, 'tolist') else attr_ClassifierChain_estimators_`;
    })();
  }
  /**
    The order of labels in the classifier chain.
   */
  get order_() {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before accessing order_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ClassifierChain_order_ = bridgeClassifierChain[${this.id}].order_`;
      return this._py`attr_ClassifierChain_order_.tolist() if hasattr(attr_ClassifierChain_order_, 'tolist') else attr_ClassifierChain_order_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit). Only defined if the underlying `base\_estimator` exposes such an attribute when fit.
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ClassifierChain_n_features_in_ = bridgeClassifierChain[${this.id}].n_features_in_`;
      return this._py`attr_ClassifierChain_n_features_in_.tolist() if hasattr(attr_ClassifierChain_n_features_in_, 'tolist') else attr_ClassifierChain_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This ClassifierChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "ClassifierChain must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ClassifierChain_feature_names_in_ = bridgeClassifierChain[${this.id}].feature_names_in_`;
      return this._py`attr_ClassifierChain_feature_names_in_.tolist() if hasattr(attr_ClassifierChain_feature_names_in_, 'tolist') else attr_ClassifierChain_feature_names_in_`;
    })();
  }
};
export {
  ClassifierChain
};
