// sklearn/build/generated/multioutput/RegressorChain.js
// src/generated/multioutput/RegressorChain.ts
import crypto from "node:crypto";
var RegressorChain = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `RegressorChain${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("RegressorChain.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.multioutput import RegressorChain
try: bridgeRegressorChain
except NameError: bridgeRegressorChain = {}
`;
    await this._py.ex`ctor_RegressorChain = {'base_estimator': ${this.opts["base_estimator"] ?? void 0}, 'order': np.array(${this.opts["order"] ?? void 0}) if ${this.opts["order"] !== void 0} else None, 'cv': ${this.opts["cv"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}, 'verbose': ${this.opts["verbose"] ?? void 0}}

ctor_RegressorChain = {k: v for k, v in ctor_RegressorChain.items() if v is not None}`;
    await this._py.ex`bridgeRegressorChain[${this.id}] = RegressorChain(**ctor_RegressorChain)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeRegressorChain[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the model to data matrix X and targets Y.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RegressorChain must call init() before fit()");
    }
    await this._py.ex`pms_RegressorChain_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'Y': np.array(${opts["Y"] ?? void 0}) if ${opts["Y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_RegressorChain_fit = {k: v for k, v in pms_RegressorChain_fit.items() if v is not None}`;
    await this._py.ex`res_RegressorChain_fit = bridgeRegressorChain[${this.id}].fit(**pms_RegressorChain_fit)`;
    return this._py`res_RegressorChain_fit.tolist() if hasattr(res_RegressorChain_fit, 'tolist') else res_RegressorChain_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RegressorChain must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_RegressorChain_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_RegressorChain_get_metadata_routing = {k: v for k, v in pms_RegressorChain_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_RegressorChain_get_metadata_routing = bridgeRegressorChain[${this.id}].get_metadata_routing(**pms_RegressorChain_get_metadata_routing)`;
    return this._py`res_RegressorChain_get_metadata_routing.tolist() if hasattr(res_RegressorChain_get_metadata_routing, 'tolist') else res_RegressorChain_get_metadata_routing`;
  }
  /**
    Predict on the data matrix X using the ClassifierChain model.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RegressorChain must call init() before predict()");
    }
    await this._py.ex`pms_RegressorChain_predict = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_RegressorChain_predict = {k: v for k, v in pms_RegressorChain_predict.items() if v is not None}`;
    await this._py.ex`res_RegressorChain_predict = bridgeRegressorChain[${this.id}].predict(**pms_RegressorChain_predict)`;
    return this._py`res_RegressorChain_predict.tolist() if hasattr(res_RegressorChain_predict, 'tolist') else res_RegressorChain_predict`;
  }
  /**
      Return the coefficient of determination of the prediction.
  
      The coefficient of determination \\(R^2\\) is defined as \\((1 - \\frac{u}{v})\\), where \\(u\\) is the residual sum of squares `((y\_true \- y\_pred)\*\* 2).sum()` and \\(v\\) is the total sum of squares `((y\_true \- y\_true.mean()) \*\* 2).sum()`. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of `y`, disregarding the input features, would get a \\(R^2\\) score of 0.0.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RegressorChain must call init() before score()");
    }
    await this._py.ex`pms_RegressorChain_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_RegressorChain_score = {k: v for k, v in pms_RegressorChain_score.items() if v is not None}`;
    await this._py.ex`res_RegressorChain_score = bridgeRegressorChain[${this.id}].score(**pms_RegressorChain_score)`;
    return this._py`res_RegressorChain_score.tolist() if hasattr(res_RegressorChain_score, 'tolist') else res_RegressorChain_score`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RegressorChain must call init() before set_score_request()"
      );
    }
    await this._py.ex`pms_RegressorChain_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_RegressorChain_set_score_request = {k: v for k, v in pms_RegressorChain_set_score_request.items() if v is not None}`;
    await this._py.ex`res_RegressorChain_set_score_request = bridgeRegressorChain[${this.id}].set_score_request(**pms_RegressorChain_set_score_request)`;
    return this._py`res_RegressorChain_set_score_request.tolist() if hasattr(res_RegressorChain_set_score_request, 'tolist') else res_RegressorChain_set_score_request`;
  }
  /**
    A list of clones of base\_estimator.
   */
  get estimators_() {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RegressorChain must call init() before accessing estimators_"
      );
    }
    return (async () => {
      await this._py.ex`attr_RegressorChain_estimators_ = bridgeRegressorChain[${this.id}].estimators_`;
      return this._py`attr_RegressorChain_estimators_.tolist() if hasattr(attr_RegressorChain_estimators_, 'tolist') else attr_RegressorChain_estimators_`;
    })();
  }
  /**
    The order of labels in the classifier chain.
   */
  get order_() {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RegressorChain must call init() before accessing order_");
    }
    return (async () => {
      await this._py.ex`attr_RegressorChain_order_ = bridgeRegressorChain[${this.id}].order_`;
      return this._py`attr_RegressorChain_order_.tolist() if hasattr(attr_RegressorChain_order_, 'tolist') else attr_RegressorChain_order_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit). Only defined if the underlying `base\_estimator` exposes such an attribute when fit.
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RegressorChain must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_RegressorChain_n_features_in_ = bridgeRegressorChain[${this.id}].n_features_in_`;
      return this._py`attr_RegressorChain_n_features_in_.tolist() if hasattr(attr_RegressorChain_n_features_in_, 'tolist') else attr_RegressorChain_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This RegressorChain instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RegressorChain must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_RegressorChain_feature_names_in_ = bridgeRegressorChain[${this.id}].feature_names_in_`;
      return this._py`attr_RegressorChain_feature_names_in_.tolist() if hasattr(attr_RegressorChain_feature_names_in_, 'tolist') else attr_RegressorChain_feature_names_in_`;
    })();
  }
};
export {
  RegressorChain
};
