// sklearn/build/generated/feature_selection/SelectFromModel.js
// src/generated/feature_selection/SelectFromModel.ts
import crypto from "node:crypto";
var SelectFromModel = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `SelectFromModel${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("SelectFromModel.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.feature_selection import SelectFromModel
try: bridgeSelectFromModel
except NameError: bridgeSelectFromModel = {}
`;
    await this._py.ex`ctor_SelectFromModel = {'estimator': ${this.opts["estimator"] ?? void 0}, 'threshold': ${this.opts["threshold"] ?? void 0}, 'prefit': ${this.opts["prefit"] ?? void 0}, 'norm_order': ${this.opts["norm_order"] ?? void 0}, 'max_features': ${this.opts["max_features"] ?? void 0}, 'importance_getter': ${this.opts["importance_getter"] ?? void 0}}

ctor_SelectFromModel = {k: v for k, v in ctor_SelectFromModel.items() if v is not None}`;
    await this._py.ex`bridgeSelectFromModel[${this.id}] = SelectFromModel(**ctor_SelectFromModel)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeSelectFromModel[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the SelectFromModel meta-transformer.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("SelectFromModel must call init() before fit()");
    }
    await this._py.ex`pms_SelectFromModel_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_SelectFromModel_fit = {k: v for k, v in pms_SelectFromModel_fit.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_fit = bridgeSelectFromModel[${this.id}].fit(**pms_SelectFromModel_fit)`;
    return this._py`res_SelectFromModel_fit.tolist() if hasattr(res_SelectFromModel_fit, 'tolist') else res_SelectFromModel_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("SelectFromModel must call init() before fit_transform()");
    }
    await this._py.ex`pms_SelectFromModel_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_SelectFromModel_fit_transform = {k: v for k, v in pms_SelectFromModel_fit_transform.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_fit_transform = bridgeSelectFromModel[${this.id}].fit_transform(**pms_SelectFromModel_fit_transform)`;
    return this._py`res_SelectFromModel_fit_transform.tolist() if hasattr(res_SelectFromModel_fit_transform, 'tolist') else res_SelectFromModel_fit_transform`;
  }
  /**
    Mask feature names according to selected features.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "SelectFromModel must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_SelectFromModel_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_SelectFromModel_get_feature_names_out = {k: v for k, v in pms_SelectFromModel_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_get_feature_names_out = bridgeSelectFromModel[${this.id}].get_feature_names_out(**pms_SelectFromModel_get_feature_names_out)`;
    return this._py`res_SelectFromModel_get_feature_names_out.tolist() if hasattr(res_SelectFromModel_get_feature_names_out, 'tolist') else res_SelectFromModel_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "SelectFromModel must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_SelectFromModel_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_SelectFromModel_get_metadata_routing = {k: v for k, v in pms_SelectFromModel_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_get_metadata_routing = bridgeSelectFromModel[${this.id}].get_metadata_routing(**pms_SelectFromModel_get_metadata_routing)`;
    return this._py`res_SelectFromModel_get_metadata_routing.tolist() if hasattr(res_SelectFromModel_get_metadata_routing, 'tolist') else res_SelectFromModel_get_metadata_routing`;
  }
  /**
    Get a mask, or integer index, of the features selected.
   */
  async get_support(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("SelectFromModel must call init() before get_support()");
    }
    await this._py.ex`pms_SelectFromModel_get_support = {'indices': ${opts["indices"] ?? void 0}}

pms_SelectFromModel_get_support = {k: v for k, v in pms_SelectFromModel_get_support.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_get_support = bridgeSelectFromModel[${this.id}].get_support(**pms_SelectFromModel_get_support)`;
    return this._py`res_SelectFromModel_get_support.tolist() if hasattr(res_SelectFromModel_get_support, 'tolist') else res_SelectFromModel_get_support`;
  }
  /**
    Reverse the transformation operation.
   */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "SelectFromModel must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_SelectFromModel_inverse_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_SelectFromModel_inverse_transform = {k: v for k, v in pms_SelectFromModel_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_inverse_transform = bridgeSelectFromModel[${this.id}].inverse_transform(**pms_SelectFromModel_inverse_transform)`;
    return this._py`res_SelectFromModel_inverse_transform.tolist() if hasattr(res_SelectFromModel_inverse_transform, 'tolist') else res_SelectFromModel_inverse_transform`;
  }
  /**
    Fit the SelectFromModel meta-transformer only once.
   */
  async partial_fit(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("SelectFromModel must call init() before partial_fit()");
    }
    await this._py.ex`pms_SelectFromModel_partial_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_SelectFromModel_partial_fit = {k: v for k, v in pms_SelectFromModel_partial_fit.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_partial_fit = bridgeSelectFromModel[${this.id}].partial_fit(**pms_SelectFromModel_partial_fit)`;
    return this._py`res_SelectFromModel_partial_fit.tolist() if hasattr(res_SelectFromModel_partial_fit, 'tolist') else res_SelectFromModel_partial_fit`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("SelectFromModel must call init() before set_output()");
    }
    await this._py.ex`pms_SelectFromModel_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_SelectFromModel_set_output = {k: v for k, v in pms_SelectFromModel_set_output.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_set_output = bridgeSelectFromModel[${this.id}].set_output(**pms_SelectFromModel_set_output)`;
    return this._py`res_SelectFromModel_set_output.tolist() if hasattr(res_SelectFromModel_set_output, 'tolist') else res_SelectFromModel_set_output`;
  }
  /**
    Reduce X to the selected features.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("SelectFromModel must call init() before transform()");
    }
    await this._py.ex`pms_SelectFromModel_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_SelectFromModel_transform = {k: v for k, v in pms_SelectFromModel_transform.items() if v is not None}`;
    await this._py.ex`res_SelectFromModel_transform = bridgeSelectFromModel[${this.id}].transform(**pms_SelectFromModel_transform)`;
    return this._py`res_SelectFromModel_transform.tolist() if hasattr(res_SelectFromModel_transform, 'tolist') else res_SelectFromModel_transform`;
  }
  /**
    The base estimator from which the transformer is built. This attribute exist only when `fit` has been called.
   */
  get estimator_() {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "SelectFromModel must call init() before accessing estimator_"
      );
    }
    return (async () => {
      await this._py.ex`attr_SelectFromModel_estimator_ = bridgeSelectFromModel[${this.id}].estimator_`;
      return this._py`attr_SelectFromModel_estimator_.tolist() if hasattr(attr_SelectFromModel_estimator_, 'tolist') else attr_SelectFromModel_estimator_`;
    })();
  }
  /**
    Maximum number of features calculated during [fit](../../glossary.html#term-fit). Only defined if the `max\_features` is not `undefined`.
   */
  get max_features_() {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "SelectFromModel must call init() before accessing max_features_"
      );
    }
    return (async () => {
      await this._py.ex`attr_SelectFromModel_max_features_ = bridgeSelectFromModel[${this.id}].max_features_`;
      return this._py`attr_SelectFromModel_max_features_.tolist() if hasattr(attr_SelectFromModel_max_features_, 'tolist') else attr_SelectFromModel_max_features_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This SelectFromModel instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "SelectFromModel must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_SelectFromModel_feature_names_in_ = bridgeSelectFromModel[${this.id}].feature_names_in_`;
      return this._py`attr_SelectFromModel_feature_names_in_.tolist() if hasattr(attr_SelectFromModel_feature_names_in_, 'tolist') else attr_SelectFromModel_feature_names_in_`;
    })();
  }
};
export {
  SelectFromModel
};
