// sklearn/build/generated/covariance/LedoitWolf.js
// src/generated/covariance/LedoitWolf.ts
import crypto from "node:crypto";
var LedoitWolf = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `LedoitWolf${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("LedoitWolf.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.covariance import LedoitWolf
try: bridgeLedoitWolf
except NameError: bridgeLedoitWolf = {}
`;
    await this._py.ex`ctor_LedoitWolf = {'store_precision': ${this.opts["store_precision"] ?? void 0}, 'assume_centered': ${this.opts["assume_centered"] ?? void 0}, 'block_size': ${this.opts["block_size"] ?? void 0}}

ctor_LedoitWolf = {k: v for k, v in ctor_LedoitWolf.items() if v is not None}`;
    await this._py.ex`bridgeLedoitWolf[${this.id}] = LedoitWolf(**ctor_LedoitWolf)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeLedoitWolf[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Compute the Mean Squared Error between two covariance estimators.
   */
  async error_norm(opts) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before error_norm()");
    }
    await this._py.ex`pms_LedoitWolf_error_norm = {'comp_cov': np.array(${opts["comp_cov"] ?? void 0}) if ${opts["comp_cov"] !== void 0} else None, 'norm': ${opts["norm"] ?? void 0}, 'scaling': ${opts["scaling"] ?? void 0}, 'squared': ${opts["squared"] ?? void 0}}

pms_LedoitWolf_error_norm = {k: v for k, v in pms_LedoitWolf_error_norm.items() if v is not None}`;
    await this._py.ex`res_LedoitWolf_error_norm = bridgeLedoitWolf[${this.id}].error_norm(**pms_LedoitWolf_error_norm)`;
    return this._py`res_LedoitWolf_error_norm.tolist() if hasattr(res_LedoitWolf_error_norm, 'tolist') else res_LedoitWolf_error_norm`;
  }
  /**
    Fit the Ledoit-Wolf shrunk covariance model to X.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before fit()");
    }
    await this._py.ex`pms_LedoitWolf_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_LedoitWolf_fit = {k: v for k, v in pms_LedoitWolf_fit.items() if v is not None}`;
    await this._py.ex`res_LedoitWolf_fit = bridgeLedoitWolf[${this.id}].fit(**pms_LedoitWolf_fit)`;
    return this._py`res_LedoitWolf_fit.tolist() if hasattr(res_LedoitWolf_fit, 'tolist') else res_LedoitWolf_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LedoitWolf must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_LedoitWolf_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_LedoitWolf_get_metadata_routing = {k: v for k, v in pms_LedoitWolf_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_LedoitWolf_get_metadata_routing = bridgeLedoitWolf[${this.id}].get_metadata_routing(**pms_LedoitWolf_get_metadata_routing)`;
    return this._py`res_LedoitWolf_get_metadata_routing.tolist() if hasattr(res_LedoitWolf_get_metadata_routing, 'tolist') else res_LedoitWolf_get_metadata_routing`;
  }
  /**
    Getter for the precision matrix.
   */
  async get_precision(opts) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before get_precision()");
    }
    await this._py.ex`pms_LedoitWolf_get_precision = {'precision_': np.array(${opts["precision_"] ?? void 0}) if ${opts["precision_"] !== void 0} else None}

pms_LedoitWolf_get_precision = {k: v for k, v in pms_LedoitWolf_get_precision.items() if v is not None}`;
    await this._py.ex`res_LedoitWolf_get_precision = bridgeLedoitWolf[${this.id}].get_precision(**pms_LedoitWolf_get_precision)`;
    return this._py`res_LedoitWolf_get_precision.tolist() if hasattr(res_LedoitWolf_get_precision, 'tolist') else res_LedoitWolf_get_precision`;
  }
  /**
    Compute the squared Mahalanobis distances of given observations.
   */
  async mahalanobis(opts) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before mahalanobis()");
    }
    await this._py.ex`pms_LedoitWolf_mahalanobis = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_LedoitWolf_mahalanobis = {k: v for k, v in pms_LedoitWolf_mahalanobis.items() if v is not None}`;
    await this._py.ex`res_LedoitWolf_mahalanobis = bridgeLedoitWolf[${this.id}].mahalanobis(**pms_LedoitWolf_mahalanobis)`;
    return this._py`res_LedoitWolf_mahalanobis.tolist() if hasattr(res_LedoitWolf_mahalanobis, 'tolist') else res_LedoitWolf_mahalanobis`;
  }
  /**
      Compute the log-likelihood of `X\_test` under the estimated Gaussian model.
  
      The Gaussian model is defined by its mean and covariance matrix which are represented respectively by `self.location\_` and `self.covariance\_`.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before score()");
    }
    await this._py.ex`pms_LedoitWolf_score = {'X_test': np.array(${opts["X_test"] ?? void 0}) if ${opts["X_test"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_LedoitWolf_score = {k: v for k, v in pms_LedoitWolf_score.items() if v is not None}`;
    await this._py.ex`res_LedoitWolf_score = bridgeLedoitWolf[${this.id}].score(**pms_LedoitWolf_score)`;
    return this._py`res_LedoitWolf_score.tolist() if hasattr(res_LedoitWolf_score, 'tolist') else res_LedoitWolf_score`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before set_score_request()");
    }
    await this._py.ex`pms_LedoitWolf_set_score_request = {'X_test': ${opts["X_test"] ?? void 0}}

pms_LedoitWolf_set_score_request = {k: v for k, v in pms_LedoitWolf_set_score_request.items() if v is not None}`;
    await this._py.ex`res_LedoitWolf_set_score_request = bridgeLedoitWolf[${this.id}].set_score_request(**pms_LedoitWolf_set_score_request)`;
    return this._py`res_LedoitWolf_set_score_request.tolist() if hasattr(res_LedoitWolf_set_score_request, 'tolist') else res_LedoitWolf_set_score_request`;
  }
  /**
    Estimated covariance matrix.
   */
  get covariance_() {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LedoitWolf must call init() before accessing covariance_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LedoitWolf_covariance_ = bridgeLedoitWolf[${this.id}].covariance_`;
      return this._py`attr_LedoitWolf_covariance_.tolist() if hasattr(attr_LedoitWolf_covariance_, 'tolist') else attr_LedoitWolf_covariance_`;
    })();
  }
  /**
    Estimated location, i.e. the estimated mean.
   */
  get location_() {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before accessing location_");
    }
    return (async () => {
      await this._py.ex`attr_LedoitWolf_location_ = bridgeLedoitWolf[${this.id}].location_`;
      return this._py`attr_LedoitWolf_location_.tolist() if hasattr(attr_LedoitWolf_location_, 'tolist') else attr_LedoitWolf_location_`;
    })();
  }
  /**
    Estimated pseudo inverse matrix. (stored only if store\_precision is `true`)
   */
  get precision_() {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before accessing precision_");
    }
    return (async () => {
      await this._py.ex`attr_LedoitWolf_precision_ = bridgeLedoitWolf[${this.id}].precision_`;
      return this._py`attr_LedoitWolf_precision_.tolist() if hasattr(attr_LedoitWolf_precision_, 'tolist') else attr_LedoitWolf_precision_`;
    })();
  }
  /**
    Coefficient in the convex combination used for the computation of the shrunk estimate. Range is \[0, 1\].
   */
  get shrinkage_() {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LedoitWolf must call init() before accessing shrinkage_");
    }
    return (async () => {
      await this._py.ex`attr_LedoitWolf_shrinkage_ = bridgeLedoitWolf[${this.id}].shrinkage_`;
      return this._py`attr_LedoitWolf_shrinkage_.tolist() if hasattr(attr_LedoitWolf_shrinkage_, 'tolist') else attr_LedoitWolf_shrinkage_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LedoitWolf must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LedoitWolf_n_features_in_ = bridgeLedoitWolf[${this.id}].n_features_in_`;
      return this._py`attr_LedoitWolf_n_features_in_.tolist() if hasattr(attr_LedoitWolf_n_features_in_, 'tolist') else attr_LedoitWolf_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This LedoitWolf instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LedoitWolf must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_LedoitWolf_feature_names_in_ = bridgeLedoitWolf[${this.id}].feature_names_in_`;
      return this._py`attr_LedoitWolf_feature_names_in_.tolist() if hasattr(attr_LedoitWolf_feature_names_in_, 'tolist') else attr_LedoitWolf_feature_names_in_`;
    })();
  }
};
export {
  LedoitWolf
};
