// sklearn/build/generated/covariance/OAS.js
// src/generated/covariance/OAS.ts
import crypto from "node:crypto";
var OAS = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `OAS${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("OAS.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.covariance import OAS
try: bridgeOAS
except NameError: bridgeOAS = {}
`;
    await this._py.ex`ctor_OAS = {'store_precision': ${this.opts["store_precision"] ?? void 0}, 'assume_centered': ${this.opts["assume_centered"] ?? void 0}}

ctor_OAS = {k: v for k, v in ctor_OAS.items() if v is not None}`;
    await this._py.ex`bridgeOAS[${this.id}] = OAS(**ctor_OAS)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeOAS[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Compute the Mean Squared Error between two covariance estimators.
   */
  async error_norm(opts) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before error_norm()");
    }
    await this._py.ex`pms_OAS_error_norm = {'comp_cov': np.array(${opts["comp_cov"] ?? void 0}) if ${opts["comp_cov"] !== void 0} else None, 'norm': ${opts["norm"] ?? void 0}, 'scaling': ${opts["scaling"] ?? void 0}, 'squared': ${opts["squared"] ?? void 0}}

pms_OAS_error_norm = {k: v for k, v in pms_OAS_error_norm.items() if v is not None}`;
    await this._py.ex`res_OAS_error_norm = bridgeOAS[${this.id}].error_norm(**pms_OAS_error_norm)`;
    return this._py`res_OAS_error_norm.tolist() if hasattr(res_OAS_error_norm, 'tolist') else res_OAS_error_norm`;
  }
  /**
    Fit the Oracle Approximating Shrinkage covariance model to X.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before fit()");
    }
    await this._py.ex`pms_OAS_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_OAS_fit = {k: v for k, v in pms_OAS_fit.items() if v is not None}`;
    await this._py.ex`res_OAS_fit = bridgeOAS[${this.id}].fit(**pms_OAS_fit)`;
    return this._py`res_OAS_fit.tolist() if hasattr(res_OAS_fit, 'tolist') else res_OAS_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before get_metadata_routing()");
    }
    await this._py.ex`pms_OAS_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_OAS_get_metadata_routing = {k: v for k, v in pms_OAS_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_OAS_get_metadata_routing = bridgeOAS[${this.id}].get_metadata_routing(**pms_OAS_get_metadata_routing)`;
    return this._py`res_OAS_get_metadata_routing.tolist() if hasattr(res_OAS_get_metadata_routing, 'tolist') else res_OAS_get_metadata_routing`;
  }
  /**
    Getter for the precision matrix.
   */
  async get_precision(opts) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before get_precision()");
    }
    await this._py.ex`pms_OAS_get_precision = {'precision_': np.array(${opts["precision_"] ?? void 0}) if ${opts["precision_"] !== void 0} else None}

pms_OAS_get_precision = {k: v for k, v in pms_OAS_get_precision.items() if v is not None}`;
    await this._py.ex`res_OAS_get_precision = bridgeOAS[${this.id}].get_precision(**pms_OAS_get_precision)`;
    return this._py`res_OAS_get_precision.tolist() if hasattr(res_OAS_get_precision, 'tolist') else res_OAS_get_precision`;
  }
  /**
    Compute the squared Mahalanobis distances of given observations.
   */
  async mahalanobis(opts) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before mahalanobis()");
    }
    await this._py.ex`pms_OAS_mahalanobis = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_OAS_mahalanobis = {k: v for k, v in pms_OAS_mahalanobis.items() if v is not None}`;
    await this._py.ex`res_OAS_mahalanobis = bridgeOAS[${this.id}].mahalanobis(**pms_OAS_mahalanobis)`;
    return this._py`res_OAS_mahalanobis.tolist() if hasattr(res_OAS_mahalanobis, 'tolist') else res_OAS_mahalanobis`;
  }
  /**
      Compute the log-likelihood of `X\_test` under the estimated Gaussian model.
  
      The Gaussian model is defined by its mean and covariance matrix which are represented respectively by `self.location\_` and `self.covariance\_`.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before score()");
    }
    await this._py.ex`pms_OAS_score = {'X_test': np.array(${opts["X_test"] ?? void 0}) if ${opts["X_test"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_OAS_score = {k: v for k, v in pms_OAS_score.items() if v is not None}`;
    await this._py.ex`res_OAS_score = bridgeOAS[${this.id}].score(**pms_OAS_score)`;
    return this._py`res_OAS_score.tolist() if hasattr(res_OAS_score, 'tolist') else res_OAS_score`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before set_score_request()");
    }
    await this._py.ex`pms_OAS_set_score_request = {'X_test': ${opts["X_test"] ?? void 0}}

pms_OAS_set_score_request = {k: v for k, v in pms_OAS_set_score_request.items() if v is not None}`;
    await this._py.ex`res_OAS_set_score_request = bridgeOAS[${this.id}].set_score_request(**pms_OAS_set_score_request)`;
    return this._py`res_OAS_set_score_request.tolist() if hasattr(res_OAS_set_score_request, 'tolist') else res_OAS_set_score_request`;
  }
  /**
    Estimated covariance matrix.
   */
  get covariance_() {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before accessing covariance_");
    }
    return (async () => {
      await this._py.ex`attr_OAS_covariance_ = bridgeOAS[${this.id}].covariance_`;
      return this._py`attr_OAS_covariance_.tolist() if hasattr(attr_OAS_covariance_, 'tolist') else attr_OAS_covariance_`;
    })();
  }
  /**
    Estimated location, i.e. the estimated mean.
   */
  get location_() {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before accessing location_");
    }
    return (async () => {
      await this._py.ex`attr_OAS_location_ = bridgeOAS[${this.id}].location_`;
      return this._py`attr_OAS_location_.tolist() if hasattr(attr_OAS_location_, 'tolist') else attr_OAS_location_`;
    })();
  }
  /**
    Estimated pseudo inverse matrix. (stored only if store\_precision is `true`)
   */
  get precision_() {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before accessing precision_");
    }
    return (async () => {
      await this._py.ex`attr_OAS_precision_ = bridgeOAS[${this.id}].precision_`;
      return this._py`attr_OAS_precision_.tolist() if hasattr(attr_OAS_precision_, 'tolist') else attr_OAS_precision_`;
    })();
  }
  /**
    coefficient in the convex combination used for the computation of the shrunk estimate. Range is \[0, 1\].
   */
  get shrinkage_() {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before accessing shrinkage_");
    }
    return (async () => {
      await this._py.ex`attr_OAS_shrinkage_ = bridgeOAS[${this.id}].shrinkage_`;
      return this._py`attr_OAS_shrinkage_.tolist() if hasattr(attr_OAS_shrinkage_, 'tolist') else attr_OAS_shrinkage_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before accessing n_features_in_");
    }
    return (async () => {
      await this._py.ex`attr_OAS_n_features_in_ = bridgeOAS[${this.id}].n_features_in_`;
      return this._py`attr_OAS_n_features_in_.tolist() if hasattr(attr_OAS_n_features_in_, 'tolist') else attr_OAS_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This OAS instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OAS must call init() before accessing feature_names_in_");
    }
    return (async () => {
      await this._py.ex`attr_OAS_feature_names_in_ = bridgeOAS[${this.id}].feature_names_in_`;
      return this._py`attr_OAS_feature_names_in_.tolist() if hasattr(attr_OAS_feature_names_in_, 'tolist') else attr_OAS_feature_names_in_`;
    })();
  }
};
export {
  OAS
};
