// sklearn/build/generated/metrics/DetCurveDisplay.js
// src/generated/metrics/DetCurveDisplay.ts
import crypto from "node:crypto";
var DetCurveDisplay = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `DetCurveDisplay${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This DetCurveDisplay instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("DetCurveDisplay.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.metrics import DetCurveDisplay
try: bridgeDetCurveDisplay
except NameError: bridgeDetCurveDisplay = {}
`;
    await this._py.ex`ctor_DetCurveDisplay = {'fpr': np.array(${this.opts["fpr"] ?? void 0}) if ${this.opts["fpr"] !== void 0} else None, 'fnr': np.array(${this.opts["fnr"] ?? void 0}) if ${this.opts["fnr"] !== void 0} else None, 'estimator_name': ${this.opts["estimator_name"] ?? void 0}, 'pos_label': ${this.opts["pos_label"] ?? void 0}}

ctor_DetCurveDisplay = {k: v for k, v in ctor_DetCurveDisplay.items() if v is not None}`;
    await this._py.ex`bridgeDetCurveDisplay[${this.id}] = DetCurveDisplay(**ctor_DetCurveDisplay)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeDetCurveDisplay[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Plot DET curve given an estimator and data.
  
      Read more in the [User Guide](../../visualizations.html#visualizations).
     */
  async from_estimator(opts) {
    if (this._isDisposed) {
      throw new Error("This DetCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "DetCurveDisplay must call init() before from_estimator()"
      );
    }
    await this._py.ex`pms_DetCurveDisplay_from_estimator = {'estimator': ${opts["estimator"] ?? void 0}, 'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None, 'response_method': ${opts["response_method"] ?? void 0}, 'pos_label': ${opts["pos_label"] ?? void 0}, 'name': ${opts["name"] ?? void 0}, 'ax': ${opts["ax"] ?? void 0}, 'kwargs': ${opts["kwargs"] ?? void 0}}

pms_DetCurveDisplay_from_estimator = {k: v for k, v in pms_DetCurveDisplay_from_estimator.items() if v is not None}`;
    await this._py.ex`res_DetCurveDisplay_from_estimator = bridgeDetCurveDisplay[${this.id}].from_estimator(**pms_DetCurveDisplay_from_estimator)`;
    return this._py`res_DetCurveDisplay_from_estimator.tolist() if hasattr(res_DetCurveDisplay_from_estimator, 'tolist') else res_DetCurveDisplay_from_estimator`;
  }
  /**
      Plot the DET curve given the true and predicted labels.
  
      Read more in the [User Guide](../../visualizations.html#visualizations).
     */
  async from_predictions(opts) {
    if (this._isDisposed) {
      throw new Error("This DetCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "DetCurveDisplay must call init() before from_predictions()"
      );
    }
    await this._py.ex`pms_DetCurveDisplay_from_predictions = {'y_true': np.array(${opts["y_true"] ?? void 0}) if ${opts["y_true"] !== void 0} else None, 'y_pred': np.array(${opts["y_pred"] ?? void 0}) if ${opts["y_pred"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None, 'pos_label': ${opts["pos_label"] ?? void 0}, 'name': ${opts["name"] ?? void 0}, 'ax': ${opts["ax"] ?? void 0}, 'kwargs': ${opts["kwargs"] ?? void 0}}

pms_DetCurveDisplay_from_predictions = {k: v for k, v in pms_DetCurveDisplay_from_predictions.items() if v is not None}`;
    await this._py.ex`res_DetCurveDisplay_from_predictions = bridgeDetCurveDisplay[${this.id}].from_predictions(**pms_DetCurveDisplay_from_predictions)`;
    return this._py`res_DetCurveDisplay_from_predictions.tolist() if hasattr(res_DetCurveDisplay_from_predictions, 'tolist') else res_DetCurveDisplay_from_predictions`;
  }
  /**
    Plot visualization.
   */
  async plot(opts) {
    if (this._isDisposed) {
      throw new Error("This DetCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("DetCurveDisplay must call init() before plot()");
    }
    await this._py.ex`pms_DetCurveDisplay_plot = {'ax': ${opts["ax"] ?? void 0}, 'name': ${opts["name"] ?? void 0}, 'kwargs': ${opts["kwargs"] ?? void 0}}

pms_DetCurveDisplay_plot = {k: v for k, v in pms_DetCurveDisplay_plot.items() if v is not None}`;
    await this._py.ex`res_DetCurveDisplay_plot = bridgeDetCurveDisplay[${this.id}].plot(**pms_DetCurveDisplay_plot)`;
    return this._py`res_DetCurveDisplay_plot.tolist() if hasattr(res_DetCurveDisplay_plot, 'tolist') else res_DetCurveDisplay_plot`;
  }
  /**
    DET Curve.
   */
  get line_() {
    if (this._isDisposed) {
      throw new Error("This DetCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("DetCurveDisplay must call init() before accessing line_");
    }
    return (async () => {
      await this._py.ex`attr_DetCurveDisplay_line_ = bridgeDetCurveDisplay[${this.id}].line_`;
      return this._py`attr_DetCurveDisplay_line_.tolist() if hasattr(attr_DetCurveDisplay_line_, 'tolist') else attr_DetCurveDisplay_line_`;
    })();
  }
  /**
    Axes with DET Curve.
   */
  get ax_() {
    if (this._isDisposed) {
      throw new Error("This DetCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("DetCurveDisplay must call init() before accessing ax_");
    }
    return (async () => {
      await this._py.ex`attr_DetCurveDisplay_ax_ = bridgeDetCurveDisplay[${this.id}].ax_`;
      return this._py`attr_DetCurveDisplay_ax_.tolist() if hasattr(attr_DetCurveDisplay_ax_, 'tolist') else attr_DetCurveDisplay_ax_`;
    })();
  }
  /**
    Figure containing the curve.
   */
  get figure_() {
    if (this._isDisposed) {
      throw new Error("This DetCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "DetCurveDisplay must call init() before accessing figure_"
      );
    }
    return (async () => {
      await this._py.ex`attr_DetCurveDisplay_figure_ = bridgeDetCurveDisplay[${this.id}].figure_`;
      return this._py`attr_DetCurveDisplay_figure_.tolist() if hasattr(attr_DetCurveDisplay_figure_, 'tolist') else attr_DetCurveDisplay_figure_`;
    })();
  }
};
export {
  DetCurveDisplay
};
