// sklearn/build/generated/metrics/RocCurveDisplay.js
// src/generated/metrics/RocCurveDisplay.ts
import crypto from "node:crypto";
var RocCurveDisplay = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `RocCurveDisplay${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("RocCurveDisplay.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.metrics import RocCurveDisplay
try: bridgeRocCurveDisplay
except NameError: bridgeRocCurveDisplay = {}
`;
    await this._py.ex`ctor_RocCurveDisplay = {'fpr': np.array(${this.opts["fpr"] ?? void 0}) if ${this.opts["fpr"] !== void 0} else None, 'tpr': np.array(${this.opts["tpr"] ?? void 0}) if ${this.opts["tpr"] !== void 0} else None, 'roc_auc': ${this.opts["roc_auc"] ?? void 0}, 'estimator_name': ${this.opts["estimator_name"] ?? void 0}, 'pos_label': ${this.opts["pos_label"] ?? void 0}}

ctor_RocCurveDisplay = {k: v for k, v in ctor_RocCurveDisplay.items() if v is not None}`;
    await this._py.ex`bridgeRocCurveDisplay[${this.id}] = RocCurveDisplay(**ctor_RocCurveDisplay)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeRocCurveDisplay[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Create a ROC Curve display from an estimator.
   */
  async from_estimator(opts) {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RocCurveDisplay must call init() before from_estimator()"
      );
    }
    await this._py.ex`pms_RocCurveDisplay_from_estimator = {'estimator': ${opts["estimator"] ?? void 0}, 'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None, 'drop_intermediate': ${opts["drop_intermediate"] ?? void 0}, 'response_method': ${opts["response_method"] ?? void 0}, 'pos_label': ${opts["pos_label"] ?? void 0}, 'name': ${opts["name"] ?? void 0}, 'ax': ${opts["ax"] ?? void 0}, 'plot_chance_level': ${opts["plot_chance_level"] ?? void 0}, 'chance_level_kw': ${opts["chance_level_kw"] ?? void 0}, 'kwargs': ${opts["kwargs"] ?? void 0}}

pms_RocCurveDisplay_from_estimator = {k: v for k, v in pms_RocCurveDisplay_from_estimator.items() if v is not None}`;
    await this._py.ex`res_RocCurveDisplay_from_estimator = bridgeRocCurveDisplay[${this.id}].from_estimator(**pms_RocCurveDisplay_from_estimator)`;
    return this._py`res_RocCurveDisplay_from_estimator.tolist() if hasattr(res_RocCurveDisplay_from_estimator, 'tolist') else res_RocCurveDisplay_from_estimator`;
  }
  /**
      Plot ROC curve given the true and predicted values.
  
      Read more in the [User Guide](../../visualizations.html#visualizations).
     */
  async from_predictions(opts) {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RocCurveDisplay must call init() before from_predictions()"
      );
    }
    await this._py.ex`pms_RocCurveDisplay_from_predictions = {'y_true': np.array(${opts["y_true"] ?? void 0}) if ${opts["y_true"] !== void 0} else None, 'y_pred': np.array(${opts["y_pred"] ?? void 0}) if ${opts["y_pred"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None, 'drop_intermediate': ${opts["drop_intermediate"] ?? void 0}, 'pos_label': ${opts["pos_label"] ?? void 0}, 'name': ${opts["name"] ?? void 0}, 'ax': ${opts["ax"] ?? void 0}, 'plot_chance_level': ${opts["plot_chance_level"] ?? void 0}, 'chance_level_kw': ${opts["chance_level_kw"] ?? void 0}, 'kwargs': ${opts["kwargs"] ?? void 0}}

pms_RocCurveDisplay_from_predictions = {k: v for k, v in pms_RocCurveDisplay_from_predictions.items() if v is not None}`;
    await this._py.ex`res_RocCurveDisplay_from_predictions = bridgeRocCurveDisplay[${this.id}].from_predictions(**pms_RocCurveDisplay_from_predictions)`;
    return this._py`res_RocCurveDisplay_from_predictions.tolist() if hasattr(res_RocCurveDisplay_from_predictions, 'tolist') else res_RocCurveDisplay_from_predictions`;
  }
  /**
      Plot visualization.
  
      Extra keyword arguments will be passed to matplotlib’s `plot`.
     */
  async plot(opts) {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RocCurveDisplay must call init() before plot()");
    }
    await this._py.ex`pms_RocCurveDisplay_plot = {'ax': ${opts["ax"] ?? void 0}, 'name': ${opts["name"] ?? void 0}, 'plot_chance_level': ${opts["plot_chance_level"] ?? void 0}, 'chance_level_kw': ${opts["chance_level_kw"] ?? void 0}, 'kwargs': ${opts["kwargs"] ?? void 0}}

pms_RocCurveDisplay_plot = {k: v for k, v in pms_RocCurveDisplay_plot.items() if v is not None}`;
    await this._py.ex`res_RocCurveDisplay_plot = bridgeRocCurveDisplay[${this.id}].plot(**pms_RocCurveDisplay_plot)`;
    return this._py`res_RocCurveDisplay_plot.tolist() if hasattr(res_RocCurveDisplay_plot, 'tolist') else res_RocCurveDisplay_plot`;
  }
  /**
    ROC Curve.
   */
  get line_() {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RocCurveDisplay must call init() before accessing line_");
    }
    return (async () => {
      await this._py.ex`attr_RocCurveDisplay_line_ = bridgeRocCurveDisplay[${this.id}].line_`;
      return this._py`attr_RocCurveDisplay_line_.tolist() if hasattr(attr_RocCurveDisplay_line_, 'tolist') else attr_RocCurveDisplay_line_`;
    })();
  }
  /**
    The chance level line. It is `undefined` if the chance level is not plotted.
   */
  get chance_level_() {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RocCurveDisplay must call init() before accessing chance_level_"
      );
    }
    return (async () => {
      await this._py.ex`attr_RocCurveDisplay_chance_level_ = bridgeRocCurveDisplay[${this.id}].chance_level_`;
      return this._py`attr_RocCurveDisplay_chance_level_.tolist() if hasattr(attr_RocCurveDisplay_chance_level_, 'tolist') else attr_RocCurveDisplay_chance_level_`;
    })();
  }
  /**
    Axes with ROC Curve.
   */
  get ax_() {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RocCurveDisplay must call init() before accessing ax_");
    }
    return (async () => {
      await this._py.ex`attr_RocCurveDisplay_ax_ = bridgeRocCurveDisplay[${this.id}].ax_`;
      return this._py`attr_RocCurveDisplay_ax_.tolist() if hasattr(attr_RocCurveDisplay_ax_, 'tolist') else attr_RocCurveDisplay_ax_`;
    })();
  }
  /**
    Figure containing the curve.
   */
  get figure_() {
    if (this._isDisposed) {
      throw new Error("This RocCurveDisplay instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RocCurveDisplay must call init() before accessing figure_"
      );
    }
    return (async () => {
      await this._py.ex`attr_RocCurveDisplay_figure_ = bridgeRocCurveDisplay[${this.id}].figure_`;
      return this._py`attr_RocCurveDisplay_figure_.tolist() if hasattr(attr_RocCurveDisplay_figure_, 'tolist') else attr_RocCurveDisplay_figure_`;
    })();
  }
};
export {
  RocCurveDisplay
};
