// sklearn/build/generated/tree/ExtraTreeClassifier.js
// src/generated/tree/ExtraTreeClassifier.ts
import crypto from "node:crypto";
var ExtraTreeClassifier = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `ExtraTreeClassifier${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error(
        "ExtraTreeClassifier.init requires a PythonBridge instance"
      );
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.tree import ExtraTreeClassifier
try: bridgeExtraTreeClassifier
except NameError: bridgeExtraTreeClassifier = {}
`;
    await this._py.ex`ctor_ExtraTreeClassifier = {'criterion': ${this.opts["criterion"] ?? void 0}, 'splitter': ${this.opts["splitter"] ?? void 0}, 'max_depth': ${this.opts["max_depth"] ?? void 0}, 'min_samples_split': ${this.opts["min_samples_split"] ?? void 0}, 'min_samples_leaf': ${this.opts["min_samples_leaf"] ?? void 0}, 'min_weight_fraction_leaf': ${this.opts["min_weight_fraction_leaf"] ?? void 0}, 'max_features': ${this.opts["max_features"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}, 'max_leaf_nodes': ${this.opts["max_leaf_nodes"] ?? void 0}, 'min_impurity_decrease': ${this.opts["min_impurity_decrease"] ?? void 0}, 'class_weight': ${this.opts["class_weight"] ?? void 0}, 'ccp_alpha': ${this.opts["ccp_alpha"] ?? void 0}}

ctor_ExtraTreeClassifier = {k: v for k, v in ctor_ExtraTreeClassifier.items() if v is not None}`;
    await this._py.ex`bridgeExtraTreeClassifier[${this.id}] = ExtraTreeClassifier(**ctor_ExtraTreeClassifier)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeExtraTreeClassifier[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Return the index of the leaf that each sample is predicted as.
   */
  async apply(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("ExtraTreeClassifier must call init() before apply()");
    }
    await this._py.ex`pms_ExtraTreeClassifier_apply = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'check_input': ${opts["check_input"] ?? void 0}}

pms_ExtraTreeClassifier_apply = {k: v for k, v in pms_ExtraTreeClassifier_apply.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_apply = bridgeExtraTreeClassifier[${this.id}].apply(**pms_ExtraTreeClassifier_apply)`;
    return this._py`res_ExtraTreeClassifier_apply.tolist() if hasattr(res_ExtraTreeClassifier_apply, 'tolist') else res_ExtraTreeClassifier_apply`;
  }
  /**
      Compute the pruning path during Minimal Cost-Complexity Pruning.
  
      See [Minimal Cost-Complexity Pruning](../tree.html#minimal-cost-complexity-pruning) for details on the pruning process.
     */
  async cost_complexity_pruning_path(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before cost_complexity_pruning_path()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_cost_complexity_pruning_path = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_ExtraTreeClassifier_cost_complexity_pruning_path = {k: v for k, v in pms_ExtraTreeClassifier_cost_complexity_pruning_path.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_cost_complexity_pruning_path = bridgeExtraTreeClassifier[${this.id}].cost_complexity_pruning_path(**pms_ExtraTreeClassifier_cost_complexity_pruning_path)`;
    return this._py`res_ExtraTreeClassifier_cost_complexity_pruning_path.tolist() if hasattr(res_ExtraTreeClassifier_cost_complexity_pruning_path, 'tolist') else res_ExtraTreeClassifier_cost_complexity_pruning_path`;
  }
  /**
    Return the decision path in the tree.
   */
  async decision_path(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before decision_path()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_decision_path = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'check_input': ${opts["check_input"] ?? void 0}}

pms_ExtraTreeClassifier_decision_path = {k: v for k, v in pms_ExtraTreeClassifier_decision_path.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_decision_path = bridgeExtraTreeClassifier[${this.id}].decision_path(**pms_ExtraTreeClassifier_decision_path)`;
    return this._py`res_ExtraTreeClassifier_decision_path.tolist() if hasattr(res_ExtraTreeClassifier_decision_path, 'tolist') else res_ExtraTreeClassifier_decision_path`;
  }
  /**
    Build a decision tree classifier from the training set (X, y).
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("ExtraTreeClassifier must call init() before fit()");
    }
    await this._py.ex`pms_ExtraTreeClassifier_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None, 'check_input': ${opts["check_input"] ?? void 0}}

pms_ExtraTreeClassifier_fit = {k: v for k, v in pms_ExtraTreeClassifier_fit.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_fit = bridgeExtraTreeClassifier[${this.id}].fit(**pms_ExtraTreeClassifier_fit)`;
    return this._py`res_ExtraTreeClassifier_fit.tolist() if hasattr(res_ExtraTreeClassifier_fit, 'tolist') else res_ExtraTreeClassifier_fit`;
  }
  /**
      Return the depth of the decision tree.
  
      The depth of a tree is the maximum distance between the root and any leaf.
     */
  async get_depth(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("ExtraTreeClassifier must call init() before get_depth()");
    }
    await this._py.ex`pms_ExtraTreeClassifier_get_depth = {}

pms_ExtraTreeClassifier_get_depth = {k: v for k, v in pms_ExtraTreeClassifier_get_depth.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_get_depth = bridgeExtraTreeClassifier[${this.id}].get_depth(**pms_ExtraTreeClassifier_get_depth)`;
    return this._py`res_ExtraTreeClassifier_get_depth.tolist() if hasattr(res_ExtraTreeClassifier_get_depth, 'tolist') else res_ExtraTreeClassifier_get_depth`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_ExtraTreeClassifier_get_metadata_routing = {k: v for k, v in pms_ExtraTreeClassifier_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_get_metadata_routing = bridgeExtraTreeClassifier[${this.id}].get_metadata_routing(**pms_ExtraTreeClassifier_get_metadata_routing)`;
    return this._py`res_ExtraTreeClassifier_get_metadata_routing.tolist() if hasattr(res_ExtraTreeClassifier_get_metadata_routing, 'tolist') else res_ExtraTreeClassifier_get_metadata_routing`;
  }
  /**
    Return the number of leaves of the decision tree.
   */
  async get_n_leaves(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before get_n_leaves()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_get_n_leaves = {}

pms_ExtraTreeClassifier_get_n_leaves = {k: v for k, v in pms_ExtraTreeClassifier_get_n_leaves.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_get_n_leaves = bridgeExtraTreeClassifier[${this.id}].get_n_leaves(**pms_ExtraTreeClassifier_get_n_leaves)`;
    return this._py`res_ExtraTreeClassifier_get_n_leaves.tolist() if hasattr(res_ExtraTreeClassifier_get_n_leaves, 'tolist') else res_ExtraTreeClassifier_get_n_leaves`;
  }
  /**
      Predict class or regression value for X.
  
      For a classification model, the predicted class for each sample in X is returned. For a regression model, the predicted value based on X is returned.
     */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("ExtraTreeClassifier must call init() before predict()");
    }
    await this._py.ex`pms_ExtraTreeClassifier_predict = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'check_input': ${opts["check_input"] ?? void 0}}

pms_ExtraTreeClassifier_predict = {k: v for k, v in pms_ExtraTreeClassifier_predict.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_predict = bridgeExtraTreeClassifier[${this.id}].predict(**pms_ExtraTreeClassifier_predict)`;
    return this._py`res_ExtraTreeClassifier_predict.tolist() if hasattr(res_ExtraTreeClassifier_predict, 'tolist') else res_ExtraTreeClassifier_predict`;
  }
  /**
    Predict class log-probabilities of the input samples X.
   */
  async predict_log_proba(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before predict_log_proba()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_predict_log_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_ExtraTreeClassifier_predict_log_proba = {k: v for k, v in pms_ExtraTreeClassifier_predict_log_proba.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_predict_log_proba = bridgeExtraTreeClassifier[${this.id}].predict_log_proba(**pms_ExtraTreeClassifier_predict_log_proba)`;
    return this._py`res_ExtraTreeClassifier_predict_log_proba.tolist() if hasattr(res_ExtraTreeClassifier_predict_log_proba, 'tolist') else res_ExtraTreeClassifier_predict_log_proba`;
  }
  /**
      Predict class probabilities of the input samples X.
  
      The predicted class probability is the fraction of samples of the same class in a leaf.
     */
  async predict_proba(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before predict_proba()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_predict_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'check_input': ${opts["check_input"] ?? void 0}}

pms_ExtraTreeClassifier_predict_proba = {k: v for k, v in pms_ExtraTreeClassifier_predict_proba.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_predict_proba = bridgeExtraTreeClassifier[${this.id}].predict_proba(**pms_ExtraTreeClassifier_predict_proba)`;
    return this._py`res_ExtraTreeClassifier_predict_proba.tolist() if hasattr(res_ExtraTreeClassifier_predict_proba, 'tolist') else res_ExtraTreeClassifier_predict_proba`;
  }
  /**
      Return the mean accuracy on the given test data and labels.
  
      In multi-label classification, this is the subset accuracy which is a harsh metric since you require for each sample that each label set be correctly predicted.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("ExtraTreeClassifier must call init() before score()");
    }
    await this._py.ex`pms_ExtraTreeClassifier_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_ExtraTreeClassifier_score = {k: v for k, v in pms_ExtraTreeClassifier_score.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_score = bridgeExtraTreeClassifier[${this.id}].score(**pms_ExtraTreeClassifier_score)`;
    return this._py`res_ExtraTreeClassifier_score.tolist() if hasattr(res_ExtraTreeClassifier_score, 'tolist') else res_ExtraTreeClassifier_score`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before set_fit_request()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_set_fit_request = {'check_input': ${opts["check_input"] ?? void 0}, 'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_ExtraTreeClassifier_set_fit_request = {k: v for k, v in pms_ExtraTreeClassifier_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_set_fit_request = bridgeExtraTreeClassifier[${this.id}].set_fit_request(**pms_ExtraTreeClassifier_set_fit_request)`;
    return this._py`res_ExtraTreeClassifier_set_fit_request.tolist() if hasattr(res_ExtraTreeClassifier_set_fit_request, 'tolist') else res_ExtraTreeClassifier_set_fit_request`;
  }
  /**
      Request metadata passed to the `predict\_proba` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_predict_proba_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before set_predict_proba_request()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_set_predict_proba_request = {'check_input': ${opts["check_input"] ?? void 0}}

pms_ExtraTreeClassifier_set_predict_proba_request = {k: v for k, v in pms_ExtraTreeClassifier_set_predict_proba_request.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_set_predict_proba_request = bridgeExtraTreeClassifier[${this.id}].set_predict_proba_request(**pms_ExtraTreeClassifier_set_predict_proba_request)`;
    return this._py`res_ExtraTreeClassifier_set_predict_proba_request.tolist() if hasattr(res_ExtraTreeClassifier_set_predict_proba_request, 'tolist') else res_ExtraTreeClassifier_set_predict_proba_request`;
  }
  /**
      Request metadata passed to the `predict` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_predict_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before set_predict_request()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_set_predict_request = {'check_input': ${opts["check_input"] ?? void 0}}

pms_ExtraTreeClassifier_set_predict_request = {k: v for k, v in pms_ExtraTreeClassifier_set_predict_request.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_set_predict_request = bridgeExtraTreeClassifier[${this.id}].set_predict_request(**pms_ExtraTreeClassifier_set_predict_request)`;
    return this._py`res_ExtraTreeClassifier_set_predict_request.tolist() if hasattr(res_ExtraTreeClassifier_set_predict_request, 'tolist') else res_ExtraTreeClassifier_set_predict_request`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before set_score_request()"
      );
    }
    await this._py.ex`pms_ExtraTreeClassifier_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_ExtraTreeClassifier_set_score_request = {k: v for k, v in pms_ExtraTreeClassifier_set_score_request.items() if v is not None}`;
    await this._py.ex`res_ExtraTreeClassifier_set_score_request = bridgeExtraTreeClassifier[${this.id}].set_score_request(**pms_ExtraTreeClassifier_set_score_request)`;
    return this._py`res_ExtraTreeClassifier_set_score_request.tolist() if hasattr(res_ExtraTreeClassifier_set_score_request, 'tolist') else res_ExtraTreeClassifier_set_score_request`;
  }
  /**
    The classes labels (single output problem), or a list of arrays of class labels (multi-output problem).
   */
  get classes_() {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before accessing classes_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ExtraTreeClassifier_classes_ = bridgeExtraTreeClassifier[${this.id}].classes_`;
      return this._py`attr_ExtraTreeClassifier_classes_.tolist() if hasattr(attr_ExtraTreeClassifier_classes_, 'tolist') else attr_ExtraTreeClassifier_classes_`;
    })();
  }
  /**
    The inferred value of max\_features.
   */
  get max_features_() {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before accessing max_features_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ExtraTreeClassifier_max_features_ = bridgeExtraTreeClassifier[${this.id}].max_features_`;
      return this._py`attr_ExtraTreeClassifier_max_features_.tolist() if hasattr(attr_ExtraTreeClassifier_max_features_, 'tolist') else attr_ExtraTreeClassifier_max_features_`;
    })();
  }
  /**
    The number of classes (for single output problems), or a list containing the number of classes for each output (for multi-output problems).
   */
  get n_classes_() {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before accessing n_classes_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ExtraTreeClassifier_n_classes_ = bridgeExtraTreeClassifier[${this.id}].n_classes_`;
      return this._py`attr_ExtraTreeClassifier_n_classes_.tolist() if hasattr(attr_ExtraTreeClassifier_n_classes_, 'tolist') else attr_ExtraTreeClassifier_n_classes_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ExtraTreeClassifier_n_features_in_ = bridgeExtraTreeClassifier[${this.id}].n_features_in_`;
      return this._py`attr_ExtraTreeClassifier_n_features_in_.tolist() if hasattr(attr_ExtraTreeClassifier_n_features_in_, 'tolist') else attr_ExtraTreeClassifier_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ExtraTreeClassifier_feature_names_in_ = bridgeExtraTreeClassifier[${this.id}].feature_names_in_`;
      return this._py`attr_ExtraTreeClassifier_feature_names_in_.tolist() if hasattr(attr_ExtraTreeClassifier_feature_names_in_, 'tolist') else attr_ExtraTreeClassifier_feature_names_in_`;
    })();
  }
  /**
    The number of outputs when `fit` is performed.
   */
  get n_outputs_() {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before accessing n_outputs_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ExtraTreeClassifier_n_outputs_ = bridgeExtraTreeClassifier[${this.id}].n_outputs_`;
      return this._py`attr_ExtraTreeClassifier_n_outputs_.tolist() if hasattr(attr_ExtraTreeClassifier_n_outputs_, 'tolist') else attr_ExtraTreeClassifier_n_outputs_`;
    })();
  }
  /**
    The underlying Tree object. Please refer to `help(sklearn.tree.\_tree.Tree)` for attributes of Tree object and [Understanding the decision tree structure](../../auto_examples/tree/plot_unveil_tree_structure.html#sphx-glr-auto-examples-tree-plot-unveil-tree-structure-py) for basic usage of these attributes.
   */
  get tree_() {
    if (this._isDisposed) {
      throw new Error(
        "This ExtraTreeClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "ExtraTreeClassifier must call init() before accessing tree_"
      );
    }
    return (async () => {
      await this._py.ex`attr_ExtraTreeClassifier_tree_ = bridgeExtraTreeClassifier[${this.id}].tree_`;
      return this._py`attr_ExtraTreeClassifier_tree_.tolist() if hasattr(attr_ExtraTreeClassifier_tree_, 'tolist') else attr_ExtraTreeClassifier_tree_`;
    })();
  }
};
export {
  ExtraTreeClassifier
};
