// sklearn/build/generated/preprocessing/FunctionTransformer.js
// src/generated/preprocessing/FunctionTransformer.ts
import crypto from "node:crypto";
var FunctionTransformer = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `FunctionTransformer${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error(
        "FunctionTransformer.init requires a PythonBridge instance"
      );
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.preprocessing import FunctionTransformer
try: bridgeFunctionTransformer
except NameError: bridgeFunctionTransformer = {}
`;
    await this._py.ex`ctor_FunctionTransformer = {'func': ${this.opts["func"] ?? void 0}, 'inverse_func': ${this.opts["inverse_func"] ?? void 0}, 'validate': ${this.opts["validate"] ?? void 0}, 'accept_sparse': ${this.opts["accept_sparse"] ?? void 0}, 'check_inverse': ${this.opts["check_inverse"] ?? void 0}, 'feature_names_out': ${this.opts["feature_names_out"] ?? void 0}, 'kw_args': ${this.opts["kw_args"] ?? void 0}, 'inv_kw_args': ${this.opts["inv_kw_args"] ?? void 0}}

ctor_FunctionTransformer = {k: v for k, v in ctor_FunctionTransformer.items() if v is not None}`;
    await this._py.ex`bridgeFunctionTransformer[${this.id}] = FunctionTransformer(**ctor_FunctionTransformer)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeFunctionTransformer[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Fit transformer by checking X.
  
      If `validate` is `true`, `X` will be checked.
     */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("FunctionTransformer must call init() before fit()");
    }
    await this._py.ex`pms_FunctionTransformer_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_FunctionTransformer_fit = {k: v for k, v in pms_FunctionTransformer_fit.items() if v is not None}`;
    await this._py.ex`res_FunctionTransformer_fit = bridgeFunctionTransformer[${this.id}].fit(**pms_FunctionTransformer_fit)`;
    return this._py`res_FunctionTransformer_fit.tolist() if hasattr(res_FunctionTransformer_fit, 'tolist') else res_FunctionTransformer_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FunctionTransformer must call init() before fit_transform()"
      );
    }
    await this._py.ex`pms_FunctionTransformer_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_FunctionTransformer_fit_transform = {k: v for k, v in pms_FunctionTransformer_fit_transform.items() if v is not None}`;
    await this._py.ex`res_FunctionTransformer_fit_transform = bridgeFunctionTransformer[${this.id}].fit_transform(**pms_FunctionTransformer_fit_transform)`;
    return this._py`res_FunctionTransformer_fit_transform.tolist() if hasattr(res_FunctionTransformer_fit_transform, 'tolist') else res_FunctionTransformer_fit_transform`;
  }
  /**
      Get output feature names for transformation.
  
      This method is only defined if `feature\_names\_out` is not `undefined`.
     */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FunctionTransformer must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_FunctionTransformer_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_FunctionTransformer_get_feature_names_out = {k: v for k, v in pms_FunctionTransformer_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_FunctionTransformer_get_feature_names_out = bridgeFunctionTransformer[${this.id}].get_feature_names_out(**pms_FunctionTransformer_get_feature_names_out)`;
    return this._py`res_FunctionTransformer_get_feature_names_out.tolist() if hasattr(res_FunctionTransformer_get_feature_names_out, 'tolist') else res_FunctionTransformer_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FunctionTransformer must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_FunctionTransformer_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_FunctionTransformer_get_metadata_routing = {k: v for k, v in pms_FunctionTransformer_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_FunctionTransformer_get_metadata_routing = bridgeFunctionTransformer[${this.id}].get_metadata_routing(**pms_FunctionTransformer_get_metadata_routing)`;
    return this._py`res_FunctionTransformer_get_metadata_routing.tolist() if hasattr(res_FunctionTransformer_get_metadata_routing, 'tolist') else res_FunctionTransformer_get_metadata_routing`;
  }
  /**
    Transform X using the inverse function.
   */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FunctionTransformer must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_FunctionTransformer_inverse_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_FunctionTransformer_inverse_transform = {k: v for k, v in pms_FunctionTransformer_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_FunctionTransformer_inverse_transform = bridgeFunctionTransformer[${this.id}].inverse_transform(**pms_FunctionTransformer_inverse_transform)`;
    return this._py`res_FunctionTransformer_inverse_transform.tolist() if hasattr(res_FunctionTransformer_inverse_transform, 'tolist') else res_FunctionTransformer_inverse_transform`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FunctionTransformer must call init() before set_output()"
      );
    }
    await this._py.ex`pms_FunctionTransformer_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_FunctionTransformer_set_output = {k: v for k, v in pms_FunctionTransformer_set_output.items() if v is not None}`;
    await this._py.ex`res_FunctionTransformer_set_output = bridgeFunctionTransformer[${this.id}].set_output(**pms_FunctionTransformer_set_output)`;
    return this._py`res_FunctionTransformer_set_output.tolist() if hasattr(res_FunctionTransformer_set_output, 'tolist') else res_FunctionTransformer_set_output`;
  }
  /**
    Transform X using the forward function.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("FunctionTransformer must call init() before transform()");
    }
    await this._py.ex`pms_FunctionTransformer_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_FunctionTransformer_transform = {k: v for k, v in pms_FunctionTransformer_transform.items() if v is not None}`;
    await this._py.ex`res_FunctionTransformer_transform = bridgeFunctionTransformer[${this.id}].transform(**pms_FunctionTransformer_transform)`;
    return this._py`res_FunctionTransformer_transform.tolist() if hasattr(res_FunctionTransformer_transform, 'tolist') else res_FunctionTransformer_transform`;
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FunctionTransformer must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_FunctionTransformer_n_features_in_ = bridgeFunctionTransformer[${this.id}].n_features_in_`;
      return this._py`attr_FunctionTransformer_n_features_in_.tolist() if hasattr(attr_FunctionTransformer_n_features_in_, 'tolist') else attr_FunctionTransformer_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This FunctionTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FunctionTransformer must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_FunctionTransformer_feature_names_in_ = bridgeFunctionTransformer[${this.id}].feature_names_in_`;
      return this._py`attr_FunctionTransformer_feature_names_in_.tolist() if hasattr(attr_FunctionTransformer_feature_names_in_, 'tolist') else attr_FunctionTransformer_feature_names_in_`;
    })();
  }
};
export {
  FunctionTransformer
};
