// sklearn/build/generated/preprocessing/OneHotEncoder.js
// src/generated/preprocessing/OneHotEncoder.ts
import crypto from "node:crypto";
var OneHotEncoder = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `OneHotEncoder${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("OneHotEncoder.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.preprocessing import OneHotEncoder
try: bridgeOneHotEncoder
except NameError: bridgeOneHotEncoder = {}
`;
    await this._py.ex`ctor_OneHotEncoder = {'categories': ${this.opts["categories"] ?? void 0}, 'drop': np.array(${this.opts["drop"] ?? void 0}) if ${this.opts["drop"] !== void 0} else None, 'sparse': ${this.opts["sparse"] ?? void 0}, 'sparse_output': ${this.opts["sparse_output"] ?? void 0}, 'dtype': ${this.opts["dtype"] ?? void 0}, 'handle_unknown': ${this.opts["handle_unknown"] ?? void 0}, 'min_frequency': ${this.opts["min_frequency"] ?? void 0}, 'max_categories': ${this.opts["max_categories"] ?? void 0}, 'feature_name_combiner': ${this.opts["feature_name_combiner"] ?? void 0}}

ctor_OneHotEncoder = {k: v for k, v in ctor_OneHotEncoder.items() if v is not None}`;
    await this._py.ex`bridgeOneHotEncoder[${this.id}] = OneHotEncoder(**ctor_OneHotEncoder)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeOneHotEncoder[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit OneHotEncoder to X.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OneHotEncoder must call init() before fit()");
    }
    await this._py.ex`pms_OneHotEncoder_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_OneHotEncoder_fit = {k: v for k, v in pms_OneHotEncoder_fit.items() if v is not None}`;
    await this._py.ex`res_OneHotEncoder_fit = bridgeOneHotEncoder[${this.id}].fit(**pms_OneHotEncoder_fit)`;
    return this._py`res_OneHotEncoder_fit.tolist() if hasattr(res_OneHotEncoder_fit, 'tolist') else res_OneHotEncoder_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OneHotEncoder must call init() before fit_transform()");
    }
    await this._py.ex`pms_OneHotEncoder_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_OneHotEncoder_fit_transform = {k: v for k, v in pms_OneHotEncoder_fit_transform.items() if v is not None}`;
    await this._py.ex`res_OneHotEncoder_fit_transform = bridgeOneHotEncoder[${this.id}].fit_transform(**pms_OneHotEncoder_fit_transform)`;
    return this._py`res_OneHotEncoder_fit_transform.tolist() if hasattr(res_OneHotEncoder_fit_transform, 'tolist') else res_OneHotEncoder_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_OneHotEncoder_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_OneHotEncoder_get_feature_names_out = {k: v for k, v in pms_OneHotEncoder_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_OneHotEncoder_get_feature_names_out = bridgeOneHotEncoder[${this.id}].get_feature_names_out(**pms_OneHotEncoder_get_feature_names_out)`;
    return this._py`res_OneHotEncoder_get_feature_names_out.tolist() if hasattr(res_OneHotEncoder_get_feature_names_out, 'tolist') else res_OneHotEncoder_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_OneHotEncoder_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_OneHotEncoder_get_metadata_routing = {k: v for k, v in pms_OneHotEncoder_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_OneHotEncoder_get_metadata_routing = bridgeOneHotEncoder[${this.id}].get_metadata_routing(**pms_OneHotEncoder_get_metadata_routing)`;
    return this._py`res_OneHotEncoder_get_metadata_routing.tolist() if hasattr(res_OneHotEncoder_get_metadata_routing, 'tolist') else res_OneHotEncoder_get_metadata_routing`;
  }
  /**
      Convert the data back to the original representation.
  
      When unknown categories are encountered (all zeros in the one-hot encoding), `undefined` is used to represent this category. If the feature with the unknown category has a dropped category, the dropped category will be its inverse.
  
      For a given input feature, if there is an infrequent category, ‘infrequent\_sklearn’ will be used to represent the infrequent category.
     */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_OneHotEncoder_inverse_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_OneHotEncoder_inverse_transform = {k: v for k, v in pms_OneHotEncoder_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_OneHotEncoder_inverse_transform = bridgeOneHotEncoder[${this.id}].inverse_transform(**pms_OneHotEncoder_inverse_transform)`;
    return this._py`res_OneHotEncoder_inverse_transform.tolist() if hasattr(res_OneHotEncoder_inverse_transform, 'tolist') else res_OneHotEncoder_inverse_transform`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OneHotEncoder must call init() before set_output()");
    }
    await this._py.ex`pms_OneHotEncoder_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_OneHotEncoder_set_output = {k: v for k, v in pms_OneHotEncoder_set_output.items() if v is not None}`;
    await this._py.ex`res_OneHotEncoder_set_output = bridgeOneHotEncoder[${this.id}].set_output(**pms_OneHotEncoder_set_output)`;
    return this._py`res_OneHotEncoder_set_output.tolist() if hasattr(res_OneHotEncoder_set_output, 'tolist') else res_OneHotEncoder_set_output`;
  }
  /**
      Transform X using one-hot encoding.
  
      If there are infrequent categories for a feature, the infrequent categories will be grouped into a single category.
     */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OneHotEncoder must call init() before transform()");
    }
    await this._py.ex`pms_OneHotEncoder_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_OneHotEncoder_transform = {k: v for k, v in pms_OneHotEncoder_transform.items() if v is not None}`;
    await this._py.ex`res_OneHotEncoder_transform = bridgeOneHotEncoder[${this.id}].transform(**pms_OneHotEncoder_transform)`;
    return this._py`res_OneHotEncoder_transform.tolist() if hasattr(res_OneHotEncoder_transform, 'tolist') else res_OneHotEncoder_transform`;
  }
  /**
    The categories of each feature determined during fitting (in order of the features in X and corresponding with the output of `transform`). This includes the category specified in `drop` (if any).
   */
  get categories_() {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before accessing categories_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OneHotEncoder_categories_ = bridgeOneHotEncoder[${this.id}].categories_`;
      return this._py`attr_OneHotEncoder_categories_.tolist() if hasattr(attr_OneHotEncoder_categories_, 'tolist') else attr_OneHotEncoder_categories_`;
    })();
  }
  /**
    `drop\_idx\_\[i\]` is the index in `categories\_\[i\]` of the category to be dropped for each feature.
   */
  get drop_idx_() {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before accessing drop_idx_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OneHotEncoder_drop_idx_ = bridgeOneHotEncoder[${this.id}].drop_idx_`;
      return this._py`attr_OneHotEncoder_drop_idx_.tolist() if hasattr(attr_OneHotEncoder_drop_idx_, 'tolist') else attr_OneHotEncoder_drop_idx_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OneHotEncoder_n_features_in_ = bridgeOneHotEncoder[${this.id}].n_features_in_`;
      return this._py`attr_OneHotEncoder_n_features_in_.tolist() if hasattr(attr_OneHotEncoder_n_features_in_, 'tolist') else attr_OneHotEncoder_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OneHotEncoder_feature_names_in_ = bridgeOneHotEncoder[${this.id}].feature_names_in_`;
      return this._py`attr_OneHotEncoder_feature_names_in_.tolist() if hasattr(attr_OneHotEncoder_feature_names_in_, 'tolist') else attr_OneHotEncoder_feature_names_in_`;
    })();
  }
  /**
    Callable with signature `def callable(input\_feature, category)` that returns a string. This is used to create feature names to be returned by [`get\_feature\_names\_out`](#sklearn.preprocessing.OneHotEncoder.get_feature_names_out "sklearn.preprocessing.OneHotEncoder.get_feature_names_out").
   */
  get feature_name_combiner() {
    if (this._isDisposed) {
      throw new Error("This OneHotEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OneHotEncoder must call init() before accessing feature_name_combiner"
      );
    }
    return (async () => {
      await this._py.ex`attr_OneHotEncoder_feature_name_combiner = bridgeOneHotEncoder[${this.id}].feature_name_combiner`;
      return this._py`attr_OneHotEncoder_feature_name_combiner.tolist() if hasattr(attr_OneHotEncoder_feature_name_combiner, 'tolist') else attr_OneHotEncoder_feature_name_combiner`;
    })();
  }
};
export {
  OneHotEncoder
};
