// sklearn/build/generated/preprocessing/OrdinalEncoder.js
// src/generated/preprocessing/OrdinalEncoder.ts
import crypto from "node:crypto";
var OrdinalEncoder = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `OrdinalEncoder${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("OrdinalEncoder.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.preprocessing import OrdinalEncoder
try: bridgeOrdinalEncoder
except NameError: bridgeOrdinalEncoder = {}
`;
    await this._py.ex`ctor_OrdinalEncoder = {'categories': ${this.opts["categories"] ?? void 0}, 'dtype': ${this.opts["dtype"] ?? void 0}, 'handle_unknown': ${this.opts["handle_unknown"] ?? void 0}, 'unknown_value': ${this.opts["unknown_value"] ?? void 0}, 'encoded_missing_value': ${this.opts["encoded_missing_value"] ?? void 0}, 'min_frequency': ${this.opts["min_frequency"] ?? void 0}, 'max_categories': ${this.opts["max_categories"] ?? void 0}}

ctor_OrdinalEncoder = {k: v for k, v in ctor_OrdinalEncoder.items() if v is not None}`;
    await this._py.ex`bridgeOrdinalEncoder[${this.id}] = OrdinalEncoder(**ctor_OrdinalEncoder)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeOrdinalEncoder[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit the OrdinalEncoder to X.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OrdinalEncoder must call init() before fit()");
    }
    await this._py.ex`pms_OrdinalEncoder_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_OrdinalEncoder_fit = {k: v for k, v in pms_OrdinalEncoder_fit.items() if v is not None}`;
    await this._py.ex`res_OrdinalEncoder_fit = bridgeOrdinalEncoder[${this.id}].fit(**pms_OrdinalEncoder_fit)`;
    return this._py`res_OrdinalEncoder_fit.tolist() if hasattr(res_OrdinalEncoder_fit, 'tolist') else res_OrdinalEncoder_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OrdinalEncoder must call init() before fit_transform()");
    }
    await this._py.ex`pms_OrdinalEncoder_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_OrdinalEncoder_fit_transform = {k: v for k, v in pms_OrdinalEncoder_fit_transform.items() if v is not None}`;
    await this._py.ex`res_OrdinalEncoder_fit_transform = bridgeOrdinalEncoder[${this.id}].fit_transform(**pms_OrdinalEncoder_fit_transform)`;
    return this._py`res_OrdinalEncoder_fit_transform.tolist() if hasattr(res_OrdinalEncoder_fit_transform, 'tolist') else res_OrdinalEncoder_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrdinalEncoder must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_OrdinalEncoder_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_OrdinalEncoder_get_feature_names_out = {k: v for k, v in pms_OrdinalEncoder_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_OrdinalEncoder_get_feature_names_out = bridgeOrdinalEncoder[${this.id}].get_feature_names_out(**pms_OrdinalEncoder_get_feature_names_out)`;
    return this._py`res_OrdinalEncoder_get_feature_names_out.tolist() if hasattr(res_OrdinalEncoder_get_feature_names_out, 'tolist') else res_OrdinalEncoder_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrdinalEncoder must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_OrdinalEncoder_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_OrdinalEncoder_get_metadata_routing = {k: v for k, v in pms_OrdinalEncoder_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_OrdinalEncoder_get_metadata_routing = bridgeOrdinalEncoder[${this.id}].get_metadata_routing(**pms_OrdinalEncoder_get_metadata_routing)`;
    return this._py`res_OrdinalEncoder_get_metadata_routing.tolist() if hasattr(res_OrdinalEncoder_get_metadata_routing, 'tolist') else res_OrdinalEncoder_get_metadata_routing`;
  }
  /**
    Convert the data back to the original representation.
   */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrdinalEncoder must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_OrdinalEncoder_inverse_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_OrdinalEncoder_inverse_transform = {k: v for k, v in pms_OrdinalEncoder_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_OrdinalEncoder_inverse_transform = bridgeOrdinalEncoder[${this.id}].inverse_transform(**pms_OrdinalEncoder_inverse_transform)`;
    return this._py`res_OrdinalEncoder_inverse_transform.tolist() if hasattr(res_OrdinalEncoder_inverse_transform, 'tolist') else res_OrdinalEncoder_inverse_transform`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OrdinalEncoder must call init() before set_output()");
    }
    await this._py.ex`pms_OrdinalEncoder_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_OrdinalEncoder_set_output = {k: v for k, v in pms_OrdinalEncoder_set_output.items() if v is not None}`;
    await this._py.ex`res_OrdinalEncoder_set_output = bridgeOrdinalEncoder[${this.id}].set_output(**pms_OrdinalEncoder_set_output)`;
    return this._py`res_OrdinalEncoder_set_output.tolist() if hasattr(res_OrdinalEncoder_set_output, 'tolist') else res_OrdinalEncoder_set_output`;
  }
  /**
    Transform X to ordinal codes.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("OrdinalEncoder must call init() before transform()");
    }
    await this._py.ex`pms_OrdinalEncoder_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_OrdinalEncoder_transform = {k: v for k, v in pms_OrdinalEncoder_transform.items() if v is not None}`;
    await this._py.ex`res_OrdinalEncoder_transform = bridgeOrdinalEncoder[${this.id}].transform(**pms_OrdinalEncoder_transform)`;
    return this._py`res_OrdinalEncoder_transform.tolist() if hasattr(res_OrdinalEncoder_transform, 'tolist') else res_OrdinalEncoder_transform`;
  }
  /**
    The categories of each feature determined during `fit` (in order of the features in X and corresponding with the output of `transform`). This does not include categories that weren’t seen during `fit`.
   */
  get categories_() {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrdinalEncoder must call init() before accessing categories_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrdinalEncoder_categories_ = bridgeOrdinalEncoder[${this.id}].categories_`;
      return this._py`attr_OrdinalEncoder_categories_.tolist() if hasattr(attr_OrdinalEncoder_categories_, 'tolist') else attr_OrdinalEncoder_categories_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrdinalEncoder must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrdinalEncoder_n_features_in_ = bridgeOrdinalEncoder[${this.id}].n_features_in_`;
      return this._py`attr_OrdinalEncoder_n_features_in_.tolist() if hasattr(attr_OrdinalEncoder_n_features_in_, 'tolist') else attr_OrdinalEncoder_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This OrdinalEncoder instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "OrdinalEncoder must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_OrdinalEncoder_feature_names_in_ = bridgeOrdinalEncoder[${this.id}].feature_names_in_`;
      return this._py`attr_OrdinalEncoder_feature_names_in_.tolist() if hasattr(attr_OrdinalEncoder_feature_names_in_, 'tolist') else attr_OrdinalEncoder_feature_names_in_`;
    })();
  }
};
export {
  OrdinalEncoder
};
