// sklearn/build/generated/preprocessing/PowerTransformer.js
// src/generated/preprocessing/PowerTransformer.ts
import crypto from "node:crypto";
var PowerTransformer = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `PowerTransformer${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("PowerTransformer.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.preprocessing import PowerTransformer
try: bridgePowerTransformer
except NameError: bridgePowerTransformer = {}
`;
    await this._py.ex`ctor_PowerTransformer = {'method': ${this.opts["method"] ?? void 0}, 'standardize': ${this.opts["standardize"] ?? void 0}, 'copy': ${this.opts["copy"] ?? void 0}}

ctor_PowerTransformer = {k: v for k, v in ctor_PowerTransformer.items() if v is not None}`;
    await this._py.ex`bridgePowerTransformer[${this.id}] = PowerTransformer(**ctor_PowerTransformer)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgePowerTransformer[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Estimate the optimal parameter lambda for each feature.
  
      The optimal lambda parameter for minimizing skewness is estimated on each feature independently using maximum likelihood.
     */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("PowerTransformer must call init() before fit()");
    }
    await this._py.ex`pms_PowerTransformer_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_PowerTransformer_fit = {k: v for k, v in pms_PowerTransformer_fit.items() if v is not None}`;
    await this._py.ex`res_PowerTransformer_fit = bridgePowerTransformer[${this.id}].fit(**pms_PowerTransformer_fit)`;
    return this._py`res_PowerTransformer_fit.tolist() if hasattr(res_PowerTransformer_fit, 'tolist') else res_PowerTransformer_fit`;
  }
  /**
    Fit `PowerTransformer` to `X`, then transform `X`.
   */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "PowerTransformer must call init() before fit_transform()"
      );
    }
    await this._py.ex`pms_PowerTransformer_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_PowerTransformer_fit_transform = {k: v for k, v in pms_PowerTransformer_fit_transform.items() if v is not None}`;
    await this._py.ex`res_PowerTransformer_fit_transform = bridgePowerTransformer[${this.id}].fit_transform(**pms_PowerTransformer_fit_transform)`;
    return this._py`res_PowerTransformer_fit_transform.tolist() if hasattr(res_PowerTransformer_fit_transform, 'tolist') else res_PowerTransformer_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "PowerTransformer must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_PowerTransformer_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_PowerTransformer_get_feature_names_out = {k: v for k, v in pms_PowerTransformer_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_PowerTransformer_get_feature_names_out = bridgePowerTransformer[${this.id}].get_feature_names_out(**pms_PowerTransformer_get_feature_names_out)`;
    return this._py`res_PowerTransformer_get_feature_names_out.tolist() if hasattr(res_PowerTransformer_get_feature_names_out, 'tolist') else res_PowerTransformer_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "PowerTransformer must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_PowerTransformer_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_PowerTransformer_get_metadata_routing = {k: v for k, v in pms_PowerTransformer_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_PowerTransformer_get_metadata_routing = bridgePowerTransformer[${this.id}].get_metadata_routing(**pms_PowerTransformer_get_metadata_routing)`;
    return this._py`res_PowerTransformer_get_metadata_routing.tolist() if hasattr(res_PowerTransformer_get_metadata_routing, 'tolist') else res_PowerTransformer_get_metadata_routing`;
  }
  /**
      Apply the inverse power transformation using the fitted lambdas.
  
      The inverse of the Box-Cox transformation is given by:
     */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "PowerTransformer must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_PowerTransformer_inverse_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_PowerTransformer_inverse_transform = {k: v for k, v in pms_PowerTransformer_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_PowerTransformer_inverse_transform = bridgePowerTransformer[${this.id}].inverse_transform(**pms_PowerTransformer_inverse_transform)`;
    return this._py`res_PowerTransformer_inverse_transform.tolist() if hasattr(res_PowerTransformer_inverse_transform, 'tolist') else res_PowerTransformer_inverse_transform`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("PowerTransformer must call init() before set_output()");
    }
    await this._py.ex`pms_PowerTransformer_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_PowerTransformer_set_output = {k: v for k, v in pms_PowerTransformer_set_output.items() if v is not None}`;
    await this._py.ex`res_PowerTransformer_set_output = bridgePowerTransformer[${this.id}].set_output(**pms_PowerTransformer_set_output)`;
    return this._py`res_PowerTransformer_set_output.tolist() if hasattr(res_PowerTransformer_set_output, 'tolist') else res_PowerTransformer_set_output`;
  }
  /**
    Apply the power transform to each feature using the fitted lambdas.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("PowerTransformer must call init() before transform()");
    }
    await this._py.ex`pms_PowerTransformer_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_PowerTransformer_transform = {k: v for k, v in pms_PowerTransformer_transform.items() if v is not None}`;
    await this._py.ex`res_PowerTransformer_transform = bridgePowerTransformer[${this.id}].transform(**pms_PowerTransformer_transform)`;
    return this._py`res_PowerTransformer_transform.tolist() if hasattr(res_PowerTransformer_transform, 'tolist') else res_PowerTransformer_transform`;
  }
  /**
    The parameters of the power transformation for the selected features.
   */
  get lambdas_() {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "PowerTransformer must call init() before accessing lambdas_"
      );
    }
    return (async () => {
      await this._py.ex`attr_PowerTransformer_lambdas_ = bridgePowerTransformer[${this.id}].lambdas_`;
      return this._py`attr_PowerTransformer_lambdas_.tolist() if hasattr(attr_PowerTransformer_lambdas_, 'tolist') else attr_PowerTransformer_lambdas_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "PowerTransformer must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_PowerTransformer_n_features_in_ = bridgePowerTransformer[${this.id}].n_features_in_`;
      return this._py`attr_PowerTransformer_n_features_in_.tolist() if hasattr(attr_PowerTransformer_n_features_in_, 'tolist') else attr_PowerTransformer_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This PowerTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "PowerTransformer must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_PowerTransformer_feature_names_in_ = bridgePowerTransformer[${this.id}].feature_names_in_`;
      return this._py`attr_PowerTransformer_feature_names_in_.tolist() if hasattr(attr_PowerTransformer_feature_names_in_, 'tolist') else attr_PowerTransformer_feature_names_in_`;
    })();
  }
};
export {
  PowerTransformer
};
