// sklearn/build/generated/preprocessing/RobustScaler.js
// src/generated/preprocessing/RobustScaler.ts
import crypto from "node:crypto";
var RobustScaler = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `RobustScaler${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("RobustScaler.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.preprocessing import RobustScaler
try: bridgeRobustScaler
except NameError: bridgeRobustScaler = {}
`;
    await this._py.ex`ctor_RobustScaler = {'with_centering': ${this.opts["with_centering"] ?? void 0}, 'with_scaling': ${this.opts["with_scaling"] ?? void 0}, 'quantile_range': ${this.opts["quantile_range"] ?? void 0}, 'copy': ${this.opts["copy"] ?? void 0}, 'unit_variance': ${this.opts["unit_variance"] ?? void 0}}

ctor_RobustScaler = {k: v for k, v in ctor_RobustScaler.items() if v is not None}`;
    await this._py.ex`bridgeRobustScaler[${this.id}] = RobustScaler(**ctor_RobustScaler)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeRobustScaler[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Compute the median and quantiles to be used for scaling.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RobustScaler must call init() before fit()");
    }
    await this._py.ex`pms_RobustScaler_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_RobustScaler_fit = {k: v for k, v in pms_RobustScaler_fit.items() if v is not None}`;
    await this._py.ex`res_RobustScaler_fit = bridgeRobustScaler[${this.id}].fit(**pms_RobustScaler_fit)`;
    return this._py`res_RobustScaler_fit.tolist() if hasattr(res_RobustScaler_fit, 'tolist') else res_RobustScaler_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RobustScaler must call init() before fit_transform()");
    }
    await this._py.ex`pms_RobustScaler_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_RobustScaler_fit_transform = {k: v for k, v in pms_RobustScaler_fit_transform.items() if v is not None}`;
    await this._py.ex`res_RobustScaler_fit_transform = bridgeRobustScaler[${this.id}].fit_transform(**pms_RobustScaler_fit_transform)`;
    return this._py`res_RobustScaler_fit_transform.tolist() if hasattr(res_RobustScaler_fit_transform, 'tolist') else res_RobustScaler_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RobustScaler must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_RobustScaler_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_RobustScaler_get_feature_names_out = {k: v for k, v in pms_RobustScaler_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_RobustScaler_get_feature_names_out = bridgeRobustScaler[${this.id}].get_feature_names_out(**pms_RobustScaler_get_feature_names_out)`;
    return this._py`res_RobustScaler_get_feature_names_out.tolist() if hasattr(res_RobustScaler_get_feature_names_out, 'tolist') else res_RobustScaler_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RobustScaler must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_RobustScaler_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_RobustScaler_get_metadata_routing = {k: v for k, v in pms_RobustScaler_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_RobustScaler_get_metadata_routing = bridgeRobustScaler[${this.id}].get_metadata_routing(**pms_RobustScaler_get_metadata_routing)`;
    return this._py`res_RobustScaler_get_metadata_routing.tolist() if hasattr(res_RobustScaler_get_metadata_routing, 'tolist') else res_RobustScaler_get_metadata_routing`;
  }
  /**
    Scale back the data to the original representation.
   */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RobustScaler must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_RobustScaler_inverse_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_RobustScaler_inverse_transform = {k: v for k, v in pms_RobustScaler_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_RobustScaler_inverse_transform = bridgeRobustScaler[${this.id}].inverse_transform(**pms_RobustScaler_inverse_transform)`;
    return this._py`res_RobustScaler_inverse_transform.tolist() if hasattr(res_RobustScaler_inverse_transform, 'tolist') else res_RobustScaler_inverse_transform`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RobustScaler must call init() before set_output()");
    }
    await this._py.ex`pms_RobustScaler_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_RobustScaler_set_output = {k: v for k, v in pms_RobustScaler_set_output.items() if v is not None}`;
    await this._py.ex`res_RobustScaler_set_output = bridgeRobustScaler[${this.id}].set_output(**pms_RobustScaler_set_output)`;
    return this._py`res_RobustScaler_set_output.tolist() if hasattr(res_RobustScaler_set_output, 'tolist') else res_RobustScaler_set_output`;
  }
  /**
    Center and scale the data.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RobustScaler must call init() before transform()");
    }
    await this._py.ex`pms_RobustScaler_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_RobustScaler_transform = {k: v for k, v in pms_RobustScaler_transform.items() if v is not None}`;
    await this._py.ex`res_RobustScaler_transform = bridgeRobustScaler[${this.id}].transform(**pms_RobustScaler_transform)`;
    return this._py`res_RobustScaler_transform.tolist() if hasattr(res_RobustScaler_transform, 'tolist') else res_RobustScaler_transform`;
  }
  /**
    The median value for each feature in the training set.
   */
  get center_() {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RobustScaler must call init() before accessing center_");
    }
    return (async () => {
      await this._py.ex`attr_RobustScaler_center_ = bridgeRobustScaler[${this.id}].center_`;
      return this._py`attr_RobustScaler_center_.tolist() if hasattr(attr_RobustScaler_center_, 'tolist') else attr_RobustScaler_center_`;
    })();
  }
  /**
    The (scaled) interquartile range for each feature in the training set.
   */
  get scale_() {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RobustScaler must call init() before accessing scale_");
    }
    return (async () => {
      await this._py.ex`attr_RobustScaler_scale_ = bridgeRobustScaler[${this.id}].scale_`;
      return this._py`attr_RobustScaler_scale_.tolist() if hasattr(attr_RobustScaler_scale_, 'tolist') else attr_RobustScaler_scale_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RobustScaler must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_RobustScaler_n_features_in_ = bridgeRobustScaler[${this.id}].n_features_in_`;
      return this._py`attr_RobustScaler_n_features_in_.tolist() if hasattr(attr_RobustScaler_n_features_in_, 'tolist') else attr_RobustScaler_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This RobustScaler instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RobustScaler must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_RobustScaler_feature_names_in_ = bridgeRobustScaler[${this.id}].feature_names_in_`;
      return this._py`attr_RobustScaler_feature_names_in_.tolist() if hasattr(attr_RobustScaler_feature_names_in_, 'tolist') else attr_RobustScaler_feature_names_in_`;
    })();
  }
};
export {
  RobustScaler
};
