// sklearn/build/generated/feature_extraction/image/index.js
// src/generated/feature_extraction/image/PatchExtractor.ts
import crypto from "node:crypto";
var PatchExtractor = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `PatchExtractor${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This PatchExtractor instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("PatchExtractor.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.feature_extraction.image import PatchExtractor
try: bridgePatchExtractor
except NameError: bridgePatchExtractor = {}
`;
    await this._py.ex`ctor_PatchExtractor = {'patch_size': ${this.opts["patch_size"] ?? void 0}, 'max_patches': ${this.opts["max_patches"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}}

ctor_PatchExtractor = {k: v for k, v in ctor_PatchExtractor.items() if v is not None}`;
    await this._py.ex`bridgePatchExtractor[${this.id}] = PatchExtractor(**ctor_PatchExtractor)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgePatchExtractor[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Only validate the parameters of the estimator.
  
      This method allows to: (i) validate the parameters of the estimator and (ii) be consistent with the scikit-learn transformer API.
     */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This PatchExtractor instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("PatchExtractor must call init() before fit()");
    }
    await this._py.ex`pms_PatchExtractor_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_PatchExtractor_fit = {k: v for k, v in pms_PatchExtractor_fit.items() if v is not None}`;
    await this._py.ex`res_PatchExtractor_fit = bridgePatchExtractor[${this.id}].fit(**pms_PatchExtractor_fit)`;
    return this._py`res_PatchExtractor_fit.tolist() if hasattr(res_PatchExtractor_fit, 'tolist') else res_PatchExtractor_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This PatchExtractor instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("PatchExtractor must call init() before fit_transform()");
    }
    await this._py.ex`pms_PatchExtractor_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_PatchExtractor_fit_transform = {k: v for k, v in pms_PatchExtractor_fit_transform.items() if v is not None}`;
    await this._py.ex`res_PatchExtractor_fit_transform = bridgePatchExtractor[${this.id}].fit_transform(**pms_PatchExtractor_fit_transform)`;
    return this._py`res_PatchExtractor_fit_transform.tolist() if hasattr(res_PatchExtractor_fit_transform, 'tolist') else res_PatchExtractor_fit_transform`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This PatchExtractor instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "PatchExtractor must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_PatchExtractor_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_PatchExtractor_get_metadata_routing = {k: v for k, v in pms_PatchExtractor_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_PatchExtractor_get_metadata_routing = bridgePatchExtractor[${this.id}].get_metadata_routing(**pms_PatchExtractor_get_metadata_routing)`;
    return this._py`res_PatchExtractor_get_metadata_routing.tolist() if hasattr(res_PatchExtractor_get_metadata_routing, 'tolist') else res_PatchExtractor_get_metadata_routing`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error("This PatchExtractor instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("PatchExtractor must call init() before set_output()");
    }
    await this._py.ex`pms_PatchExtractor_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_PatchExtractor_set_output = {k: v for k, v in pms_PatchExtractor_set_output.items() if v is not None}`;
    await this._py.ex`res_PatchExtractor_set_output = bridgePatchExtractor[${this.id}].set_output(**pms_PatchExtractor_set_output)`;
    return this._py`res_PatchExtractor_set_output.tolist() if hasattr(res_PatchExtractor_set_output, 'tolist') else res_PatchExtractor_set_output`;
  }
  /**
    Transform the image samples in `X` into a matrix of patch data.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This PatchExtractor instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("PatchExtractor must call init() before transform()");
    }
    await this._py.ex`pms_PatchExtractor_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_PatchExtractor_transform = {k: v for k, v in pms_PatchExtractor_transform.items() if v is not None}`;
    await this._py.ex`res_PatchExtractor_transform = bridgePatchExtractor[${this.id}].transform(**pms_PatchExtractor_transform)`;
    return this._py`res_PatchExtractor_transform.tolist() if hasattr(res_PatchExtractor_transform, 'tolist') else res_PatchExtractor_transform`;
  }
};
export {
  PatchExtractor
};
