// sklearn/build/generated/feature_extraction/text/CountVectorizer.js
// src/generated/feature_extraction/text/CountVectorizer.ts
import crypto from "node:crypto";
var CountVectorizer = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `CountVectorizer${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("CountVectorizer.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
try: bridgeCountVectorizer
except NameError: bridgeCountVectorizer = {}
`;
    await this._py.ex`ctor_CountVectorizer = {'input': ${this.opts["input"] ?? void 0}, 'encoding': ${this.opts["encoding"] ?? void 0}, 'decode_error': ${this.opts["decode_error"] ?? void 0}, 'strip_accents': ${this.opts["strip_accents"] ?? void 0}, 'lowercase': ${this.opts["lowercase"] ?? void 0}, 'preprocessor': ${this.opts["preprocessor"] ?? void 0}, 'tokenizer': ${this.opts["tokenizer"] ?? void 0}, 'stop_words': ${this.opts["stop_words"] ?? void 0}, 'token_pattern': ${this.opts["token_pattern"] ?? void 0}, 'ngram_range': ${this.opts["ngram_range"] ?? void 0}, 'analyzer': ${this.opts["analyzer"] ?? void 0}, 'max_df': ${this.opts["max_df"] ?? void 0}, 'min_df': ${this.opts["min_df"] ?? void 0}, 'max_features': ${this.opts["max_features"] ?? void 0}, 'vocabulary': ${this.opts["vocabulary"] ?? void 0}, 'binary': ${this.opts["binary"] ?? void 0}, 'dtype': ${this.opts["dtype"] ?? void 0}}

ctor_CountVectorizer = {k: v for k, v in ctor_CountVectorizer.items() if v is not None}`;
    await this._py.ex`bridgeCountVectorizer[${this.id}] = CountVectorizer(**ctor_CountVectorizer)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeCountVectorizer[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Return a callable to process input data.
  
      The callable handles preprocessing, tokenization, and n-grams generation.
     */
  async build_analyzer(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before build_analyzer()"
      );
    }
    await this._py.ex`pms_CountVectorizer_build_analyzer = {}

pms_CountVectorizer_build_analyzer = {k: v for k, v in pms_CountVectorizer_build_analyzer.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_build_analyzer = bridgeCountVectorizer[${this.id}].build_analyzer(**pms_CountVectorizer_build_analyzer)`;
    return this._py`res_CountVectorizer_build_analyzer.tolist() if hasattr(res_CountVectorizer_build_analyzer, 'tolist') else res_CountVectorizer_build_analyzer`;
  }
  /**
    Return a function to preprocess the text before tokenization.
   */
  async build_preprocessor(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before build_preprocessor()"
      );
    }
    await this._py.ex`pms_CountVectorizer_build_preprocessor = {}

pms_CountVectorizer_build_preprocessor = {k: v for k, v in pms_CountVectorizer_build_preprocessor.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_build_preprocessor = bridgeCountVectorizer[${this.id}].build_preprocessor(**pms_CountVectorizer_build_preprocessor)`;
    return this._py`res_CountVectorizer_build_preprocessor.tolist() if hasattr(res_CountVectorizer_build_preprocessor, 'tolist') else res_CountVectorizer_build_preprocessor`;
  }
  /**
    Return a function that splits a string into a sequence of tokens.
   */
  async build_tokenizer(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before build_tokenizer()"
      );
    }
    await this._py.ex`pms_CountVectorizer_build_tokenizer = {}

pms_CountVectorizer_build_tokenizer = {k: v for k, v in pms_CountVectorizer_build_tokenizer.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_build_tokenizer = bridgeCountVectorizer[${this.id}].build_tokenizer(**pms_CountVectorizer_build_tokenizer)`;
    return this._py`res_CountVectorizer_build_tokenizer.tolist() if hasattr(res_CountVectorizer_build_tokenizer, 'tolist') else res_CountVectorizer_build_tokenizer`;
  }
  /**
      Decode the input into a string of unicode symbols.
  
      The decoding strategy depends on the vectorizer parameters.
     */
  async decode(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("CountVectorizer must call init() before decode()");
    }
    await this._py.ex`pms_CountVectorizer_decode = {'doc': ${opts["doc"] ?? void 0}}

pms_CountVectorizer_decode = {k: v for k, v in pms_CountVectorizer_decode.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_decode = bridgeCountVectorizer[${this.id}].decode(**pms_CountVectorizer_decode)`;
    return this._py`res_CountVectorizer_decode.tolist() if hasattr(res_CountVectorizer_decode, 'tolist') else res_CountVectorizer_decode`;
  }
  /**
    Learn a vocabulary dictionary of all tokens in the raw documents.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("CountVectorizer must call init() before fit()");
    }
    await this._py.ex`pms_CountVectorizer_fit = {'raw_documents': ${opts["raw_documents"] ?? void 0}, 'y': ${opts["y"] ?? void 0}}

pms_CountVectorizer_fit = {k: v for k, v in pms_CountVectorizer_fit.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_fit = bridgeCountVectorizer[${this.id}].fit(**pms_CountVectorizer_fit)`;
    return this._py`res_CountVectorizer_fit.tolist() if hasattr(res_CountVectorizer_fit, 'tolist') else res_CountVectorizer_fit`;
  }
  /**
      Learn the vocabulary dictionary and return document-term matrix.
  
      This is equivalent to fit followed by transform, but more efficiently implemented.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("CountVectorizer must call init() before fit_transform()");
    }
    await this._py.ex`pms_CountVectorizer_fit_transform = {'raw_documents': ${opts["raw_documents"] ?? void 0}, 'y': ${opts["y"] ?? void 0}}

pms_CountVectorizer_fit_transform = {k: v for k, v in pms_CountVectorizer_fit_transform.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_fit_transform = bridgeCountVectorizer[${this.id}].fit_transform(**pms_CountVectorizer_fit_transform)`;
    return this._py`res_CountVectorizer_fit_transform.tolist() if hasattr(res_CountVectorizer_fit_transform, 'tolist') else res_CountVectorizer_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_CountVectorizer_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_CountVectorizer_get_feature_names_out = {k: v for k, v in pms_CountVectorizer_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_get_feature_names_out = bridgeCountVectorizer[${this.id}].get_feature_names_out(**pms_CountVectorizer_get_feature_names_out)`;
    return this._py`res_CountVectorizer_get_feature_names_out.tolist() if hasattr(res_CountVectorizer_get_feature_names_out, 'tolist') else res_CountVectorizer_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_CountVectorizer_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_CountVectorizer_get_metadata_routing = {k: v for k, v in pms_CountVectorizer_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_get_metadata_routing = bridgeCountVectorizer[${this.id}].get_metadata_routing(**pms_CountVectorizer_get_metadata_routing)`;
    return this._py`res_CountVectorizer_get_metadata_routing.tolist() if hasattr(res_CountVectorizer_get_metadata_routing, 'tolist') else res_CountVectorizer_get_metadata_routing`;
  }
  /**
    Build or fetch the effective stop words list.
   */
  async get_stop_words(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before get_stop_words()"
      );
    }
    await this._py.ex`pms_CountVectorizer_get_stop_words = {}

pms_CountVectorizer_get_stop_words = {k: v for k, v in pms_CountVectorizer_get_stop_words.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_get_stop_words = bridgeCountVectorizer[${this.id}].get_stop_words(**pms_CountVectorizer_get_stop_words)`;
    return this._py`res_CountVectorizer_get_stop_words.tolist() if hasattr(res_CountVectorizer_get_stop_words, 'tolist') else res_CountVectorizer_get_stop_words`;
  }
  /**
    Return terms per document with nonzero entries in X.
   */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_CountVectorizer_inverse_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_CountVectorizer_inverse_transform = {k: v for k, v in pms_CountVectorizer_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_inverse_transform = bridgeCountVectorizer[${this.id}].inverse_transform(**pms_CountVectorizer_inverse_transform)`;
    return this._py`res_CountVectorizer_inverse_transform.tolist() if hasattr(res_CountVectorizer_inverse_transform, 'tolist') else res_CountVectorizer_inverse_transform`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before set_fit_request()"
      );
    }
    await this._py.ex`pms_CountVectorizer_set_fit_request = {'raw_documents': ${opts["raw_documents"] ?? void 0}}

pms_CountVectorizer_set_fit_request = {k: v for k, v in pms_CountVectorizer_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_set_fit_request = bridgeCountVectorizer[${this.id}].set_fit_request(**pms_CountVectorizer_set_fit_request)`;
    return this._py`res_CountVectorizer_set_fit_request.tolist() if hasattr(res_CountVectorizer_set_fit_request, 'tolist') else res_CountVectorizer_set_fit_request`;
  }
  /**
      Request metadata passed to the `transform` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_transform_request(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before set_transform_request()"
      );
    }
    await this._py.ex`pms_CountVectorizer_set_transform_request = {'raw_documents': ${opts["raw_documents"] ?? void 0}}

pms_CountVectorizer_set_transform_request = {k: v for k, v in pms_CountVectorizer_set_transform_request.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_set_transform_request = bridgeCountVectorizer[${this.id}].set_transform_request(**pms_CountVectorizer_set_transform_request)`;
    return this._py`res_CountVectorizer_set_transform_request.tolist() if hasattr(res_CountVectorizer_set_transform_request, 'tolist') else res_CountVectorizer_set_transform_request`;
  }
  /**
      Transform documents to document-term matrix.
  
      Extract token counts out of raw text documents using the vocabulary fitted with fit or the one provided to the constructor.
     */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("CountVectorizer must call init() before transform()");
    }
    await this._py.ex`pms_CountVectorizer_transform = {'raw_documents': ${opts["raw_documents"] ?? void 0}}

pms_CountVectorizer_transform = {k: v for k, v in pms_CountVectorizer_transform.items() if v is not None}`;
    await this._py.ex`res_CountVectorizer_transform = bridgeCountVectorizer[${this.id}].transform(**pms_CountVectorizer_transform)`;
    return this._py`res_CountVectorizer_transform.tolist() if hasattr(res_CountVectorizer_transform, 'tolist') else res_CountVectorizer_transform`;
  }
  /**
    A mapping of terms to feature indices.
   */
  get vocabulary_() {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before accessing vocabulary_"
      );
    }
    return (async () => {
      await this._py.ex`attr_CountVectorizer_vocabulary_ = bridgeCountVectorizer[${this.id}].vocabulary_`;
      return this._py`attr_CountVectorizer_vocabulary_.tolist() if hasattr(attr_CountVectorizer_vocabulary_, 'tolist') else attr_CountVectorizer_vocabulary_`;
    })();
  }
  /**
    True if a fixed vocabulary of term to indices mapping is provided by the user.
   */
  get fixed_vocabulary_() {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before accessing fixed_vocabulary_"
      );
    }
    return (async () => {
      await this._py.ex`attr_CountVectorizer_fixed_vocabulary_ = bridgeCountVectorizer[${this.id}].fixed_vocabulary_`;
      return this._py`attr_CountVectorizer_fixed_vocabulary_.tolist() if hasattr(attr_CountVectorizer_fixed_vocabulary_, 'tolist') else attr_CountVectorizer_fixed_vocabulary_`;
    })();
  }
  /**
    Terms that were ignored because they either:
   */
  get stop_words_() {
    if (this._isDisposed) {
      throw new Error("This CountVectorizer instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "CountVectorizer must call init() before accessing stop_words_"
      );
    }
    return (async () => {
      await this._py.ex`attr_CountVectorizer_stop_words_ = bridgeCountVectorizer[${this.id}].stop_words_`;
      return this._py`attr_CountVectorizer_stop_words_.tolist() if hasattr(attr_CountVectorizer_stop_words_, 'tolist') else attr_CountVectorizer_stop_words_`;
    })();
  }
};
export {
  CountVectorizer
};
