// sklearn/build/generated/feature_extraction/text/TfidfTransformer.js
// src/generated/feature_extraction/text/TfidfTransformer.ts
import crypto from "node:crypto";
var TfidfTransformer = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `TfidfTransformer${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("TfidfTransformer.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.feature_extraction.text import TfidfTransformer
try: bridgeTfidfTransformer
except NameError: bridgeTfidfTransformer = {}
`;
    await this._py.ex`ctor_TfidfTransformer = {'norm': ${this.opts["norm"] ?? void 0}, 'use_idf': ${this.opts["use_idf"] ?? void 0}, 'smooth_idf': ${this.opts["smooth_idf"] ?? void 0}, 'sublinear_tf': ${this.opts["sublinear_tf"] ?? void 0}}

ctor_TfidfTransformer = {k: v for k, v in ctor_TfidfTransformer.items() if v is not None}`;
    await this._py.ex`bridgeTfidfTransformer[${this.id}] = TfidfTransformer(**ctor_TfidfTransformer)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeTfidfTransformer[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Learn the idf vector (global term weights).
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("TfidfTransformer must call init() before fit()");
    }
    await this._py.ex`pms_TfidfTransformer_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}}

pms_TfidfTransformer_fit = {k: v for k, v in pms_TfidfTransformer_fit.items() if v is not None}`;
    await this._py.ex`res_TfidfTransformer_fit = bridgeTfidfTransformer[${this.id}].fit(**pms_TfidfTransformer_fit)`;
    return this._py`res_TfidfTransformer_fit.tolist() if hasattr(res_TfidfTransformer_fit, 'tolist') else res_TfidfTransformer_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "TfidfTransformer must call init() before fit_transform()"
      );
    }
    await this._py.ex`pms_TfidfTransformer_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_TfidfTransformer_fit_transform = {k: v for k, v in pms_TfidfTransformer_fit_transform.items() if v is not None}`;
    await this._py.ex`res_TfidfTransformer_fit_transform = bridgeTfidfTransformer[${this.id}].fit_transform(**pms_TfidfTransformer_fit_transform)`;
    return this._py`res_TfidfTransformer_fit_transform.tolist() if hasattr(res_TfidfTransformer_fit_transform, 'tolist') else res_TfidfTransformer_fit_transform`;
  }
  /**
    Get output feature names for transformation.
   */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "TfidfTransformer must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_TfidfTransformer_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_TfidfTransformer_get_feature_names_out = {k: v for k, v in pms_TfidfTransformer_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_TfidfTransformer_get_feature_names_out = bridgeTfidfTransformer[${this.id}].get_feature_names_out(**pms_TfidfTransformer_get_feature_names_out)`;
    return this._py`res_TfidfTransformer_get_feature_names_out.tolist() if hasattr(res_TfidfTransformer_get_feature_names_out, 'tolist') else res_TfidfTransformer_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "TfidfTransformer must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_TfidfTransformer_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_TfidfTransformer_get_metadata_routing = {k: v for k, v in pms_TfidfTransformer_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_TfidfTransformer_get_metadata_routing = bridgeTfidfTransformer[${this.id}].get_metadata_routing(**pms_TfidfTransformer_get_metadata_routing)`;
    return this._py`res_TfidfTransformer_get_metadata_routing.tolist() if hasattr(res_TfidfTransformer_get_metadata_routing, 'tolist') else res_TfidfTransformer_get_metadata_routing`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("TfidfTransformer must call init() before set_output()");
    }
    await this._py.ex`pms_TfidfTransformer_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_TfidfTransformer_set_output = {k: v for k, v in pms_TfidfTransformer_set_output.items() if v is not None}`;
    await this._py.ex`res_TfidfTransformer_set_output = bridgeTfidfTransformer[${this.id}].set_output(**pms_TfidfTransformer_set_output)`;
    return this._py`res_TfidfTransformer_set_output.tolist() if hasattr(res_TfidfTransformer_set_output, 'tolist') else res_TfidfTransformer_set_output`;
  }
  /**
      Request metadata passed to the `transform` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_transform_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "TfidfTransformer must call init() before set_transform_request()"
      );
    }
    await this._py.ex`pms_TfidfTransformer_set_transform_request = {'copy': ${opts["copy"] ?? void 0}}

pms_TfidfTransformer_set_transform_request = {k: v for k, v in pms_TfidfTransformer_set_transform_request.items() if v is not None}`;
    await this._py.ex`res_TfidfTransformer_set_transform_request = bridgeTfidfTransformer[${this.id}].set_transform_request(**pms_TfidfTransformer_set_transform_request)`;
    return this._py`res_TfidfTransformer_set_transform_request.tolist() if hasattr(res_TfidfTransformer_set_transform_request, 'tolist') else res_TfidfTransformer_set_transform_request`;
  }
  /**
    Transform a count matrix to a tf or tf-idf representation.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("TfidfTransformer must call init() before transform()");
    }
    await this._py.ex`pms_TfidfTransformer_transform = {'X': ${opts["X"] ?? void 0}, 'copy': ${opts["copy"] ?? void 0}}

pms_TfidfTransformer_transform = {k: v for k, v in pms_TfidfTransformer_transform.items() if v is not None}`;
    await this._py.ex`res_TfidfTransformer_transform = bridgeTfidfTransformer[${this.id}].transform(**pms_TfidfTransformer_transform)`;
    return this._py`res_TfidfTransformer_transform.tolist() if hasattr(res_TfidfTransformer_transform, 'tolist') else res_TfidfTransformer_transform`;
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "TfidfTransformer must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_TfidfTransformer_n_features_in_ = bridgeTfidfTransformer[${this.id}].n_features_in_`;
      return this._py`attr_TfidfTransformer_n_features_in_.tolist() if hasattr(attr_TfidfTransformer_n_features_in_, 'tolist') else attr_TfidfTransformer_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This TfidfTransformer instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "TfidfTransformer must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_TfidfTransformer_feature_names_in_ = bridgeTfidfTransformer[${this.id}].feature_names_in_`;
      return this._py`attr_TfidfTransformer_feature_names_in_.tolist() if hasattr(attr_TfidfTransformer_feature_names_in_, 'tolist') else attr_TfidfTransformer_feature_names_in_`;
    })();
  }
};
export {
  TfidfTransformer
};
