// sklearn/build/generated/kernel_approximation/SkewedChi2Sampler.js
// src/generated/kernel_approximation/SkewedChi2Sampler.ts
import crypto from "node:crypto";
var SkewedChi2Sampler = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `SkewedChi2Sampler${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("SkewedChi2Sampler.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.kernel_approximation import SkewedChi2Sampler
try: bridgeSkewedChi2Sampler
except NameError: bridgeSkewedChi2Sampler = {}
`;
    await this._py.ex`ctor_SkewedChi2Sampler = {'skewedness': ${this.opts["skewedness"] ?? void 0}, 'n_components': ${this.opts["n_components"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}}

ctor_SkewedChi2Sampler = {k: v for k, v in ctor_SkewedChi2Sampler.items() if v is not None}`;
    await this._py.ex`bridgeSkewedChi2Sampler[${this.id}] = SkewedChi2Sampler(**ctor_SkewedChi2Sampler)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeSkewedChi2Sampler[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Fit the model with X.
  
      Samples random projection according to n\_features.
     */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("SkewedChi2Sampler must call init() before fit()");
    }
    await this._py.ex`pms_SkewedChi2Sampler_fit = {'X': ${opts["X"] ?? void 0}, 'y': ${opts["y"] ?? void 0}}

pms_SkewedChi2Sampler_fit = {k: v for k, v in pms_SkewedChi2Sampler_fit.items() if v is not None}`;
    await this._py.ex`res_SkewedChi2Sampler_fit = bridgeSkewedChi2Sampler[${this.id}].fit(**pms_SkewedChi2Sampler_fit)`;
    return this._py`res_SkewedChi2Sampler_fit.tolist() if hasattr(res_SkewedChi2Sampler_fit, 'tolist') else res_SkewedChi2Sampler_fit`;
  }
  /**
      Fit to data, then transform it.
  
      Fits transformer to `X` and `y` with optional parameters `fit\_params` and returns a transformed version of `X`.
     */
  async fit_transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "SkewedChi2Sampler must call init() before fit_transform()"
      );
    }
    await this._py.ex`pms_SkewedChi2Sampler_fit_transform = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_SkewedChi2Sampler_fit_transform = {k: v for k, v in pms_SkewedChi2Sampler_fit_transform.items() if v is not None}`;
    await this._py.ex`res_SkewedChi2Sampler_fit_transform = bridgeSkewedChi2Sampler[${this.id}].fit_transform(**pms_SkewedChi2Sampler_fit_transform)`;
    return this._py`res_SkewedChi2Sampler_fit_transform.tolist() if hasattr(res_SkewedChi2Sampler_fit_transform, 'tolist') else res_SkewedChi2Sampler_fit_transform`;
  }
  /**
      Get output feature names for transformation.
  
      The feature names out will prefixed by the lowercased class name. For example, if the transformer outputs 3 features, then the feature names out are: `\["class\_name0", "class\_name1", "class\_name2"\]`.
     */
  async get_feature_names_out(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "SkewedChi2Sampler must call init() before get_feature_names_out()"
      );
    }
    await this._py.ex`pms_SkewedChi2Sampler_get_feature_names_out = {'input_features': ${opts["input_features"] ?? void 0}}

pms_SkewedChi2Sampler_get_feature_names_out = {k: v for k, v in pms_SkewedChi2Sampler_get_feature_names_out.items() if v is not None}`;
    await this._py.ex`res_SkewedChi2Sampler_get_feature_names_out = bridgeSkewedChi2Sampler[${this.id}].get_feature_names_out(**pms_SkewedChi2Sampler_get_feature_names_out)`;
    return this._py`res_SkewedChi2Sampler_get_feature_names_out.tolist() if hasattr(res_SkewedChi2Sampler_get_feature_names_out, 'tolist') else res_SkewedChi2Sampler_get_feature_names_out`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "SkewedChi2Sampler must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_SkewedChi2Sampler_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_SkewedChi2Sampler_get_metadata_routing = {k: v for k, v in pms_SkewedChi2Sampler_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_SkewedChi2Sampler_get_metadata_routing = bridgeSkewedChi2Sampler[${this.id}].get_metadata_routing(**pms_SkewedChi2Sampler_get_metadata_routing)`;
    return this._py`res_SkewedChi2Sampler_get_metadata_routing.tolist() if hasattr(res_SkewedChi2Sampler_get_metadata_routing, 'tolist') else res_SkewedChi2Sampler_get_metadata_routing`;
  }
  /**
      Set output container.
  
      See [Introducing the set\_output API](../../auto_examples/miscellaneous/plot_set_output.html#sphx-glr-auto-examples-miscellaneous-plot-set-output-py) for an example on how to use the API.
     */
  async set_output(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("SkewedChi2Sampler must call init() before set_output()");
    }
    await this._py.ex`pms_SkewedChi2Sampler_set_output = {'transform': ${opts["transform"] ?? void 0}}

pms_SkewedChi2Sampler_set_output = {k: v for k, v in pms_SkewedChi2Sampler_set_output.items() if v is not None}`;
    await this._py.ex`res_SkewedChi2Sampler_set_output = bridgeSkewedChi2Sampler[${this.id}].set_output(**pms_SkewedChi2Sampler_set_output)`;
    return this._py`res_SkewedChi2Sampler_set_output.tolist() if hasattr(res_SkewedChi2Sampler_set_output, 'tolist') else res_SkewedChi2Sampler_set_output`;
  }
  /**
    Apply the approximate feature map to X.
   */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("SkewedChi2Sampler must call init() before transform()");
    }
    await this._py.ex`pms_SkewedChi2Sampler_transform = {'X': ${opts["X"] ?? void 0}}

pms_SkewedChi2Sampler_transform = {k: v for k, v in pms_SkewedChi2Sampler_transform.items() if v is not None}`;
    await this._py.ex`res_SkewedChi2Sampler_transform = bridgeSkewedChi2Sampler[${this.id}].transform(**pms_SkewedChi2Sampler_transform)`;
    return this._py`res_SkewedChi2Sampler_transform.tolist() if hasattr(res_SkewedChi2Sampler_transform, 'tolist') else res_SkewedChi2Sampler_transform`;
  }
  /**
    Weight array, sampled from a secant hyperbolic distribution, which will be used to linearly transform the log of the data.
   */
  get random_weights_() {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "SkewedChi2Sampler must call init() before accessing random_weights_"
      );
    }
    return (async () => {
      await this._py.ex`attr_SkewedChi2Sampler_random_weights_ = bridgeSkewedChi2Sampler[${this.id}].random_weights_`;
      return this._py`attr_SkewedChi2Sampler_random_weights_.tolist() if hasattr(attr_SkewedChi2Sampler_random_weights_, 'tolist') else attr_SkewedChi2Sampler_random_weights_`;
    })();
  }
  /**
    Bias term, which will be added to the data. It is uniformly distributed between 0 and 2\*pi.
   */
  get random_offset_() {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "SkewedChi2Sampler must call init() before accessing random_offset_"
      );
    }
    return (async () => {
      await this._py.ex`attr_SkewedChi2Sampler_random_offset_ = bridgeSkewedChi2Sampler[${this.id}].random_offset_`;
      return this._py`attr_SkewedChi2Sampler_random_offset_.tolist() if hasattr(attr_SkewedChi2Sampler_random_offset_, 'tolist') else attr_SkewedChi2Sampler_random_offset_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "SkewedChi2Sampler must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_SkewedChi2Sampler_n_features_in_ = bridgeSkewedChi2Sampler[${this.id}].n_features_in_`;
      return this._py`attr_SkewedChi2Sampler_n_features_in_.tolist() if hasattr(attr_SkewedChi2Sampler_n_features_in_, 'tolist') else attr_SkewedChi2Sampler_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This SkewedChi2Sampler instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "SkewedChi2Sampler must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_SkewedChi2Sampler_feature_names_in_ = bridgeSkewedChi2Sampler[${this.id}].feature_names_in_`;
      return this._py`attr_SkewedChi2Sampler_feature_names_in_.tolist() if hasattr(attr_SkewedChi2Sampler_feature_names_in_, 'tolist') else attr_SkewedChi2Sampler_feature_names_in_`;
    })();
  }
};
export {
  SkewedChi2Sampler
};
