// sklearn/build/generated/gaussian_process/kernels/Matern.js
// src/generated/gaussian_process/kernels/Matern.ts
import crypto from "node:crypto";
var Matern = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `Matern${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This Matern instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("Matern.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.gaussian_process.kernels import Matern
try: bridgeMatern
except NameError: bridgeMatern = {}
`;
    await this._py.ex`ctor_Matern = {'length_scale': np.array(${this.opts["length_scale"] ?? void 0}) if ${this.opts["length_scale"] !== void 0} else None, 'length_scale_bounds': ${this.opts["length_scale_bounds"] ?? void 0}, 'nu': ${this.opts["nu"] ?? void 0}}

ctor_Matern = {k: v for k, v in ctor_Matern.items() if v is not None}`;
    await this._py.ex`bridgeMatern[${this.id}] = Matern(**ctor_Matern)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeMatern[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Return the kernel k(X, Y) and optionally its gradient.
   */
  async __call__(opts) {
    if (this._isDisposed) {
      throw new Error("This Matern instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Matern must call init() before __call__()");
    }
    await this._py.ex`pms_Matern___call__ = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'Y': np.array(${opts["Y"] ?? void 0}) if ${opts["Y"] !== void 0} else None, 'eval_gradient': ${opts["eval_gradient"] ?? void 0}}

pms_Matern___call__ = {k: v for k, v in pms_Matern___call__.items() if v is not None}`;
    await this._py.ex`res_Matern___call__ = bridgeMatern[${this.id}].__call__(**pms_Matern___call__)`;
    return this._py`res_Matern___call__.tolist() if hasattr(res_Matern___call__, 'tolist') else res_Matern___call__`;
  }
  /**
    Returns a clone of self with given hyperparameters theta.
   */
  async clone_with_theta(opts) {
    if (this._isDisposed) {
      throw new Error("This Matern instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Matern must call init() before clone_with_theta()");
    }
    await this._py.ex`pms_Matern_clone_with_theta = {'theta': np.array(${opts["theta"] ?? void 0}) if ${opts["theta"] !== void 0} else None}

pms_Matern_clone_with_theta = {k: v for k, v in pms_Matern_clone_with_theta.items() if v is not None}`;
    await this._py.ex`res_Matern_clone_with_theta = bridgeMatern[${this.id}].clone_with_theta(**pms_Matern_clone_with_theta)`;
    return this._py`res_Matern_clone_with_theta.tolist() if hasattr(res_Matern_clone_with_theta, 'tolist') else res_Matern_clone_with_theta`;
  }
  /**
      Returns the diagonal of the kernel k(X, X).
  
      The result of this method is identical to np.diag(self(X)); however, it can be evaluated more efficiently since only the diagonal is evaluated.
     */
  async diag(opts) {
    if (this._isDisposed) {
      throw new Error("This Matern instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Matern must call init() before diag()");
    }
    await this._py.ex`pms_Matern_diag = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_Matern_diag = {k: v for k, v in pms_Matern_diag.items() if v is not None}`;
    await this._py.ex`res_Matern_diag = bridgeMatern[${this.id}].diag(**pms_Matern_diag)`;
    return this._py`res_Matern_diag.tolist() if hasattr(res_Matern_diag, 'tolist') else res_Matern_diag`;
  }
  /**
    Returns whether the kernel is stationary.
   */
  async is_stationary(opts) {
    if (this._isDisposed) {
      throw new Error("This Matern instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Matern must call init() before is_stationary()");
    }
    await this._py.ex`pms_Matern_is_stationary = {}

pms_Matern_is_stationary = {k: v for k, v in pms_Matern_is_stationary.items() if v is not None}`;
    await this._py.ex`res_Matern_is_stationary = bridgeMatern[${this.id}].is_stationary(**pms_Matern_is_stationary)`;
    return this._py`res_Matern_is_stationary.tolist() if hasattr(res_Matern_is_stationary, 'tolist') else res_Matern_is_stationary`;
  }
  get anisotropic() {
    if (this._isDisposed) {
      throw new Error("This Matern instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Matern must call init() before accessing anisotropic");
    }
    return (async () => {
      await this._py.ex`attr_Matern_anisotropic = bridgeMatern[${this.id}].anisotropic`;
      return this._py`attr_Matern_anisotropic.tolist() if hasattr(attr_Matern_anisotropic, 'tolist') else attr_Matern_anisotropic`;
    })();
  }
  get hyperparameter_length_scale() {
    if (this._isDisposed) {
      throw new Error("This Matern instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "Matern must call init() before accessing hyperparameter_length_scale"
      );
    }
    return (async () => {
      await this._py.ex`attr_Matern_hyperparameter_length_scale = bridgeMatern[${this.id}].hyperparameter_length_scale`;
      return this._py`attr_Matern_hyperparameter_length_scale.tolist() if hasattr(attr_Matern_hyperparameter_length_scale, 'tolist') else attr_Matern_hyperparameter_length_scale`;
    })();
  }
};
export {
  Matern
};
