// sklearn/build/generated/gaussian_process/kernels/Product.js
// src/generated/gaussian_process/kernels/Product.ts
import crypto from "node:crypto";
var Product = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `Product${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This Product instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("Product.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.gaussian_process.kernels import Product
try: bridgeProduct
except NameError: bridgeProduct = {}
`;
    await this._py.ex`ctor_Product = {'k1': ${this.opts["k1"] ?? void 0}, 'k2': ${this.opts["k2"] ?? void 0}}

ctor_Product = {k: v for k, v in ctor_Product.items() if v is not None}`;
    await this._py.ex`bridgeProduct[${this.id}] = Product(**ctor_Product)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeProduct[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Return the kernel k(X, Y) and optionally its gradient.
   */
  async __call__(opts) {
    if (this._isDisposed) {
      throw new Error("This Product instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Product must call init() before __call__()");
    }
    await this._py.ex`pms_Product___call__ = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'Y': np.array(${opts["Y"] ?? void 0}) if ${opts["Y"] !== void 0} else None, 'eval_gradient': ${opts["eval_gradient"] ?? void 0}}

pms_Product___call__ = {k: v for k, v in pms_Product___call__.items() if v is not None}`;
    await this._py.ex`res_Product___call__ = bridgeProduct[${this.id}].__call__(**pms_Product___call__)`;
    return this._py`res_Product___call__.tolist() if hasattr(res_Product___call__, 'tolist') else res_Product___call__`;
  }
  /**
    Returns a clone of self with given hyperparameters theta.
   */
  async clone_with_theta(opts) {
    if (this._isDisposed) {
      throw new Error("This Product instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Product must call init() before clone_with_theta()");
    }
    await this._py.ex`pms_Product_clone_with_theta = {'theta': np.array(${opts["theta"] ?? void 0}) if ${opts["theta"] !== void 0} else None}

pms_Product_clone_with_theta = {k: v for k, v in pms_Product_clone_with_theta.items() if v is not None}`;
    await this._py.ex`res_Product_clone_with_theta = bridgeProduct[${this.id}].clone_with_theta(**pms_Product_clone_with_theta)`;
    return this._py`res_Product_clone_with_theta.tolist() if hasattr(res_Product_clone_with_theta, 'tolist') else res_Product_clone_with_theta`;
  }
  /**
      Returns the diagonal of the kernel k(X, X).
  
      The result of this method is identical to np.diag(self(X)); however, it can be evaluated more efficiently since only the diagonal is evaluated.
     */
  async diag(opts) {
    if (this._isDisposed) {
      throw new Error("This Product instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Product must call init() before diag()");
    }
    await this._py.ex`pms_Product_diag = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_Product_diag = {k: v for k, v in pms_Product_diag.items() if v is not None}`;
    await this._py.ex`res_Product_diag = bridgeProduct[${this.id}].diag(**pms_Product_diag)`;
    return this._py`res_Product_diag.tolist() if hasattr(res_Product_diag, 'tolist') else res_Product_diag`;
  }
  /**
    Returns whether the kernel is stationary.
   */
  async is_stationary(opts) {
    if (this._isDisposed) {
      throw new Error("This Product instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("Product must call init() before is_stationary()");
    }
    await this._py.ex`pms_Product_is_stationary = {}

pms_Product_is_stationary = {k: v for k, v in pms_Product_is_stationary.items() if v is not None}`;
    await this._py.ex`res_Product_is_stationary = bridgeProduct[${this.id}].is_stationary(**pms_Product_is_stationary)`;
    return this._py`res_Product_is_stationary.tolist() if hasattr(res_Product_is_stationary, 'tolist') else res_Product_is_stationary`;
  }
};
export {
  Product
};
