// sklearn/build/generated/naive_bayes/MultinomialNB.js
// src/generated/naive_bayes/MultinomialNB.ts
import crypto from "node:crypto";
var MultinomialNB = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `MultinomialNB${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("MultinomialNB.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.naive_bayes import MultinomialNB
try: bridgeMultinomialNB
except NameError: bridgeMultinomialNB = {}
`;
    await this._py.ex`ctor_MultinomialNB = {'alpha': np.array(${this.opts["alpha"] ?? void 0}) if ${this.opts["alpha"] !== void 0} else None, 'force_alpha': ${this.opts["force_alpha"] ?? void 0}, 'fit_prior': ${this.opts["fit_prior"] ?? void 0}, 'class_prior': np.array(${this.opts["class_prior"] ?? void 0}) if ${this.opts["class_prior"] !== void 0} else None}

ctor_MultinomialNB = {k: v for k, v in ctor_MultinomialNB.items() if v is not None}`;
    await this._py.ex`bridgeMultinomialNB[${this.id}] = MultinomialNB(**ctor_MultinomialNB)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeMultinomialNB[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Fit Naive Bayes classifier according to X, y.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("MultinomialNB must call init() before fit()");
    }
    await this._py.ex`pms_MultinomialNB_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_MultinomialNB_fit = {k: v for k, v in pms_MultinomialNB_fit.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_fit = bridgeMultinomialNB[${this.id}].fit(**pms_MultinomialNB_fit)`;
    return this._py`res_MultinomialNB_fit.tolist() if hasattr(res_MultinomialNB_fit, 'tolist') else res_MultinomialNB_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_MultinomialNB_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_MultinomialNB_get_metadata_routing = {k: v for k, v in pms_MultinomialNB_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_get_metadata_routing = bridgeMultinomialNB[${this.id}].get_metadata_routing(**pms_MultinomialNB_get_metadata_routing)`;
    return this._py`res_MultinomialNB_get_metadata_routing.tolist() if hasattr(res_MultinomialNB_get_metadata_routing, 'tolist') else res_MultinomialNB_get_metadata_routing`;
  }
  /**
      Incremental fit on a batch of samples.
  
      This method is expected to be called several times consecutively on different chunks of a dataset so as to implement out-of-core or online learning.
  
      This is especially useful when the whole dataset is too big to fit in memory at once.
  
      This method has some performance overhead hence it is better to call partial\_fit on chunks of data that are as large as possible (as long as fitting in the memory budget) to hide the overhead.
     */
  async partial_fit(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("MultinomialNB must call init() before partial_fit()");
    }
    await this._py.ex`pms_MultinomialNB_partial_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'classes': np.array(${opts["classes"] ?? void 0}) if ${opts["classes"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_MultinomialNB_partial_fit = {k: v for k, v in pms_MultinomialNB_partial_fit.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_partial_fit = bridgeMultinomialNB[${this.id}].partial_fit(**pms_MultinomialNB_partial_fit)`;
    return this._py`res_MultinomialNB_partial_fit.tolist() if hasattr(res_MultinomialNB_partial_fit, 'tolist') else res_MultinomialNB_partial_fit`;
  }
  /**
    Perform classification on an array of test vectors X.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("MultinomialNB must call init() before predict()");
    }
    await this._py.ex`pms_MultinomialNB_predict = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_MultinomialNB_predict = {k: v for k, v in pms_MultinomialNB_predict.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_predict = bridgeMultinomialNB[${this.id}].predict(**pms_MultinomialNB_predict)`;
    return this._py`res_MultinomialNB_predict.tolist() if hasattr(res_MultinomialNB_predict, 'tolist') else res_MultinomialNB_predict`;
  }
  /**
      Return joint log probability estimates for the test vector X.
  
      For each row x of X and class y, the joint log probability is given by `log P(x, y) \= log P(y) + log P(x|y),` where `log P(y)` is the class prior probability and `log P(x|y)` is the class-conditional probability.
     */
  async predict_joint_log_proba(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before predict_joint_log_proba()"
      );
    }
    await this._py.ex`pms_MultinomialNB_predict_joint_log_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_MultinomialNB_predict_joint_log_proba = {k: v for k, v in pms_MultinomialNB_predict_joint_log_proba.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_predict_joint_log_proba = bridgeMultinomialNB[${this.id}].predict_joint_log_proba(**pms_MultinomialNB_predict_joint_log_proba)`;
    return this._py`res_MultinomialNB_predict_joint_log_proba.tolist() if hasattr(res_MultinomialNB_predict_joint_log_proba, 'tolist') else res_MultinomialNB_predict_joint_log_proba`;
  }
  /**
    Return log-probability estimates for the test vector X.
   */
  async predict_log_proba(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before predict_log_proba()"
      );
    }
    await this._py.ex`pms_MultinomialNB_predict_log_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_MultinomialNB_predict_log_proba = {k: v for k, v in pms_MultinomialNB_predict_log_proba.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_predict_log_proba = bridgeMultinomialNB[${this.id}].predict_log_proba(**pms_MultinomialNB_predict_log_proba)`;
    return this._py`res_MultinomialNB_predict_log_proba.tolist() if hasattr(res_MultinomialNB_predict_log_proba, 'tolist') else res_MultinomialNB_predict_log_proba`;
  }
  /**
    Return probability estimates for the test vector X.
   */
  async predict_proba(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("MultinomialNB must call init() before predict_proba()");
    }
    await this._py.ex`pms_MultinomialNB_predict_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_MultinomialNB_predict_proba = {k: v for k, v in pms_MultinomialNB_predict_proba.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_predict_proba = bridgeMultinomialNB[${this.id}].predict_proba(**pms_MultinomialNB_predict_proba)`;
    return this._py`res_MultinomialNB_predict_proba.tolist() if hasattr(res_MultinomialNB_predict_proba, 'tolist') else res_MultinomialNB_predict_proba`;
  }
  /**
      Return the mean accuracy on the given test data and labels.
  
      In multi-label classification, this is the subset accuracy which is a harsh metric since you require for each sample that each label set be correctly predicted.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("MultinomialNB must call init() before score()");
    }
    await this._py.ex`pms_MultinomialNB_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_MultinomialNB_score = {k: v for k, v in pms_MultinomialNB_score.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_score = bridgeMultinomialNB[${this.id}].score(**pms_MultinomialNB_score)`;
    return this._py`res_MultinomialNB_score.tolist() if hasattr(res_MultinomialNB_score, 'tolist') else res_MultinomialNB_score`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("MultinomialNB must call init() before set_fit_request()");
    }
    await this._py.ex`pms_MultinomialNB_set_fit_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_MultinomialNB_set_fit_request = {k: v for k, v in pms_MultinomialNB_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_set_fit_request = bridgeMultinomialNB[${this.id}].set_fit_request(**pms_MultinomialNB_set_fit_request)`;
    return this._py`res_MultinomialNB_set_fit_request.tolist() if hasattr(res_MultinomialNB_set_fit_request, 'tolist') else res_MultinomialNB_set_fit_request`;
  }
  /**
      Request metadata passed to the `partial\_fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_partial_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before set_partial_fit_request()"
      );
    }
    await this._py.ex`pms_MultinomialNB_set_partial_fit_request = {'classes': ${opts["classes"] ?? void 0}, 'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_MultinomialNB_set_partial_fit_request = {k: v for k, v in pms_MultinomialNB_set_partial_fit_request.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_set_partial_fit_request = bridgeMultinomialNB[${this.id}].set_partial_fit_request(**pms_MultinomialNB_set_partial_fit_request)`;
    return this._py`res_MultinomialNB_set_partial_fit_request.tolist() if hasattr(res_MultinomialNB_set_partial_fit_request, 'tolist') else res_MultinomialNB_set_partial_fit_request`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before set_score_request()"
      );
    }
    await this._py.ex`pms_MultinomialNB_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_MultinomialNB_set_score_request = {k: v for k, v in pms_MultinomialNB_set_score_request.items() if v is not None}`;
    await this._py.ex`res_MultinomialNB_set_score_request = bridgeMultinomialNB[${this.id}].set_score_request(**pms_MultinomialNB_set_score_request)`;
    return this._py`res_MultinomialNB_set_score_request.tolist() if hasattr(res_MultinomialNB_set_score_request, 'tolist') else res_MultinomialNB_set_score_request`;
  }
  /**
    Number of samples encountered for each class during fitting. This value is weighted by the sample weight when provided.
   */
  get class_count_() {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before accessing class_count_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MultinomialNB_class_count_ = bridgeMultinomialNB[${this.id}].class_count_`;
      return this._py`attr_MultinomialNB_class_count_.tolist() if hasattr(attr_MultinomialNB_class_count_, 'tolist') else attr_MultinomialNB_class_count_`;
    })();
  }
  /**
    Smoothed empirical log probability for each class.
   */
  get class_log_prior_() {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before accessing class_log_prior_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MultinomialNB_class_log_prior_ = bridgeMultinomialNB[${this.id}].class_log_prior_`;
      return this._py`attr_MultinomialNB_class_log_prior_.tolist() if hasattr(attr_MultinomialNB_class_log_prior_, 'tolist') else attr_MultinomialNB_class_log_prior_`;
    })();
  }
  /**
    Class labels known to the classifier
   */
  get classes_() {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before accessing classes_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MultinomialNB_classes_ = bridgeMultinomialNB[${this.id}].classes_`;
      return this._py`attr_MultinomialNB_classes_.tolist() if hasattr(attr_MultinomialNB_classes_, 'tolist') else attr_MultinomialNB_classes_`;
    })();
  }
  /**
    Number of samples encountered for each (class, feature) during fitting. This value is weighted by the sample weight when provided.
   */
  get feature_count_() {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before accessing feature_count_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MultinomialNB_feature_count_ = bridgeMultinomialNB[${this.id}].feature_count_`;
      return this._py`attr_MultinomialNB_feature_count_.tolist() if hasattr(attr_MultinomialNB_feature_count_, 'tolist') else attr_MultinomialNB_feature_count_`;
    })();
  }
  /**
    Empirical log probability of features given a class, `P(x\_i|y)`.
   */
  get feature_log_prob_() {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before accessing feature_log_prob_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MultinomialNB_feature_log_prob_ = bridgeMultinomialNB[${this.id}].feature_log_prob_`;
      return this._py`attr_MultinomialNB_feature_log_prob_.tolist() if hasattr(attr_MultinomialNB_feature_log_prob_, 'tolist') else attr_MultinomialNB_feature_log_prob_`;
    })();
  }
  /**
    Number of features seen during [fit](../../glossary.html#term-fit).
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MultinomialNB_n_features_in_ = bridgeMultinomialNB[${this.id}].n_features_in_`;
      return this._py`attr_MultinomialNB_n_features_in_.tolist() if hasattr(attr_MultinomialNB_n_features_in_, 'tolist') else attr_MultinomialNB_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Defined only when `X` has feature names that are all strings.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This MultinomialNB instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "MultinomialNB must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_MultinomialNB_feature_names_in_ = bridgeMultinomialNB[${this.id}].feature_names_in_`;
      return this._py`attr_MultinomialNB_feature_names_in_.tolist() if hasattr(attr_MultinomialNB_feature_names_in_, 'tolist') else attr_MultinomialNB_feature_names_in_`;
    })();
  }
};
export {
  MultinomialNB
};
