// sklearn/build/generated/model_selection/GridSearchCV.js
// src/generated/model_selection/GridSearchCV.ts
import crypto from "node:crypto";
var GridSearchCV = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `GridSearchCV${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("GridSearchCV.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import GridSearchCV
try: bridgeGridSearchCV
except NameError: bridgeGridSearchCV = {}
`;
    await this._py.ex`ctor_GridSearchCV = {'estimator': ${this.opts["estimator"] ?? void 0}, 'param_grid': ${this.opts["param_grid"] ?? void 0}, 'scoring': ${this.opts["scoring"] ?? void 0}, 'n_jobs': ${this.opts["n_jobs"] ?? void 0}, 'refit': ${this.opts["refit"] ?? void 0}, 'cv': ${this.opts["cv"] ?? void 0}, 'verbose': ${this.opts["verbose"] ?? void 0}, 'pre_dispatch': ${this.opts["pre_dispatch"] ?? void 0}, 'error_score': ${this.opts["error_score"] ?? void 0}, 'return_train_score': ${this.opts["return_train_score"] ?? void 0}}

ctor_GridSearchCV = {k: v for k, v in ctor_GridSearchCV.items() if v is not None}`;
    await this._py.ex`bridgeGridSearchCV[${this.id}] = GridSearchCV(**ctor_GridSearchCV)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeGridSearchCV[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Call decision\_function on the estimator with the best found parameters.
  
      Only available if `refit=True` and the underlying estimator supports `decision\_function`.
     */
  async decision_function(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before decision_function()"
      );
    }
    await this._py.ex`pms_GridSearchCV_decision_function = {'X': ${opts["X"] ?? void 0}}

pms_GridSearchCV_decision_function = {k: v for k, v in pms_GridSearchCV_decision_function.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_decision_function = bridgeGridSearchCV[${this.id}].decision_function(**pms_GridSearchCV_decision_function)`;
    return this._py`res_GridSearchCV_decision_function.tolist() if hasattr(res_GridSearchCV_decision_function, 'tolist') else res_GridSearchCV_decision_function`;
  }
  /**
    Run fit with all sets of parameters.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before fit()");
    }
    await this._py.ex`pms_GridSearchCV_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'groups': np.array(${opts["groups"] ?? void 0}) if ${opts["groups"] !== void 0} else None, 'fit_params': ${opts["fit_params"] ?? void 0}}

pms_GridSearchCV_fit = {k: v for k, v in pms_GridSearchCV_fit.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_fit = bridgeGridSearchCV[${this.id}].fit(**pms_GridSearchCV_fit)`;
    return this._py`res_GridSearchCV_fit.tolist() if hasattr(res_GridSearchCV_fit, 'tolist') else res_GridSearchCV_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_GridSearchCV_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_GridSearchCV_get_metadata_routing = {k: v for k, v in pms_GridSearchCV_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_get_metadata_routing = bridgeGridSearchCV[${this.id}].get_metadata_routing(**pms_GridSearchCV_get_metadata_routing)`;
    return this._py`res_GridSearchCV_get_metadata_routing.tolist() if hasattr(res_GridSearchCV_get_metadata_routing, 'tolist') else res_GridSearchCV_get_metadata_routing`;
  }
  /**
      Call inverse\_transform on the estimator with the best found params.
  
      Only available if the underlying estimator implements `inverse\_transform` and `refit=True`.
     */
  async inverse_transform(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before inverse_transform()"
      );
    }
    await this._py.ex`pms_GridSearchCV_inverse_transform = {'Xt': ${opts["Xt"] ?? void 0}}

pms_GridSearchCV_inverse_transform = {k: v for k, v in pms_GridSearchCV_inverse_transform.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_inverse_transform = bridgeGridSearchCV[${this.id}].inverse_transform(**pms_GridSearchCV_inverse_transform)`;
    return this._py`res_GridSearchCV_inverse_transform.tolist() if hasattr(res_GridSearchCV_inverse_transform, 'tolist') else res_GridSearchCV_inverse_transform`;
  }
  /**
      Call predict on the estimator with the best found parameters.
  
      Only available if `refit=True` and the underlying estimator supports `predict`.
     */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before predict()");
    }
    await this._py.ex`pms_GridSearchCV_predict = {'X': ${opts["X"] ?? void 0}}

pms_GridSearchCV_predict = {k: v for k, v in pms_GridSearchCV_predict.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_predict = bridgeGridSearchCV[${this.id}].predict(**pms_GridSearchCV_predict)`;
    return this._py`res_GridSearchCV_predict.tolist() if hasattr(res_GridSearchCV_predict, 'tolist') else res_GridSearchCV_predict`;
  }
  /**
      Call predict\_log\_proba on the estimator with the best found parameters.
  
      Only available if `refit=True` and the underlying estimator supports `predict\_log\_proba`.
     */
  async predict_log_proba(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before predict_log_proba()"
      );
    }
    await this._py.ex`pms_GridSearchCV_predict_log_proba = {'X': ${opts["X"] ?? void 0}}

pms_GridSearchCV_predict_log_proba = {k: v for k, v in pms_GridSearchCV_predict_log_proba.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_predict_log_proba = bridgeGridSearchCV[${this.id}].predict_log_proba(**pms_GridSearchCV_predict_log_proba)`;
    return this._py`res_GridSearchCV_predict_log_proba.tolist() if hasattr(res_GridSearchCV_predict_log_proba, 'tolist') else res_GridSearchCV_predict_log_proba`;
  }
  /**
      Call predict\_proba on the estimator with the best found parameters.
  
      Only available if `refit=True` and the underlying estimator supports `predict\_proba`.
     */
  async predict_proba(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before predict_proba()");
    }
    await this._py.ex`pms_GridSearchCV_predict_proba = {'X': ${opts["X"] ?? void 0}}

pms_GridSearchCV_predict_proba = {k: v for k, v in pms_GridSearchCV_predict_proba.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_predict_proba = bridgeGridSearchCV[${this.id}].predict_proba(**pms_GridSearchCV_predict_proba)`;
    return this._py`res_GridSearchCV_predict_proba.tolist() if hasattr(res_GridSearchCV_predict_proba, 'tolist') else res_GridSearchCV_predict_proba`;
  }
  /**
      Return the score on the given data, if the estimator has been refit.
  
      This uses the score defined by `scoring` where provided, and the `best\_estimator\_.score` method otherwise.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before score()");
    }
    await this._py.ex`pms_GridSearchCV_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None}

pms_GridSearchCV_score = {k: v for k, v in pms_GridSearchCV_score.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_score = bridgeGridSearchCV[${this.id}].score(**pms_GridSearchCV_score)`;
    return this._py`res_GridSearchCV_score.tolist() if hasattr(res_GridSearchCV_score, 'tolist') else res_GridSearchCV_score`;
  }
  /**
      Call score\_samples on the estimator with the best found parameters.
  
      Only available if `refit=True` and the underlying estimator supports `score\_samples`.
     */
  async score_samples(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before score_samples()");
    }
    await this._py.ex`pms_GridSearchCV_score_samples = {'X': ${opts["X"] ?? void 0}}

pms_GridSearchCV_score_samples = {k: v for k, v in pms_GridSearchCV_score_samples.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_score_samples = bridgeGridSearchCV[${this.id}].score_samples(**pms_GridSearchCV_score_samples)`;
    return this._py`res_GridSearchCV_score_samples.tolist() if hasattr(res_GridSearchCV_score_samples, 'tolist') else res_GridSearchCV_score_samples`;
  }
  /**
      Request metadata passed to the `fit` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_fit_request(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before set_fit_request()");
    }
    await this._py.ex`pms_GridSearchCV_set_fit_request = {'groups': ${opts["groups"] ?? void 0}}

pms_GridSearchCV_set_fit_request = {k: v for k, v in pms_GridSearchCV_set_fit_request.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_set_fit_request = bridgeGridSearchCV[${this.id}].set_fit_request(**pms_GridSearchCV_set_fit_request)`;
    return this._py`res_GridSearchCV_set_fit_request.tolist() if hasattr(res_GridSearchCV_set_fit_request, 'tolist') else res_GridSearchCV_set_fit_request`;
  }
  /**
      Call transform on the estimator with the best found parameters.
  
      Only available if the underlying estimator supports `transform` and `refit=True`.
     */
  async transform(opts) {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before transform()");
    }
    await this._py.ex`pms_GridSearchCV_transform = {'X': ${opts["X"] ?? void 0}}

pms_GridSearchCV_transform = {k: v for k, v in pms_GridSearchCV_transform.items() if v is not None}`;
    await this._py.ex`res_GridSearchCV_transform = bridgeGridSearchCV[${this.id}].transform(**pms_GridSearchCV_transform)`;
    return this._py`res_GridSearchCV_transform.tolist() if hasattr(res_GridSearchCV_transform, 'tolist') else res_GridSearchCV_transform`;
  }
  /**
      A dict with keys as column headers and values as columns, that can be imported into a pandas `DataFrame`.
  
      For instance the below given table
     */
  get cv_results_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing cv_results_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_cv_results_ = bridgeGridSearchCV[${this.id}].cv_results_`;
      return this._py`attr_GridSearchCV_cv_results_.tolist() if hasattr(attr_GridSearchCV_cv_results_, 'tolist') else attr_GridSearchCV_cv_results_`;
    })();
  }
  /**
      Estimator that was chosen by the search, i.e. estimator which gave highest score (or smallest loss if specified) on the left out data. Not available if `refit=False`.
  
      See `refit` parameter for more information on allowed values.
     */
  get best_estimator_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing best_estimator_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_best_estimator_ = bridgeGridSearchCV[${this.id}].best_estimator_`;
      return this._py`attr_GridSearchCV_best_estimator_.tolist() if hasattr(attr_GridSearchCV_best_estimator_, 'tolist') else attr_GridSearchCV_best_estimator_`;
    })();
  }
  /**
      Mean cross-validated score of the best\_estimator
  
      For multi-metric evaluation, this is present only if `refit` is specified.
  
      This attribute is not available if `refit` is a function.
     */
  get best_score_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing best_score_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_best_score_ = bridgeGridSearchCV[${this.id}].best_score_`;
      return this._py`attr_GridSearchCV_best_score_.tolist() if hasattr(attr_GridSearchCV_best_score_, 'tolist') else attr_GridSearchCV_best_score_`;
    })();
  }
  /**
      Parameter setting that gave the best results on the hold out data.
  
      For multi-metric evaluation, this is present only if `refit` is specified.
     */
  get best_params_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing best_params_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_best_params_ = bridgeGridSearchCV[${this.id}].best_params_`;
      return this._py`attr_GridSearchCV_best_params_.tolist() if hasattr(attr_GridSearchCV_best_params_, 'tolist') else attr_GridSearchCV_best_params_`;
    })();
  }
  /**
      The index (of the `cv\_results\_` arrays) which corresponds to the best candidate parameter setting.
  
      The dict at `search.cv\_results\_\['params'\]\[search.best\_index\_\]` gives the parameter setting for the best model, that gives the highest mean score (`search.best\_score\_`).
  
      For multi-metric evaluation, this is present only if `refit` is specified.
     */
  get best_index_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing best_index_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_best_index_ = bridgeGridSearchCV[${this.id}].best_index_`;
      return this._py`attr_GridSearchCV_best_index_.tolist() if hasattr(attr_GridSearchCV_best_index_, 'tolist') else attr_GridSearchCV_best_index_`;
    })();
  }
  /**
      Scorer function used on the held out data to choose the best parameters for the model.
  
      For multi-metric evaluation, this attribute holds the validated `scoring` dict which maps the scorer key to the scorer callable.
     */
  get scorer_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("GridSearchCV must call init() before accessing scorer_");
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_scorer_ = bridgeGridSearchCV[${this.id}].scorer_`;
      return this._py`attr_GridSearchCV_scorer_.tolist() if hasattr(attr_GridSearchCV_scorer_, 'tolist') else attr_GridSearchCV_scorer_`;
    })();
  }
  /**
    The number of cross-validation splits (folds/iterations).
   */
  get n_splits_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing n_splits_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_n_splits_ = bridgeGridSearchCV[${this.id}].n_splits_`;
      return this._py`attr_GridSearchCV_n_splits_.tolist() if hasattr(attr_GridSearchCV_n_splits_, 'tolist') else attr_GridSearchCV_n_splits_`;
    })();
  }
  /**
      Seconds used for refitting the best model on the whole dataset.
  
      This is present only if `refit` is not `false`.
     */
  get refit_time_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing refit_time_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_refit_time_ = bridgeGridSearchCV[${this.id}].refit_time_`;
      return this._py`attr_GridSearchCV_refit_time_.tolist() if hasattr(attr_GridSearchCV_refit_time_, 'tolist') else attr_GridSearchCV_refit_time_`;
    })();
  }
  /**
    Whether or not the scorers compute several metrics.
   */
  get multimetric_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing multimetric_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_multimetric_ = bridgeGridSearchCV[${this.id}].multimetric_`;
      return this._py`attr_GridSearchCV_multimetric_.tolist() if hasattr(attr_GridSearchCV_multimetric_, 'tolist') else attr_GridSearchCV_multimetric_`;
    })();
  }
  /**
    Names of features seen during [fit](../../glossary.html#term-fit). Only defined if `best\_estimator\_` is defined (see the documentation for the `refit` parameter for more details) and that `best\_estimator\_` exposes `feature\_names\_in\_` when fit.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error("This GridSearchCV instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "GridSearchCV must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_GridSearchCV_feature_names_in_ = bridgeGridSearchCV[${this.id}].feature_names_in_`;
      return this._py`attr_GridSearchCV_feature_names_in_.tolist() if hasattr(attr_GridSearchCV_feature_names_in_, 'tolist') else attr_GridSearchCV_feature_names_in_`;
    })();
  }
};
export {
  GridSearchCV
};
