// sklearn/build/generated/model_selection/LeaveOneOut.js
// src/generated/model_selection/LeaveOneOut.ts
import crypto from "node:crypto";
var LeaveOneOut = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `LeaveOneOut${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This LeaveOneOut instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("LeaveOneOut.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import LeaveOneOut
try: bridgeLeaveOneOut
except NameError: bridgeLeaveOneOut = {}
`;
    await this._py.ex`ctor_LeaveOneOut = {'routing': ${this.opts["routing"] ?? void 0}}

ctor_LeaveOneOut = {k: v for k, v in ctor_LeaveOneOut.items() if v is not None}`;
    await this._py.ex`bridgeLeaveOneOut[${this.id}] = LeaveOneOut(**ctor_LeaveOneOut)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeLeaveOneOut[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This LeaveOneOut instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LeaveOneOut must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_LeaveOneOut_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_LeaveOneOut_get_metadata_routing = {k: v for k, v in pms_LeaveOneOut_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_LeaveOneOut_get_metadata_routing = bridgeLeaveOneOut[${this.id}].get_metadata_routing(**pms_LeaveOneOut_get_metadata_routing)`;
    return this._py`res_LeaveOneOut_get_metadata_routing.tolist() if hasattr(res_LeaveOneOut_get_metadata_routing, 'tolist') else res_LeaveOneOut_get_metadata_routing`;
  }
  /**
    Returns the number of splitting iterations in the cross-validator
   */
  async get_n_splits(opts) {
    if (this._isDisposed) {
      throw new Error("This LeaveOneOut instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LeaveOneOut must call init() before get_n_splits()");
    }
    await this._py.ex`pms_LeaveOneOut_get_n_splits = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}, 'groups': ${opts["groups"] ?? void 0}}

pms_LeaveOneOut_get_n_splits = {k: v for k, v in pms_LeaveOneOut_get_n_splits.items() if v is not None}`;
    await this._py.ex`res_LeaveOneOut_get_n_splits = bridgeLeaveOneOut[${this.id}].get_n_splits(**pms_LeaveOneOut_get_n_splits)`;
    return this._py`res_LeaveOneOut_get_n_splits.tolist() if hasattr(res_LeaveOneOut_get_n_splits, 'tolist') else res_LeaveOneOut_get_n_splits`;
  }
  /**
    Generate indices to split data into training and test set.
   */
  async split(opts) {
    if (this._isDisposed) {
      throw new Error("This LeaveOneOut instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LeaveOneOut must call init() before split()");
    }
    await this._py.ex`pms_LeaveOneOut_split = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'groups': np.array(${opts["groups"] ?? void 0}) if ${opts["groups"] !== void 0} else None}

pms_LeaveOneOut_split = {k: v for k, v in pms_LeaveOneOut_split.items() if v is not None}`;
    await this._py.ex`res_LeaveOneOut_split = bridgeLeaveOneOut[${this.id}].split(**pms_LeaveOneOut_split)`;
    return this._py`res_LeaveOneOut_split.tolist() if hasattr(res_LeaveOneOut_split, 'tolist') else res_LeaveOneOut_split`;
  }
};
export {
  LeaveOneOut
};
