// sklearn/build/generated/model_selection/LeavePOut.js
// src/generated/model_selection/LeavePOut.ts
import crypto from "node:crypto";
var LeavePOut = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `LeavePOut${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This LeavePOut instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("LeavePOut.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import LeavePOut
try: bridgeLeavePOut
except NameError: bridgeLeavePOut = {}
`;
    await this._py.ex`ctor_LeavePOut = {'p': ${this.opts["p"] ?? void 0}}

ctor_LeavePOut = {k: v for k, v in ctor_LeavePOut.items() if v is not None}`;
    await this._py.ex`bridgeLeavePOut[${this.id}] = LeavePOut(**ctor_LeavePOut)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeLeavePOut[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This LeavePOut instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "LeavePOut must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_LeavePOut_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_LeavePOut_get_metadata_routing = {k: v for k, v in pms_LeavePOut_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_LeavePOut_get_metadata_routing = bridgeLeavePOut[${this.id}].get_metadata_routing(**pms_LeavePOut_get_metadata_routing)`;
    return this._py`res_LeavePOut_get_metadata_routing.tolist() if hasattr(res_LeavePOut_get_metadata_routing, 'tolist') else res_LeavePOut_get_metadata_routing`;
  }
  /**
    Returns the number of splitting iterations in the cross-validator
   */
  async get_n_splits(opts) {
    if (this._isDisposed) {
      throw new Error("This LeavePOut instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LeavePOut must call init() before get_n_splits()");
    }
    await this._py.ex`pms_LeavePOut_get_n_splits = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': ${opts["y"] ?? void 0}, 'groups': ${opts["groups"] ?? void 0}}

pms_LeavePOut_get_n_splits = {k: v for k, v in pms_LeavePOut_get_n_splits.items() if v is not None}`;
    await this._py.ex`res_LeavePOut_get_n_splits = bridgeLeavePOut[${this.id}].get_n_splits(**pms_LeavePOut_get_n_splits)`;
    return this._py`res_LeavePOut_get_n_splits.tolist() if hasattr(res_LeavePOut_get_n_splits, 'tolist') else res_LeavePOut_get_n_splits`;
  }
  /**
    Generate indices to split data into training and test set.
   */
  async split(opts) {
    if (this._isDisposed) {
      throw new Error("This LeavePOut instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("LeavePOut must call init() before split()");
    }
    await this._py.ex`pms_LeavePOut_split = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'groups': np.array(${opts["groups"] ?? void 0}) if ${opts["groups"] !== void 0} else None}

pms_LeavePOut_split = {k: v for k, v in pms_LeavePOut_split.items() if v is not None}`;
    await this._py.ex`res_LeavePOut_split = bridgeLeavePOut[${this.id}].split(**pms_LeavePOut_split)`;
    return this._py`res_LeavePOut_split.tolist() if hasattr(res_LeavePOut_split, 'tolist') else res_LeavePOut_split`;
  }
};
export {
  LeavePOut
};
