// sklearn/build/generated/model_selection/LeaveOneGroupOut.js
// src/generated/model_selection/LeaveOneGroupOut.ts
import crypto from "node:crypto";
var LeaveOneGroupOut = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `LeaveOneGroupOut${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This LeaveOneGroupOut instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("LeaveOneGroupOut.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import LeaveOneGroupOut
try: bridgeLeaveOneGroupOut
except NameError: bridgeLeaveOneGroupOut = {}
`;
    await this._py.ex`ctor_LeaveOneGroupOut = {'routing': ${this.opts["routing"] ?? void 0}}

ctor_LeaveOneGroupOut = {k: v for k, v in ctor_LeaveOneGroupOut.items() if v is not None}`;
    await this._py.ex`bridgeLeaveOneGroupOut[${this.id}] = LeaveOneGroupOut(**ctor_LeaveOneGroupOut)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeLeaveOneGroupOut[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LeaveOneGroupOut instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LeaveOneGroupOut must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_LeaveOneGroupOut_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_LeaveOneGroupOut_get_metadata_routing = {k: v for k, v in pms_LeaveOneGroupOut_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_LeaveOneGroupOut_get_metadata_routing = bridgeLeaveOneGroupOut[${this.id}].get_metadata_routing(**pms_LeaveOneGroupOut_get_metadata_routing)`;
    return this._py`res_LeaveOneGroupOut_get_metadata_routing.tolist() if hasattr(res_LeaveOneGroupOut_get_metadata_routing, 'tolist') else res_LeaveOneGroupOut_get_metadata_routing`;
  }
  /**
    Returns the number of splitting iterations in the cross-validator
   */
  async get_n_splits(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LeaveOneGroupOut instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("LeaveOneGroupOut must call init() before get_n_splits()");
    }
    await this._py.ex`pms_LeaveOneGroupOut_get_n_splits = {'X': ${opts["X"] ?? void 0}, 'y': ${opts["y"] ?? void 0}, 'groups': np.array(${opts["groups"] ?? void 0}) if ${opts["groups"] !== void 0} else None}

pms_LeaveOneGroupOut_get_n_splits = {k: v for k, v in pms_LeaveOneGroupOut_get_n_splits.items() if v is not None}`;
    await this._py.ex`res_LeaveOneGroupOut_get_n_splits = bridgeLeaveOneGroupOut[${this.id}].get_n_splits(**pms_LeaveOneGroupOut_get_n_splits)`;
    return this._py`res_LeaveOneGroupOut_get_n_splits.tolist() if hasattr(res_LeaveOneGroupOut_get_n_splits, 'tolist') else res_LeaveOneGroupOut_get_n_splits`;
  }
  /**
      Request metadata passed to the `split` method.
  
      Note that this method is only relevant if `enable\_metadata\_routing=True` (see [`sklearn.set\_config`](sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_split_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LeaveOneGroupOut instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "LeaveOneGroupOut must call init() before set_split_request()"
      );
    }
    await this._py.ex`pms_LeaveOneGroupOut_set_split_request = {'groups': ${opts["groups"] ?? void 0}}

pms_LeaveOneGroupOut_set_split_request = {k: v for k, v in pms_LeaveOneGroupOut_set_split_request.items() if v is not None}`;
    await this._py.ex`res_LeaveOneGroupOut_set_split_request = bridgeLeaveOneGroupOut[${this.id}].set_split_request(**pms_LeaveOneGroupOut_set_split_request)`;
    return this._py`res_LeaveOneGroupOut_set_split_request.tolist() if hasattr(res_LeaveOneGroupOut_set_split_request, 'tolist') else res_LeaveOneGroupOut_set_split_request`;
  }
  /**
    Generate indices to split data into training and test set.
   */
  async split(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This LeaveOneGroupOut instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("LeaveOneGroupOut must call init() before split()");
    }
    await this._py.ex`pms_LeaveOneGroupOut_split = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'groups': np.array(${opts["groups"] ?? void 0}) if ${opts["groups"] !== void 0} else None}

pms_LeaveOneGroupOut_split = {k: v for k, v in pms_LeaveOneGroupOut_split.items() if v is not None}`;
    await this._py.ex`res_LeaveOneGroupOut_split = bridgeLeaveOneGroupOut[${this.id}].split(**pms_LeaveOneGroupOut_split)`;
    return this._py`res_LeaveOneGroupOut_split.tolist() if hasattr(res_LeaveOneGroupOut_split, 'tolist') else res_LeaveOneGroupOut_split`;
  }
};
export {
  LeaveOneGroupOut
};
