// sklearn/build/generated/model_selection/PredefinedSplit.js
// src/generated/model_selection/PredefinedSplit.ts
import crypto from "node:crypto";
var PredefinedSplit = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `PredefinedSplit${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This PredefinedSplit instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("PredefinedSplit.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import PredefinedSplit
try: bridgePredefinedSplit
except NameError: bridgePredefinedSplit = {}
`;
    await this._py.ex`ctor_PredefinedSplit = {'test_fold': np.array(${this.opts["test_fold"] ?? void 0}) if ${this.opts["test_fold"] !== void 0} else None}

ctor_PredefinedSplit = {k: v for k, v in ctor_PredefinedSplit.items() if v is not None}`;
    await this._py.ex`bridgePredefinedSplit[${this.id}] = PredefinedSplit(**ctor_PredefinedSplit)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgePredefinedSplit[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This PredefinedSplit instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "PredefinedSplit must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_PredefinedSplit_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_PredefinedSplit_get_metadata_routing = {k: v for k, v in pms_PredefinedSplit_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_PredefinedSplit_get_metadata_routing = bridgePredefinedSplit[${this.id}].get_metadata_routing(**pms_PredefinedSplit_get_metadata_routing)`;
    return this._py`res_PredefinedSplit_get_metadata_routing.tolist() if hasattr(res_PredefinedSplit_get_metadata_routing, 'tolist') else res_PredefinedSplit_get_metadata_routing`;
  }
  /**
    Returns the number of splitting iterations in the cross-validator
   */
  async get_n_splits(opts) {
    if (this._isDisposed) {
      throw new Error("This PredefinedSplit instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("PredefinedSplit must call init() before get_n_splits()");
    }
    await this._py.ex`pms_PredefinedSplit_get_n_splits = {'X': ${opts["X"] ?? void 0}, 'y': ${opts["y"] ?? void 0}, 'groups': ${opts["groups"] ?? void 0}}

pms_PredefinedSplit_get_n_splits = {k: v for k, v in pms_PredefinedSplit_get_n_splits.items() if v is not None}`;
    await this._py.ex`res_PredefinedSplit_get_n_splits = bridgePredefinedSplit[${this.id}].get_n_splits(**pms_PredefinedSplit_get_n_splits)`;
    return this._py`res_PredefinedSplit_get_n_splits.tolist() if hasattr(res_PredefinedSplit_get_n_splits, 'tolist') else res_PredefinedSplit_get_n_splits`;
  }
  /**
    Generate indices to split data into training and test set.
   */
  async split(opts) {
    if (this._isDisposed) {
      throw new Error("This PredefinedSplit instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("PredefinedSplit must call init() before split()");
    }
    await this._py.ex`pms_PredefinedSplit_split = {'X': ${opts["X"] ?? void 0}, 'y': ${opts["y"] ?? void 0}, 'groups': ${opts["groups"] ?? void 0}}

pms_PredefinedSplit_split = {k: v for k, v in pms_PredefinedSplit_split.items() if v is not None}`;
    await this._py.ex`res_PredefinedSplit_split = bridgePredefinedSplit[${this.id}].split(**pms_PredefinedSplit_split)`;
    return this._py`res_PredefinedSplit_split.tolist() if hasattr(res_PredefinedSplit_split, 'tolist') else res_PredefinedSplit_split`;
  }
};
export {
  PredefinedSplit
};
