// sklearn/build/generated/model_selection/RepeatedKFold.js
// src/generated/model_selection/RepeatedKFold.ts
import crypto from "node:crypto";
var RepeatedKFold = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `RepeatedKFold${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error("This RepeatedKFold instance has already been disposed");
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error("RepeatedKFold.init requires a PythonBridge instance");
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import RepeatedKFold
try: bridgeRepeatedKFold
except NameError: bridgeRepeatedKFold = {}
`;
    await this._py.ex`ctor_RepeatedKFold = {'n_splits': ${this.opts["n_splits"] ?? void 0}, 'n_repeats': ${this.opts["n_repeats"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}}

ctor_RepeatedKFold = {k: v for k, v in ctor_RepeatedKFold.items() if v is not None}`;
    await this._py.ex`bridgeRepeatedKFold[${this.id}] = RepeatedKFold(**ctor_RepeatedKFold)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeRepeatedKFold[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error("This RepeatedKFold instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error(
        "RepeatedKFold must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_RepeatedKFold_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_RepeatedKFold_get_metadata_routing = {k: v for k, v in pms_RepeatedKFold_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_RepeatedKFold_get_metadata_routing = bridgeRepeatedKFold[${this.id}].get_metadata_routing(**pms_RepeatedKFold_get_metadata_routing)`;
    return this._py`res_RepeatedKFold_get_metadata_routing.tolist() if hasattr(res_RepeatedKFold_get_metadata_routing, 'tolist') else res_RepeatedKFold_get_metadata_routing`;
  }
  /**
    Returns the number of splitting iterations in the cross-validator
   */
  async get_n_splits(opts) {
    if (this._isDisposed) {
      throw new Error("This RepeatedKFold instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RepeatedKFold must call init() before get_n_splits()");
    }
    await this._py.ex`pms_RepeatedKFold_get_n_splits = {'X': ${opts["X"] ?? void 0}, 'y': ${opts["y"] ?? void 0}, 'groups': np.array(${opts["groups"] ?? void 0}) if ${opts["groups"] !== void 0} else None}

pms_RepeatedKFold_get_n_splits = {k: v for k, v in pms_RepeatedKFold_get_n_splits.items() if v is not None}`;
    await this._py.ex`res_RepeatedKFold_get_n_splits = bridgeRepeatedKFold[${this.id}].get_n_splits(**pms_RepeatedKFold_get_n_splits)`;
    return this._py`res_RepeatedKFold_get_n_splits.tolist() if hasattr(res_RepeatedKFold_get_n_splits, 'tolist') else res_RepeatedKFold_get_n_splits`;
  }
  /**
    Generates indices to split data into training and test set.
   */
  async split(opts) {
    if (this._isDisposed) {
      throw new Error("This RepeatedKFold instance has already been disposed");
    }
    if (!this._isInitialized) {
      throw new Error("RepeatedKFold must call init() before split()");
    }
    await this._py.ex`pms_RepeatedKFold_split = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'groups': np.array(${opts["groups"] ?? void 0}) if ${opts["groups"] !== void 0} else None}

pms_RepeatedKFold_split = {k: v for k, v in pms_RepeatedKFold_split.items() if v is not None}`;
    await this._py.ex`res_RepeatedKFold_split = bridgeRepeatedKFold[${this.id}].split(**pms_RepeatedKFold_split)`;
    return this._py`res_RepeatedKFold_split.tolist() if hasattr(res_RepeatedKFold_split, 'tolist') else res_RepeatedKFold_split`;
  }
};
export {
  RepeatedKFold
};
