// sklearn/build/generated/model_selection/StratifiedShuffleSplit.js
// src/generated/model_selection/StratifiedShuffleSplit.ts
import crypto from "node:crypto";
var StratifiedShuffleSplit = class {
  constructor(opts) {
    this._isInitialized = false;
    this._isDisposed = false;
    this.id = `StratifiedShuffleSplit${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This StratifiedShuffleSplit instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error(
        "StratifiedShuffleSplit.init requires a PythonBridge instance"
      );
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import StratifiedShuffleSplit
try: bridgeStratifiedShuffleSplit
except NameError: bridgeStratifiedShuffleSplit = {}
`;
    await this._py.ex`ctor_StratifiedShuffleSplit = {'n_splits': ${this.opts["n_splits"] ?? void 0}, 'test_size': ${this.opts["test_size"] ?? void 0}, 'train_size': ${this.opts["train_size"] ?? void 0}, 'random_state': ${this.opts["random_state"] ?? void 0}}

ctor_StratifiedShuffleSplit = {k: v for k, v in ctor_StratifiedShuffleSplit.items() if v is not None}`;
    await this._py.ex`bridgeStratifiedShuffleSplit[${this.id}] = StratifiedShuffleSplit(**ctor_StratifiedShuffleSplit)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeStratifiedShuffleSplit[${this.id}]`;
    this._isDisposed = true;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This StratifiedShuffleSplit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "StratifiedShuffleSplit must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_StratifiedShuffleSplit_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_StratifiedShuffleSplit_get_metadata_routing = {k: v for k, v in pms_StratifiedShuffleSplit_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_StratifiedShuffleSplit_get_metadata_routing = bridgeStratifiedShuffleSplit[${this.id}].get_metadata_routing(**pms_StratifiedShuffleSplit_get_metadata_routing)`;
    return this._py`res_StratifiedShuffleSplit_get_metadata_routing.tolist() if hasattr(res_StratifiedShuffleSplit_get_metadata_routing, 'tolist') else res_StratifiedShuffleSplit_get_metadata_routing`;
  }
  /**
    Returns the number of splitting iterations in the cross-validator
   */
  async get_n_splits(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This StratifiedShuffleSplit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "StratifiedShuffleSplit must call init() before get_n_splits()"
      );
    }
    await this._py.ex`pms_StratifiedShuffleSplit_get_n_splits = {'X': ${opts["X"] ?? void 0}, 'y': ${opts["y"] ?? void 0}, 'groups': ${opts["groups"] ?? void 0}}

pms_StratifiedShuffleSplit_get_n_splits = {k: v for k, v in pms_StratifiedShuffleSplit_get_n_splits.items() if v is not None}`;
    await this._py.ex`res_StratifiedShuffleSplit_get_n_splits = bridgeStratifiedShuffleSplit[${this.id}].get_n_splits(**pms_StratifiedShuffleSplit_get_n_splits)`;
    return this._py`res_StratifiedShuffleSplit_get_n_splits.tolist() if hasattr(res_StratifiedShuffleSplit_get_n_splits, 'tolist') else res_StratifiedShuffleSplit_get_n_splits`;
  }
  /**
    Generate indices to split data into training and test set.
   */
  async split(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This StratifiedShuffleSplit instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("StratifiedShuffleSplit must call init() before split()");
    }
    await this._py.ex`pms_StratifiedShuffleSplit_split = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'groups': ${opts["groups"] ?? void 0}}

pms_StratifiedShuffleSplit_split = {k: v for k, v in pms_StratifiedShuffleSplit_split.items() if v is not None}`;
    await this._py.ex`res_StratifiedShuffleSplit_split = bridgeStratifiedShuffleSplit[${this.id}].split(**pms_StratifiedShuffleSplit_split)`;
    return this._py`res_StratifiedShuffleSplit_split.tolist() if hasattr(res_StratifiedShuffleSplit_split, 'tolist') else res_StratifiedShuffleSplit_split`;
  }
};
export {
  StratifiedShuffleSplit
};
